diff -urN --no-dereference nspr-clean/nspr/configure nspr-workdir/nspr/configure
--- nspr-clean/nspr/configure
+++ nspr-workdir/nspr/configure
@@ -7490,7 +7490,7 @@
     fi
     ;;
 
-*-linux*|*-gnu*|*-k*bsd*-gnu|*-android*|*-linuxandroid*)
+*-linux*|*-gnu*|*-k*bsd*-gnu|*-android*|*-linuxandroid*|*-mlibc)
     if test -z "$USE_NSPR_THREADS"; then
         USE_PTHREADS=1
         IMPL_STRATEGY=_PTH
diff -urN --no-dereference nspr-clean/nspr/pr/include/md/_pth.h nspr-workdir/nspr/pr/include/md/_pth.h
--- nspr-clean/nspr/pr/include/md/_pth.h
+++ nspr-workdir/nspr/pr/include/md/_pth.h
@@ -66,7 +66,7 @@
     || defined(HPUX) || defined(FREEBSD) \
     || defined(NETBSD) || defined(OPENBSD) \
     || defined(NTO) || defined(DARWIN) \
-    || defined(RISCOS)
+    || defined(RISCOS) || defined(__mlibc__)
 #define _PT_PTHREAD_INVALIDATE_THR_HANDLE(t)  (t) = 0
 #define _PT_PTHREAD_THR_HANDLE_IS_INVALID(t)  (t) == 0
 #define _PT_PTHREAD_COPY_THR_HANDLE(st, dt)   (dt) = (st)
@@ -109,7 +109,7 @@
 #define PT_PRIO_MIN            sched_get_priority_min(SCHED_OTHER)
 #define PT_PRIO_MAX            sched_get_priority_max(SCHED_OTHER)
 #elif defined(LINUX) || defined(__GNU__) || defined(__GLIBC__) \
-    || defined(FREEBSD)
+    || defined(FREEBSD) || defined(__mlibc__)
 #define PT_PRIO_MIN            sched_get_priority_min(SCHED_OTHER)
 #define PT_PRIO_MAX            sched_get_priority_max(SCHED_OTHER)
 #elif defined(NTO)
@@ -153,7 +153,7 @@
     || defined(LINUX) || defined(__GNU__) || defined(__GLIBC__) \
     || defined(FREEBSD) || defined(NETBSD) || defined(OPENBSD) \
     || defined(NTO) || defined(DARWIN) \
-    || defined(RISCOS)
+    || defined(RISCOS) || defined(__mlibc__)
 #define _PT_PTHREAD_YIELD()             sched_yield()
 #else
 #error "Need to define _PT_PTHREAD_YIELD for this platform"
diff -urN --no-dereference nspr-clean/nspr/pr/include/md/prosdep.h nspr-workdir/nspr/pr/include/md/prosdep.h
--- nspr-clean/nspr/pr/include/md/prosdep.h
+++ nspr-workdir/nspr/pr/include/md/prosdep.h
@@ -43,7 +43,7 @@
 #elif defined(HPUX)
 #include "md/_hpux.h"
 
-#elif defined(LINUX) || defined(__GNU__) || defined(__GLIBC__)
+#elif defined(LINUX) || defined(__GNU__) || defined(__GLIBC__) || defined(__mlibc__)
 #include "md/_linux.h"
 
 #elif defined(DARWIN)
diff -urN --no-dereference nspr-clean/nspr/pr/src/md/unix/unix.c nspr-workdir/nspr/pr/src/md/unix/unix.c
--- nspr-clean/nspr/pr/src/md/unix/unix.c
+++ nspr-workdir/nspr/pr/src/md/unix/unix.c
@@ -33,7 +33,7 @@
  * Make sure _PRSockLen_t is 32-bit, because we will cast a PRUint32* or
  * PRInt32* pointer to a _PRSockLen_t* pointer.
  */
-#if defined(HAVE_SOCKLEN_T) || (defined(__GLIBC__) && __GLIBC__ >= 2)
+#if defined(HAVE_SOCKLEN_T) || (defined(__GLIBC__) && __GLIBC__ >= 2) || defined(__mlibc__)
 #  define _PRSockLen_t socklen_t
 #elif defined(HPUX) || defined(SOLARIS) || defined(AIX4_1) || \
     defined(LINUX) || defined(DARWIN) || defined(QNX)
diff -urN --no-dereference nspr-clean/nspr/pr/src/md/unix/uxrng.c nspr-workdir/nspr/pr/src/md/unix/uxrng.c
--- nspr-clean/nspr/pr/src/md/unix/uxrng.c
+++ nspr-workdir/nspr/pr/src/md/unix/uxrng.c
@@ -52,7 +52,7 @@
 
 #elif (defined(LINUX) || defined(FREEBSD) || defined(__FreeBSD_kernel__) || \
        defined(NETBSD) || defined(__NetBSD_kernel__) || defined(OPENBSD) || \
-       defined(__GNU__))
+       defined(__GNU__) || defined(__mlibc__))
 #  include <sys/types.h>
 #  include <sys/stat.h>
 #  include <fcntl.h>
diff -urN --no-dereference nspr-clean/nspr/pr/src/pthreads/ptio.c nspr-workdir/nspr/pr/src/pthreads/ptio.c
--- nspr-clean/nspr/pr/src/pthreads/ptio.c
+++ nspr-workdir/nspr/pr/src/pthreads/ptio.c
@@ -149,7 +149,7 @@
 #    endif /* HAVE_SEND_FILE */
 #  endif   /* AIX */
 
-#  ifdef LINUX
+#  if defined(LINUX) && !defined(__ironclad__)
 #    include <sys/sendfile.h>
 #  endif
 
@@ -190,7 +190,7 @@
       defined(HPUX10_30) || defined(HPUX11) || defined(LINUX) ||    \
       defined(__GNU__) || defined(__GLIBC__) || defined(FREEBSD) || \
       defined(NETBSD) || defined(OPENBSD) || defined(NTO) ||        \
-      defined(DARWIN) || defined(RISCOS)
+      defined(DARWIN) || defined(RISCOS) || defined(__mlibc__)
 #    define _PRSelectFdSetArg_t fd_set*
 #  else
 #    error "Cannot determine architecture"
@@ -278,7 +278,7 @@
  * defines this type as socklen_t.  It is size_t or int on
  * most current systems.
  */
-#  if defined(HAVE_SOCKLEN_T) || (defined(__GLIBC__) && __GLIBC__ >= 2)
+#  if defined(HAVE_SOCKLEN_T) || (defined(__GLIBC__) && __GLIBC__ >= 2) || defined(__mlibc__)
 typedef socklen_t pt_SockLen;
 #  elif (defined(AIX) && !defined(AIX4_1))
 typedef PRSize pt_SockLen;
@@ -1090,6 +1090,11 @@
 
 #  ifdef LINUX
 static PRBool pt_linux_sendfile_cont(pt_Continuation* op, PRInt16 revents) {
+#ifdef __ironclad__
+  op->syserrno = ENOSYS;
+  op->result.code = -1;
+  return PR_TRUE;
+#else
   ssize_t rv;
   off_t oldoffset;
 
@@ -1111,6 +1116,7 @@
     return PR_FALSE;
   }
   return PR_TRUE;
+#endif
 }
 #  endif /* LINUX */
 
@@ -2641,8 +2647,13 @@
   if (file_nbytes_to_send != 0) {
     offset = sfd->file_offset;
     do {
+#ifndef __ironclad__
       rv = sendfile(sd->secret->md.osfd, sfd->fd->secret->md.osfd, &offset,
                     file_nbytes_to_send);
+#else
+      count = -1;
+      goto failed;
+#endif
     } while (rv == -1 && (syserrno = errno) == EINTR);
     if (rv == -1) {
       if (syserrno != EAGAIN && syserrno != EWOULDBLOCK) {
@@ -3276,7 +3287,7 @@
 #  if defined(HPUX) || defined(SOLARIS) || defined(LINUX) ||     \
       defined(__GNU__) || defined(__GLIBC__) || defined(AIX) ||  \
       defined(FREEBSD) || defined(NETBSD) || defined(OPENBSD) || \
-      defined(NTO) || defined(DARWIN) || defined(RISCOS)
+      defined(NTO) || defined(DARWIN) || defined(RISCOS) || defined(__mlibc__)
 #    define _PR_FCNTL_FLAGS O_NONBLOCK
 #  else
 #    error "Can't determine architecture"
