diff --git slim-clean/app.cpp slim-workdir/app.cpp
index d378793..632e48f 100644
--- slim-clean/app.cpp
+++ slim-workdir/app.cpp
@@ -1125,7 +1125,10 @@ int App::StartServer()
 	server[0] = (char *)cfg->getOption("default_xserver").c_str();
 	string argOption = cfg->getOption("xserver_arguments");
 	/* Add mandatory -xauth option */
-	argOption = argOption + " -auth " + cfg->getOption("authfile");
+    // Uncommenting the following line and thus passing auth to Xorg
+    // makes slim spawn Xorg in a loop, I dont know why, thus this
+    // hotfix.
+	// argOption = argOption + " -auth " + cfg->getOption("authfile");
 	char* args = new char[argOption.length()+2]; /* NULL plus vt */
 	strcpy(args, argOption.c_str());
 
@@ -1167,10 +1170,12 @@ int App::StartServer()
 		}
 	}
 
+	/*
 	if (!hasVtSet && daemonmode)
 	{
 		server[argc++] = (char*)"vt07";
 	}
+	*/
 	server[argc] = NULL;
 
 	switch (ServerPID) {
diff --git slim-clean/app.h slim-workdir/app.h
index 7453e34..a32dcb0 100644
--- slim-clean/app.h
+++ slim-workdir/app.h
@@ -14,6 +14,7 @@
 #define _APP_H_
 
 #include <X11/Xlib.h>
+#include <string>
 
 #ifdef USE_PAM
 #include "PAM.h"
