diff --git xf86-input-keyboard-clean/configure.ac xf86-input-keyboard-workdir/configure.ac
index 0f71ce3..9673a82 100644
--- xf86-input-keyboard-clean/configure.ac
+++ xf86-input-keyboard-workdir/configure.ac
@@ -57,6 +57,10 @@ AC_SUBST(inputdir)
 
 # The keyboard driver code is O/S specific
 case $host_os in
+  ironclad*)
+    IS_IRONCLAD="yes"
+    ;;
+
   linux*)
     AC_MSG_ERROR([This is not the keyboard driver you are looking for. Use evdev or libinput.])
     ;;
@@ -95,6 +99,7 @@ case $host_os in
 esac
 AC_SUBST([OS_FLAGS])
 
+AM_CONDITIONAL(IRONCLAD, [test "x$IS_IRONCLAD" = xyes])
 AM_CONDITIONAL(BSD, [test "x$IS_BSD" = xyes])
 AM_CONDITIONAL(SOLARIS, [test "x$IS_SOLARIS" = xyes])
 AM_CONDITIONAL(HURD, [test "x$IS_HURD" = xyes])
diff --git xf86-input-keyboard-clean/src/Makefile.am xf86-input-keyboard-workdir/src/Makefile.am
index 52f5f4d..45b0d0f 100644
--- xf86-input-keyboard-clean/src/Makefile.am
+++ xf86-input-keyboard-workdir/src/Makefile.am
@@ -26,10 +26,15 @@ kbd_drv_la_SOURCES = kbd.c xf86OSKbd.h xf86Keymap.h atKeynames.h
 kbd_drv_la_LIBADD = $(XORG_LIBS)
 kbd_drv_ladir = @inputdir@
 
+IRONCLAD_SRCS = ironclad_kbd.c at_scancode.c
 BSD_SRCS = bsd_KbdMap.c bsd_kbd.c bsd_kbd.h at_scancode.c
 HURD_SRCS = hurd_kbd.c at_scancode.c
 SOLARIS_SRCS = sun_kbd.c sun_kbd.h sun_kbdMap.c
 
+if IRONCLAD
+kbd_drv_la_SOURCES += $(IRONCLAD_SRCS)
+endif
+
 if BSD
 kbd_drv_la_SOURCES += $(BSD_SRCS)
 endif
@@ -42,4 +47,4 @@ if HURD
 kbd_drv_la_SOURCES += $(HURD_SRCS)
 endif
 
-EXTRA_DIST = $(BSD_SRCS) $(HURD_SRCS) $(SOLARIS_SRCS)
+EXTRA_DIST = $(IRONCLAD_SRCS) $(BSD_SRCS) $(HURD_SRCS) $(SOLARIS_SRCS)
diff --git xf86-input-keyboard-workdir/src/ironclad_kbd.c xf86-input-keyboard-workdir/src/ironclad_kbd.c
new file mode 100644
index 0000000..a473dbb
--- /dev/null
+++ xf86-input-keyboard-workdir/src/ironclad_kbd.c
@@ -0,0 +1,124 @@
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <X11/X.h>
+#include <xorg-server.h>
+
+#include "compiler.h"
+
+#include "xf86.h"
+#include "xf86Priv.h"
+#include "xf86_OSlib.h"
+
+#include "atKeynames.h"
+#include "xf86Keymap.h"
+#include "xf86OSKbd.h"
+#include "xf86Xinput.h"
+
+#include <assert.h>
+#include <errno.h>
+#include <stdio.h>
+#include <sys/file.h>
+#include <sys/ioctl.h>
+#include <sys/time.h>
+
+static int KbdOn(InputInfoPtr pInfo, int what) {
+    return Success;
+}
+
+static int KbdOff(InputInfoPtr pInfo, int what) {
+    printf("kbdOff is a stub!\n");
+    return -1;
+}
+
+static void SoundKbdBell(InputInfoPtr pInfo, int loudness, int pitch, int duration) {
+    printf("SoundKbdBell is a stub!\n");
+}
+
+static void SetKbdLeds(InputInfoPtr pInfo, int leds) {
+    printf("SetKbdLeds: is a stub!\n");
+}
+
+static int GetKbdLeds(InputInfoPtr pInfo) {
+    printf("GetKbdLeds is a stub!\n");
+    return -1;
+}
+
+// Save the initial keyboard state. This function is called at the start
+// of each server generation.
+static int KbdInit(InputInfoPtr pInfo, int what) {
+    return Success;
+}
+
+static void KbdGetMapping(InputInfoPtr pInfo, KeySymsPtr pKeySyms, CARD8 *pModMap) {
+    printf("KbdGetMapping is a stub!\n");
+}
+
+static void ReadInput(InputInfoPtr pInfo) {
+    KbdDevPtr pKbd = (KbdDevPtr) pInfo->private;
+
+    for(;;) {
+        uint8_t scancodes[2];
+        size_t result = read(pInfo->fd, scancodes, sizeof(scancodes));
+
+        if (result == 0 || result == -1) {
+            return;
+        }
+
+        if (result == 1) {
+            pKbd->PostEvent(pInfo, scancodes[0] & 0x7f /* scancode */, scancodes[0] & 0x80 ? FALSE : TRUE /* released */);
+        } else {
+            switch (scancodes[1]) {
+                case 0x48: pKbd->PostEvent(pInfo, KEY_Up, TRUE);     break;
+                case 0xC8: pKbd->PostEvent(pInfo, KEY_Up, FALSE);    break;
+                case 0x4B: pKbd->PostEvent(pInfo, KEY_Left, TRUE);   break;
+                case 0xCB: pKbd->PostEvent(pInfo, KEY_Left, FALSE);  break;
+                case 0x4D: pKbd->PostEvent(pInfo, KEY_Right, TRUE);  break;
+                case 0xCD: pKbd->PostEvent(pInfo, KEY_Right, FALSE); break;
+                case 0x50: pKbd->PostEvent(pInfo, KEY_Down, TRUE);   break;
+                case 0xD0: pKbd->PostEvent(pInfo, KEY_Down, FALSE);  break;
+            }
+        }
+    }
+}
+
+static Bool OpenKeyboard(InputInfoPtr pInfo) {
+    char *kbdPath = xf86SetStrOption(pInfo->options, "Device", "/dev/ps2keyboard");
+    Bool ret;
+
+    pInfo->fd = open(kbdPath, O_RDONLY | O_NONBLOCK);
+
+    if (pInfo->fd == -1) {
+        xf86Msg(X_ERROR, "%s: cannot open \"%s\"\n", pInfo->name, kbdPath);
+        ret = FALSE;
+    } else {
+        xf86MsgVerb(X_INFO, 3, "%s: opened device \"%s\"\n", pInfo->name, kbdPath);
+        pInfo->read_input = ReadInput;
+        ret = TRUE;
+
+        // in case it wasn't set and we fell back to default.
+        xf86ReplaceStrOption(pInfo->options, "Device", kbdPath);
+    }
+
+    free(kbdPath);
+    return ret;
+}
+
+Bool xf86OSKbdPreInit(InputInfoPtr pInfo) {
+    KbdDevPtr pKbd = pInfo->private;
+
+    pKbd->KbdInit       = KbdInit;
+    pKbd->KbdOn         = KbdOn;
+    pKbd->KbdOff        = KbdOff;
+    pKbd->Bell          = SoundKbdBell;
+    pKbd->SetLeds       = SetKbdLeds;
+    pKbd->GetLeds       = GetKbdLeds;
+    pKbd->KbdGetMapping = KbdGetMapping;
+    pKbd->OpenKeyboard = OpenKeyboard;
+
+    pKbd->RemapScanCode = NULL;
+    pKbd->private = NULL;
+
+    return TRUE;
+}
