diff --git openssl-clean/Configurations/10-main.conf openssl-workdir/Configurations/10-main.conf
index 7e93968..4de7b1f 100644
--- openssl-clean/Configurations/10-main.conf
+++ openssl-workdir/Configurations/10-main.conf
@@ -1010,6 +1010,51 @@ my %targets = (
         ranlib           => "true",
     },
 
+#### Ironclad
+    "ironclad-generic64" => {
+        inherit_from     => [ "BASE_unix" ],
+        CFLAGS           => picker(default   => "-Wall",
+                         debug       => "-O0 -g",
+                         release     => "-O3"),
+        CXXFLAGS         => picker(default   => "-Wall",
+                         debug       => "-O0 -g",
+                         release     => "-O3"),
+        cflags           => threads("-pthread"),
+        cxxflags         => combine("-std=c++11", threads("-pthread")),
+        lib_cppflags     => combine("-DOPENSSL_USE_NODELETE", "-DL_ENDIAN"),
+        bn_ops           => "SIXTY_FOUR_BIT_LONG RC4_CHAR",
+        thread_scheme    => "pthreads",
+        dso_scheme       => "dlfcn",
+        shared_target    => "linux-shared",
+        shared_cflag     => "-fPIC",
+        shared_ldflag    => sub { $disabled{pinshared} ? () : "-Wl,-znodelete" },
+        shared_extension => ".so.\$(SHLIB_VERSION_NUMBER)",
+    },
+
+    # These break the usual configuration naming convention but it's more
+    # convenient for us, since it matches the target triples
+    "x86_64-ironclad-gnu" => {
+        inherit_from     => [ "ironclad-generic64" ],
+        asm_arch         => 'x86_64',
+        perlasm_scheme   => "elf",
+
+        # Configure doesn't want to play nice and passes "--cross-compile-suffix"
+        # on to the GCC invocation
+        CC               => "x86_64-ironclad-gnu-gcc",
+        CXX              => "x86_64-ironclad-gnu-g++",
+
+    },
+    "riscv64-ironclad-gnu" => {
+        inherit_from     => [ "ironclad-generic64" ],
+        asm_arch         => 'riscv64',
+        perlasm_scheme   => "elf",
+
+        # Configure doesn't want to play nice and passes "--cross-compile-suffix"
+        # on to the GCC invocation
+        CC               => "riscv64-ironclad-gnu-gcc",
+        CXX              => "riscv64-ironclad-gnu-g++",
+    },
+
 #### *BSD
     "BSD-generic32" => {
         # As for thread cflag. Idea is to maintain "collective" set of
