diff --git libunistring-clean/lib/getlocalename_l-unsafe.c libunistring-workdir/lib/getlocalename_l-unsafe.c
index e273882..aaa3cfe 100644
--- libunistring-clean/lib/getlocalename_l-unsafe.c
+++ libunistring-workdir/lib/getlocalename_l-unsafe.c
@@ -473,6 +473,8 @@ getlocalename_l_unsafe (int category, locale_t locale)
       struct gl_locale_category_t *plc =
         &locale->category[gl_log2_lcmask_to_index (gl_log2_lc_mask (category))];
       return (struct string_with_storage) { plc->name, STORAGE_OBJECT };
+#elif defined __mlibc__
+      return (struct string_with_storage) { "C", STORAGE_INDEFINITE };
 #elif __GLIBC__ >= 2 && !defined __UCLIBC__
       /* Work around an incorrect definition of the _NL_LOCALE_NAME macro in
          glibc < 2.12.
