diff --git dav1d-clean/meson.build dav1d-workdir/meson.build
index 90555e4..e095fda 100644
--- dav1d-clean/meson.build
+++ dav1d-workdir/meson.build
@@ -96,7 +96,7 @@ test_args = []
 optional_arguments = []
 optional_link_arguments = []
 
-if host_machine.system() in ['linux', 'gnu', 'emscripten']
+if host_machine.system() in ['linux', 'gnu', 'ironclad', 'emscripten']
     test_args += '-D_GNU_SOURCE'
     add_project_arguments('-D_GNU_SOURCE', language: 'c')
 endif
