diff --git binutils-clean/bfd/config.in binutils-workdir/bfd/config.in
index 7458362..1d12c4d 100644
--- binutils-clean/bfd/config.in
+++ binutils-workdir/bfd/config.in
@@ -2,7 +2,7 @@
 
 /* Check that config.h is #included before system headers
     (this works only for glibc, but that should be enough).  */
-#if defined(__GLIBC__) && !defined(__FreeBSD_kernel__) && !defined(__CONFIG_H__)
+#if defined(__GLIBC__) && !defined(__FreeBSD_kernel__) && !defined(__ironclad__) && !defined(__CONFIG_H__)
 #  error config.h must be #included before system headers
 #endif
 #define __CONFIG_H__ 1
diff --git binutils-clean/binutils/config.in binutils-workdir/binutils/config.in
index 93ff046..8036d5c 100644
--- binutils-clean/binutils/config.in
+++ binutils-workdir/binutils/config.in
@@ -2,7 +2,7 @@
 
 /* Check that config.h is #included before system headers
    (this works only for glibc, but that should be enough).  */
-#if defined(__GLIBC__) && !defined(__FreeBSD_kernel__) && !defined(__CONFIG_H__)
+#if defined(__GLIBC__) && !defined(__FreeBSD_kernel__) && !defined(__ironclad__) && !defined(__CONFIG_H__)
 #  error config.h must be #included before system headers
 #endif
 #define __CONFIG_H__ 1
diff --git binutils-clean/gas/config.in binutils-workdir/gas/config.in
index bb89624..543eede 100644
--- binutils-clean/gas/config.in
+++ binutils-workdir/gas/config.in
@@ -2,7 +2,7 @@
 
 /* Check that config.h is #included before system headers
    (this works only for glibc, but that should be enough).  */
-#if defined(__GLIBC__) && !defined(__FreeBSD_kernel__) && !defined(__CONFIG_H__)
+#if defined(__GLIBC__) && !defined(__FreeBSD_kernel__) && !defined(__ironclad__) && !defined(__CONFIG_H__)
 #  error config.h must be #included before system headers
 #endif
 #define __CONFIG_H__ 1
diff --git binutils-clean/ld/config.in binutils-workdir/ld/config.in
index 790efd3..a8da3e4 100644
--- binutils-clean/ld/config.in
+++ binutils-workdir/ld/config.in
@@ -2,7 +2,7 @@
 
 /* Check that config.h is #included before system headers
    (this works only for glibc, but that should be enough).  */
-#if defined(__GLIBC__) && !defined(__FreeBSD_kernel__) && !defined(__CONFIG_H__)
+#if defined(__GLIBC__) && !defined(__FreeBSD_kernel__) && !defined(__ironclad__) && !defined(__CONFIG_H__)
 #  error config.h must be #included before system headers
 #endif
 #define __CONFIG_H__ 1
diff --git binutils-clean/opcodes/config.in binutils-workdir/opcodes/config.in
index 64b3303..ecd5c2b 100644
--- binutils-clean/opcodes/config.in
+++ binutils-workdir/opcodes/config.in
@@ -2,7 +2,7 @@
 
 /* Check that config.h is #included before system headers
    (this works only for glibc, but that should be enough).  */
-#if defined(__GLIBC__) && !defined(__FreeBSD_kernel__) && !defined(__CONFIG_H__)
+#if defined(__GLIBC__) && !defined(__FreeBSD_kernel__) && !defined(__ironclad__) && !defined(__CONFIG_H__)
 #  error config.h must be #included before system headers
 #endif
 #define __CONFIG_H__ 1
