diff --git fox-clean/configure fox-workdir/configure
index 9304716..d666a86 100755
--- fox-clean/configure
+++ fox-workdir/configure
@@ -18821,8 +18821,8 @@ fi
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $with_xft" >&5
 $as_echo "$with_xft" >&6; }
 if test "x$with_xft" != "xno"; then
-XFTCFLAGS="-I/usr/include/freetype2"
-XFTLIBS="-lfreetype -lfontconfig -lXft"
+XFTCFLAGS="$($PKG_CONFIG --cflags freetype2 fontconfig xft)"
+XFTLIBS="$($PKG_CONFIG --libs freetype2 fontconfig xft)"
 saved_cppflags="${CXXFLAGS}"
 CXXFLAGS="${CXXFLAGS} -DHAVE_XFT_H=1 $XFTCFLAGS"
 X_BASE_LIBS="${X_BASE_LIBS} $XFTLIBS"
diff --git fox-clean/include/FXStream.h fox-workdir/include/FXStream.h
index 41fe97a..e82638a 100644
--- fox-clean/include/FXStream.h
+++ fox-workdir/include/FXStream.h
@@ -52,9 +52,9 @@ enum FXStreamStatus {
 
 /// Stream seeking
 enum FXWhence {
-  FXFromStart=0,                /// Seek from start position
-  FXFromCurrent=1,              /// Seek from current position
-  FXFromEnd=2                   /// Seek from end position
+  FXFromStart=1,                /// Seek from start position
+  FXFromCurrent=2,              /// Seek from current position
+  FXFromEnd=4                   /// Seek from end position
   };
 
 
diff --git fox-clean/src/FXString.cpp fox-workdir/src/FXString.cpp
index c79ccf6..b1a28c5 100644
--- fox-clean/src/FXString.cpp
+++ fox-workdir/src/FXString.cpp
@@ -2501,18 +2501,18 @@ FXString FXStringVal(FXuint num,FXint base){
 
 
 // Formatting for reals
-static const char *const expo[]={"%.*f","%.*E","%.*G"};
+static const char *const expo[]={"%f","%f","%f"};
 
 
 // Conversion of float to string
 FXString FXStringVal(FXfloat num,FXint prec,FXint exp){
-  return FXStringFormat(expo[exp],prec,num);
+  return FXStringFormat("%f",num);
   }
 
 
 // Conversion of double to string
 FXString FXStringVal(FXdouble num,FXint prec,FXint exp){
-  return FXStringFormat(expo[exp],prec,num);
+  return FXStringFormat("%f",num);
   }
 
 
