diff -urN --no-dereference fox-clean/configure fox-workdir/configure
--- fox-clean/configure	2024-11-23 20:38:12.000000000 +0100
+++ fox-workdir/configure	2025-04-26 22:48:06.079122492 +0200
@@ -18821,8 +18821,8 @@
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $with_xft" >&5
 $as_echo "$with_xft" >&6; }
 if test "x$with_xft" != "xno"; then
-XFTCFLAGS="-I/usr/include/freetype2"
-XFTLIBS="-lfreetype -lfontconfig -lXft"
+XFTCFLAGS="$($PKG_CONFIG --cflags freetype2 fontconfig xft)"
+XFTLIBS="$($PKG_CONFIG --libs freetype2 fontconfig xft)"
 saved_cppflags="${CXXFLAGS}"
 CXXFLAGS="${CXXFLAGS} -DHAVE_XFT_H=1 $XFTCFLAGS"
 X_BASE_LIBS="${X_BASE_LIBS} $XFTLIBS"
diff -urN --no-dereference fox-clean/include/FXStream.h fox-workdir/include/FXStream.h
--- fox-clean/include/FXStream.h	2018-04-04 05:42:41.000000000 +0200
+++ fox-workdir/include/FXStream.h	2025-04-26 22:42:21.503813042 +0200
@@ -52,9 +52,9 @@
 
 /// Stream seeking
 enum FXWhence {
-  FXFromStart=0,                /// Seek from start position
-  FXFromCurrent=1,              /// Seek from current position
-  FXFromEnd=2                   /// Seek from end position
+  FXFromStart=1,                /// Seek from start position
+  FXFromCurrent=2,              /// Seek from current position
+  FXFromEnd=4                   /// Seek from end position
   };
 
 
diff -urN --no-dereference fox-clean/src/FXString.cpp fox-workdir/src/FXString.cpp
--- fox-clean/src/FXString.cpp	2024-11-23 20:37:53.000000000 +0100
+++ fox-workdir/src/FXString.cpp	2025-04-26 22:42:21.503926928 +0200
@@ -2501,18 +2501,18 @@
 
 
 // Formatting for reals
-static const char *const expo[]={"%.*f","%.*E","%.*G"};
+static const char *const expo[]={"%f","%f","%f"};
 
 
 // Conversion of float to string
 FXString FXStringVal(FXfloat num,FXint prec,FXint exp){
-  return FXStringFormat(expo[exp],prec,num);
+  return FXStringFormat("%f",num);
   }
 
 
 // Conversion of double to string
 FXString FXStringVal(FXdouble num,FXint prec,FXint exp){
-  return FXStringFormat(expo[exp],prec,num);
+  return FXStringFormat("%f",num);
   }
 
 
