diff -urN --no-dereference fox-clean/configure fox-workdir/configure
--- fox-clean/configure
+++ fox-workdir/configure
@@ -19480,7 +19480,7 @@
 if test "${with_freetype_include+set}" = set; then :
   withval=$with_freetype_include; CPPFLAGS="${CPPFLAGS} -I${with_freetype_include}"
 else
-  CPPFLAGS="${CPPFLAGS} -I/usr/include/freetype2"
+  CPPFLAGS="${CPPFLAGS} $($PKG_CONFIG --cflags freetype2 fontconfig xft)"
 fi
 
 ac_fn_cxx_check_header_mongrel "$LINENO" "ft2build.h" "ac_cv_header_ft2build_h" "$ac_includes_default"
diff -urN --no-dereference fox-clean/include/FXStream.h fox-workdir/include/FXStream.h
--- fox-clean/include/FXStream.h
+++ fox-workdir/include/FXStream.h
@@ -48,9 +48,9 @@
 
 /// Stream seeking
 enum FXWhence {
-  FXFromStart=0,                /// Seek from start position
-  FXFromCurrent=1,              /// Seek from current position
-  FXFromEnd=2                   /// Seek from end position
+  FXFromStart=1,                /// Seek from start position
+  FXFromCurrent=2,              /// Seek from current position
+  FXFromEnd=4                   /// Seek from end position
   };
 
 
