diff --git neofetch-clean/neofetch neofetch-workdir/neofetch
index 1e4b564..d9b8f6d 100755
--- neofetch-clean/neofetch
+++ neofetch-workdir/neofetch
@@ -929,6 +929,7 @@ get_os() {
         AIX)      os=AIX ;;
         IRIX*)    os=IRIX ;;
         FreeMiNT) os=FreeMiNT ;;
+        Ironclad) os=Ironclad ;;
 
         Linux|GNU*)
             os=Linux
@@ -954,6 +955,7 @@ get_distro() {
     [[ $distro ]] && return
 
     case $os in
+    	Ironclad) distro=Ironclad ;;
         Linux|BSD|MINIX)
             if [[ -f /bedrock/etc/bedrock-release && $PATH == */bedrock/cross/* ]]; then
                 case $distro_shorthand in
@@ -1387,6 +1389,9 @@ get_kernel() {
 get_uptime() {
     # Get uptime in seconds.
     case $os in
+        Ironclad)
+            s=$(uptime -s)
+        ;;
         Linux|Windows|MINIX)
             if [[ -r /proc/uptime ]]; then
                 s=$(< /proc/uptime)
@@ -2068,6 +2073,9 @@ get_wm_theme() {
 
 get_cpu() {
     case $os in
+        "Ironclad")
+            cpu="$(cpuinfo -n) ($(cpuinfo -c)) @ $(cpuinfo -f)"
+            ;;
         "Linux" | "MINIX" | "Windows")
             # Get CPU name.
             cpu_file="/proc/cpuinfo"
@@ -2646,6 +2654,10 @@ get_memory() {
             mem_used="$((mem_used / 1024))"
         ;;
 
+        "Ironclad")
+            mem_free="$(showmem -f)"
+            mem_used="$(showmem -u)"
+            mem_total="$(showmem -t)"
     esac
 
     [[ "$memory_percent" == "on" ]] && ((mem_perc=mem_used * 100 / mem_total))
@@ -3164,6 +3176,7 @@ get_term() {
     [[ "$WT_SESSION" ]]     && term="Windows Terminal"
 
     # Check $PPID for terminal emulator.
+    term="gcon"
     while [[ -z "$term" ]]; do
         parent="$(get_ppid "$parent")"
         [[ -z "$parent" ]] && break
@@ -5288,6 +5301,23 @@ get_distro_ascii() {
     #
     # $ascii_distro is the same as $distro.
     case $(trim "$ascii_distro") in
+        "Ironclad"*)
+            set_colors 5 7 0
+            read -rd '' ascii_data <<'EOF'
+${c3}
+                  &#BGPPPPPG#&
+               B5?77!!?YJJ7!7YBB&
+            &G5YJ77!7JYYYYYBPJ&PY#
+          #PYYYYYY?!?YYYYY7?7JP5JJ
+         B?YYYYYY7!!7JYYYYJ!!?JJJ5
+  &&    B7?J?77?7!!!!!77777!7Y5YYBBPGGG&
+ G77?YBB!!!!!!!!!!!!!JYJ??7JYJJY# PYPPG&
+ J777JB?!7JJ???!!!7?JYYYYYPJ!7JB
+ GYYG #JJJJJ??7!!!JYYY5PGB&GB&
+    #Y!?GB5YYJY5PG###&
+    GJJP
+EOF
+	;;
         "AIX"*)
             set_colors 2 7
             read -rd '' ascii_data <<'EOF'
