diff --git mesa-clean/include/drm-uapi/drm.h mesa-workdir/include/drm-uapi/drm.h
index 4e4f7c2..d41c0b5 100644
--- mesa-clean/include/drm-uapi/drm.h
+++ mesa-workdir/include/drm-uapi/drm.h
@@ -44,7 +44,7 @@ typedef unsigned int drm_handle_t;
 #else /* One of the BSDs */
 
 #include <stdint.h>
-#include <sys/ioccom.h>
+// #include <sys/ioccom.h>
 #include <sys/types.h>
 typedef int8_t   __s8;
 typedef uint8_t  __u8;
diff --git mesa-clean/include/renderdoc_app.h mesa-workdir/include/renderdoc_app.h
index d4a446c..9e8d4b4 100644
--- mesa-clean/include/renderdoc_app.h
+++ mesa-workdir/include/renderdoc_app.h
@@ -35,7 +35,7 @@
 
 #if defined(WIN32)
 #define RENDERDOC_CC __cdecl
-#elif defined(__linux__)
+#elif defined(__linux__) || defined(__ironclad__)
 #define RENDERDOC_CC
 #elif defined(__APPLE__)
 #define RENDERDOC_CC
diff --git mesa-clean/meson.build mesa-workdir/meson.build
index 427cfde..1bdff0c 100644
--- mesa-clean/meson.build
+++ mesa-workdir/meson.build
@@ -1013,7 +1013,7 @@ if cc.has_function('fmemopen')
 endif
 
 # TODO: this is very incomplete
-if ['linux', 'cygwin', 'gnu', 'freebsd', 'gnu/kfreebsd', 'haiku', 'android', 'managarm'].contains(host_machine.system())
+if ['linux', 'cygwin', 'gnu', 'freebsd', 'gnu/kfreebsd', 'haiku', 'android', 'managarm', 'ironclad'].contains(host_machine.system())
   pre_args += '-D_GNU_SOURCE'
 elif host_machine.system() == 'sunos'
   pre_args += '-D__EXTENSIONS__'
diff --git mesa-clean/src/gallium/auxiliary/gallivm/lp_bld_init_orc.cpp mesa-workdir/src/gallium/auxiliary/gallivm/lp_bld_init_orc.cpp
index 3d2b8cf..0be69b0 100644
--- mesa-clean/src/gallium/auxiliary/gallivm/lp_bld_init_orc.cpp
+++ mesa-workdir/src/gallium/auxiliary/gallivm/lp_bld_init_orc.cpp
@@ -340,7 +340,12 @@ LPJit::LPJit() :jit_dylib_count(0) {
          .setJITTargetMachineBuilder(std::move(JTMB))
 #ifdef USE_JITLINK
          .setObjectLinkingLayerCreator(
+#if LLVM_VERSION_MAJOR >= 21
+            /* LLVM 21 removed the Triple argument */
+            [&](ExecutionSession &ES) {
+#else
             [&](ExecutionSession &ES, const llvm::Triple &TT) {
+#endif
                return std::make_unique<ObjectLinkingLayer>(
                   ES, ExitOnErr(llvm::jitlink::InProcessMemoryManager::Create()));
             })
@@ -552,7 +557,7 @@ init_gallivm_state(struct gallivm_state *gallivm, const char *name,
    gallivm->cache = cache;
 
    gallivm->_ts_context = context->ref;
-   gallivm->context = LLVMOrcThreadSafeContextGetContext(context->ref);
+   gallivm->context = LLVMContextCreate();
 
    gallivm->module_name = LPJit::get_unique_name(name);
    gallivm->module = LLVMModuleCreateWithNameInContext(gallivm->module_name,
diff --git mesa-clean/src/gallium/drivers/llvmpipe/lp_texture.c mesa-workdir/src/gallium/drivers/llvmpipe/lp_texture.c
index f1f1baa..f064c50 100644
--- mesa-clean/src/gallium/drivers/llvmpipe/lp_texture.c
+++ mesa-workdir/src/gallium/drivers/llvmpipe/lp_texture.c
@@ -1754,7 +1754,7 @@ llvmpipe_resource_get_param(struct pipe_screen *screen,
    default:
       break;
    }
-   assert(0);
+   // assert(0);
    *value = 0;
    return false;
 }
diff --git mesa-clean/src/util/detect_os.h mesa-workdir/src/util/detect_os.h
index 86286df..802814f 100644
--- mesa-clean/src/util/detect_os.h
+++ mesa-workdir/src/util/detect_os.h
@@ -28,6 +28,12 @@
 #define DETECT_OS_ANDROID 1
 #endif
 
+#if defined(__ironclad__)
+#define DETECT_OS_IRONCLAD 1
+#define DETECT_OS_UNIX 1
+#define DETECT_OS_POSIX 1
+#endif
+
 #if defined(__FreeBSD__) || defined(__FreeBSD_kernel__)
 #define DETECT_OS_FREEBSD 1
 #define DETECT_OS_BSD 1
@@ -147,4 +153,8 @@
 #define DETECT_OS_MANAGARM 0
 #endif
 
+#ifndef DETECT_OS_IRONCLAD
+#define DETECT_OS_IRONCLAD 0
+#endif
+
 #endif /* DETECT_OS_H */
diff --git mesa-clean/src/util/os_misc.c mesa-workdir/src/util/os_misc.c
index 4b79166..09c16b0 100644
--- mesa-clean/src/util/os_misc.c
+++ mesa-workdir/src/util/os_misc.c
@@ -57,7 +57,7 @@
 #  include <unistd.h>
 #  include <log/log.h>
 #  include <cutils/properties.h>
-#elif DETECT_OS_LINUX || DETECT_OS_CYGWIN || DETECT_OS_SOLARIS || DETECT_OS_HURD || DETECT_OS_MANAGARM
+#elif DETECT_OS_LINUX || DETECT_OS_CYGWIN || DETECT_OS_SOLARIS || DETECT_OS_HURD || DETECT_OS_MANAGARM || DETECT_OS_IRONCLAD
 #  include <unistd.h>
 #elif DETECT_OS_OPENBSD || DETECT_OS_FREEBSD
 #  include <sys/resource.h>
@@ -284,7 +284,7 @@ exit_mutex:
 bool
 os_get_total_physical_memory(uint64_t *size)
 {
-#if DETECT_OS_LINUX || DETECT_OS_CYGWIN || DETECT_OS_SOLARIS || DETECT_OS_HURD || DETECT_OS_MANAGARM
+#if DETECT_OS_LINUX || DETECT_OS_CYGWIN || DETECT_OS_SOLARIS || DETECT_OS_HURD || DETECT_OS_MANAGARM || DETECT_OS_IRONCLAD
    const long phys_pages = sysconf(_SC_PHYS_PAGES);
    const long page_size = sysconf(_SC_PAGE_SIZE);
 
diff --git mesa-clean/src/util/os_time.c mesa-workdir/src/util/os_time.c
index da8ad7a..aa9aca4 100644
--- mesa-clean/src/util/os_time.c
+++ mesa-workdir/src/util/os_time.c
@@ -65,7 +65,7 @@ os_time_get_nano(void)
 void
 os_time_sleep(int64_t usecs)
 {
-#if DETECT_OS_LINUX || DETECT_OS_MANAGARM || DETECT_OS_FUCHSIA
+#if DETECT_OS_LINUX || DETECT_OS_MANAGARM || DETECT_OS_IRONCLAD || DETECT_OS_FUCHSIA
    struct timespec time;
    time.tv_sec = usecs / 1000000;
    time.tv_nsec = (usecs % 1000000) * 1000;
diff --git mesa-clean/src/util/u_thread.c mesa-workdir/src/util/u_thread.c
index 255c5f6..7c20f8d 100644
--- mesa-clean/src/util/u_thread.c
+++ mesa-workdir/src/util/u_thread.c
@@ -75,7 +75,7 @@ int u_thread_create(thrd_t *thrd, int (*routine)(void *), void *param)
 void u_thread_setname( const char *name )
 {
 #if defined(HAVE_PTHREAD)
-#if DETECT_OS_LINUX || DETECT_OS_CYGWIN || DETECT_OS_SOLARIS || defined(__GLIBC__) || DETECT_OS_MANAGARM || DETECT_OS_FUCHSIA
+#if DETECT_OS_LINUX || DETECT_OS_CYGWIN || DETECT_OS_SOLARIS || defined(__GLIBC__) || DETECT_OS_MANAGARM || DETECT_OS_IRONCLAD || DETECT_OS_FUCHSIA
    int ret = pthread_setname_np(pthread_self(), name);
    if (ret == ERANGE) {
       char buf[16];
@@ -154,7 +154,7 @@ util_set_thread_affinity(thrd_t thread,
 int64_t
 util_thread_get_time_nano(thrd_t thread)
 {
-#if defined(HAVE_PTHREAD) && !defined(__APPLE__) && !defined(__HAIKU__) && !defined(__managarm__)
+#if defined(HAVE_PTHREAD) && !defined(__APPLE__) && !defined(__HAIKU__) && !defined(__managarm__) && !defined(__ironclad__)
    struct timespec ts;
    clockid_t cid;
 
