diff --git fastfetch-clean/CMakeLists.txt fastfetch-workdir/CMakeLists.txt
index ac2315e..d0f2853 100644
--- fastfetch-clean/CMakeLists.txt
+++ fastfetch-workdir/CMakeLists.txt
@@ -45,6 +45,8 @@ elseif("${CMAKE_SYSTEM_NAME}" STREQUAL "Haiku")
     set(Haiku TRUE CACHE BOOL "..." FORCE)
 elseif("${CMAKE_SYSTEM_NAME}" STREQUAL "GNU")
     set(GNU TRUE CACHE BOOL "..." FORCE)
+elseif("${CMAKE_SYSTEM_NAME}" STREQUAL "Ironclad")
+    set(Ironclad TRUE CACHE BOOL "..." FORCE)
 elseif(NOT APPLE AND NOT WIN32)
     message(FATAL_ERROR "Unsupported platform: ${CMAKE_SYSTEM_NAME}")
 endif()
@@ -1364,6 +1366,79 @@ elseif(GNU)
         src/detection/wmtheme/wmtheme_linux.c
         src/detection/camera/camera_nosupport.c
    )
+elseif(Ironclad)
+    list(APPEND LIBFASTFETCH_SRC
+        src/common/impl/io_unix.c
+        src/common/impl/netif_gnu.c # This will fail, but else stuff doesn't link. Needs /proc
+        src/common/impl/networking_linux.c
+        src/common/impl/processing_linux.c
+        src/common/impl/FFPlatform_unix.c
+        src/common/impl/binary_linux.c
+        src/detection/battery/battery_nosupport.c
+        src/detection/bios/bios_nosupport.c
+        src/detection/board/board_nosupport.c
+        src/detection/bootmgr/bootmgr_nosupport.c
+        src/detection/brightness/brightness_nosupport.c
+        src/detection/btrfs/btrfs_nosupport.c
+        src/detection/chassis/chassis_nosupport.c
+        src/detection/cpu/cpu_ironclad.c # /proc/cpuinfo my beloved
+        src/detection/cpucache/cpucache_nosupport.c # sysfs
+        src/detection/cpuusage/cpuusage_nosupport.c # procfs
+        src/detection/cursor/cursor_linux.c
+        src/detection/physicaldisk/physicaldisk_nosupport.c # Sysfs
+        src/detection/physicalmemory/physicalmemory_nosupport.c # SMBIOS?
+        src/detection/diskio/diskio_nosupport.c # sysfs
+        src/detection/bluetooth/bluetooth_nosupport.c # We don't do bluetooth yet
+        src/detection/bluetoothradio/bluetoothradio_nosupport.c # We don't do bluetooth yet
+        src/detection/codec/codec_linux.c
+        src/detection/disk/disk_nosupport.c # sysfs and procfs
+        src/detection/dns/dns_linux.c
+        src/detection/displayserver/linux/displayserver_linux.c
+        src/detection/displayserver/linux/common.c
+        src/detection/displayserver/linux/drm.c
+        src/detection/displayserver/linux/wayland/wayland.c
+        src/detection/displayserver/linux/wmde.c
+        src/detection/displayserver/linux/xcb.c
+        src/detection/displayserver/linux/xlib.c
+        src/detection/font/font_linux.c
+        src/detection/icons/icons_linux.c
+        src/detection/gpu/gpu_nosupport.c # Needs sysfs and procfs?
+        src/detection/gtk_qt/gtk.c
+        src/detection/gtk_qt/qt.c
+        src/detection/host/host_nosupport.c
+        src/detection/initsystem/initsystem_linux.c # Misses support for epoch
+        src/detection/keyboard/keyboard_nosupport.c
+        src/detection/loadavg/loadavg_sunos.c # Uses getloadavg libc function, which we implement
+        src/detection/libc/libc_linux.c
+        src/detection/lm/lm_linux.c
+        src/detection/locale/locale_linux.c
+        src/detection/gamepad/gamepad_nosupport.c
+        src/detection/memory/memory_ironclad.c
+        src/detection/media/media_linux.c
+        src/detection/mouse/mouse_nosupport.c # No sysfs, maybe add custom solution
+        src/detection/netio/netio_nosupport.c # No sysfs
+        src/detection/opengl/opengl_linux.c
+        src/detection/os/os_linux.c
+        src/detection/packages/packages_linux.c
+        src/detection/packages/packages_nix.c
+        src/detection/terminalfont/terminalfont_linux.c
+        src/detection/terminalshell/terminalshell_linux.c
+        src/detection/terminalsize/terminalsize_linux.c
+        src/detection/poweradapter/poweradapter_nosupport.c
+        src/detection/processes/processes_nosupport.c # No procfs
+        src/detection/uptime/uptime_linux.c
+        src/detection/users/users_linux.c
+        src/detection/wallpaper/wallpaper_linux.c
+        src/detection/wifi/wifi_nosupport.c
+        src/detection/theme/theme_linux.c
+        src/detection/tpm/tpm_nosupport.c
+        src/detection/wm/wm_linux.c
+        src/detection/de/de_linux.c
+        src/detection/sound/sound_linux.c # Won't do anything as no pulseaudio
+        src/detection/swap/swap_nosupport.c # Reads procfs
+        src/detection/wmtheme/wmtheme_linux.c
+        src/detection/camera/camera_nosupport.c
+    )
 endif()
 
 if(ENABLE_TRACER)
@@ -1529,6 +1604,8 @@ elseif(NetBSD)
     set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,-rpath,/usr/X11R7/lib -Wl,-rpath,/usr/pkg/lib") # ditto
 elseif(Haiku)
     target_compile_definitions(libfastfetch PUBLIC _GNU_SOURCE)
+elseif(Ironclad)
+    target_compile_definitions(libfastfetch PUBLIC _GNU_SOURCE)
 elseif(GNU)
     # On Hurd PATH_MAX is not defined. Set an arbitrary limit as workaround.
     target_compile_definitions(libfastfetch PUBLIC _GNU_SOURCE PATH_MAX=4096 O_PATH=0)
@@ -1983,6 +2060,10 @@ elseif(Haiku)
         PRIVATE "be"
         PRIVATE "gnu"
     )
+elseif(Ironclad)
+    target_link_libraries(libfastfetch
+        PRIVATE "m"
+    )
 endif()
 
 target_include_directories(libfastfetch
diff --git fastfetch-clean/src/common/impl/FFPlatform_unix.c fastfetch-workdir/src/common/impl/FFPlatform_unix.c
index 3b9c87d..47df390 100644
--- fastfetch-clean/src/common/impl/FFPlatform_unix.c
+++ fastfetch-workdir/src/common/impl/FFPlatform_unix.c
@@ -142,6 +142,8 @@ static void getExePath(FFPlatform* platform) {
             break;
         }
     }
+#else
+    size_t exePathLen = 0;
 #endif
     if (exePathLen > 0) {
         ffStrbufEnsureFree(&platform->exePath, PATH_MAX);
diff --git fastfetch-workdir/src/detection/cpu/cpu_ironclad.c fastfetch-workdir/src/detection/cpu/cpu_ironclad.c
new file mode 100644
index 0000000..7ad301a
--- /dev/null
+++ fastfetch-workdir/src/detection/cpu/cpu_ironclad.c
@@ -0,0 +1,39 @@
+#include <sys/syscall.h>
+#include <unistd.h>
+#include "cpu.h"
+
+struct cpuinfo {
+    uint64_t conf_cores;
+    uint64_t onln_cores;
+    char model_name[64];
+    char vendor_name[64];
+    uint32_t base_mhz;
+    uint32_t max_mhz;
+    uint32_t ref_mhz;
+};
+
+const char* ffDetectCPUImpl(const FFCPUOptions* options, FFCPUResult* cpu_result) {
+	(void)options;  /* unused */
+	struct cpuinfo cpu;
+	syscall(SYS_getcpuinfo, &cpu);
+
+#if __i386__ || __x86_64__
+	ffCPUDetectByCpuid(cpu_result);
+#else
+	double base_frequency = ((double)cpu.base_mhz) / 1000;
+    double max_frequency = ((double)cpu.max_mhz) / 1000;
+	cpu_result->frequencyBase = base_frequency;
+    cpu_result->frequencyMax = max_frequency;
+#endif
+
+	ffStrbufAppendNS(&cpu_result->name, (uint32_t)strlen(cpu.model_name), cpu.model_name);
+	ffStrbufAppendNS(&cpu_result->vendor, (uint32_t)strlen(cpu.vendor_name), cpu.vendor_name);
+	cpu_result->coresOnline = (uint16_t)cpu.onln_cores;
+	// TODO: Differentiate between logical and physical
+	cpu_result->coresLogical = (uint16_t)cpu.conf_cores;
+	cpu_result->coresPhysical = (uint16_t)cpu.conf_cores;
+	// TODO: Add support for microarchitecture, packages and temperature
+	cpu_result->temperature = FF_CPU_TEMP_UNSET;
+
+	return NULL;
+}
diff --git fastfetch-clean/src/detection/loadavg/loadavg_sunos.c fastfetch-workdir/src/detection/loadavg/loadavg_sunos.c
index 91e3edd..5d88c67 100644
--- fastfetch-clean/src/detection/loadavg/loadavg_sunos.c
+++ fastfetch-workdir/src/detection/loadavg/loadavg_sunos.c
@@ -1,6 +1,10 @@
 #include "detection/loadavg/loadavg.h"
 
+#ifndef __ironclad__
 #include <sys/loadavg.h>
+#else
+#include <stdlib.h> // Provides getloadavg as a BSD extension
+#endif
 
 const char* ffDetectLoadavg(double result[3]) {
     return getloadavg(result, 3) == 3 ? NULL : "getloadavg() failed";
diff --git fastfetch-workdir/src/detection/memory/memory_ironclad.c fastfetch-workdir/src/detection/memory/memory_ironclad.c
new file mode 100644
index 0000000..23fd66a
--- /dev/null
+++ fastfetch-workdir/src/detection/memory/memory_ironclad.c
@@ -0,0 +1,10 @@
+#include "memory.h"
+#include <unistd.h>
+
+const char* ffDetectMemory(FFMemoryResult* ram) {
+    uint64_t pageSize = instance.state.platform.sysinfo.pageSize;
+
+    ram->bytesTotal = (uint64_t) sysconf(_SC_PHYS_PAGES) * pageSize;
+    ram->bytesUsed = ram->bytesTotal - (uint64_t) sysconf(_SC_AVPHYS_PAGES) * pageSize;
+    return NULL;
+}
diff --git fastfetch-clean/src/detection/packages/packages.h fastfetch-workdir/src/detection/packages/packages.h
index 2119349..1574082 100644
--- fastfetch-clean/src/detection/packages/packages.h
+++ fastfetch-workdir/src/detection/packages/packages.h
@@ -66,7 +66,7 @@ const char* ffDetectPackages(FFPackagesResult* result, FFPackagesOptions* option
 bool ffPackagesReadCache(FFstrbuf* cacheDir, FFstrbuf* cacheContent, const char* filePath, const char* packageId, uint32_t* result);
 bool ffPackagesWriteCache(FFstrbuf* cacheDir, FFstrbuf* cacheContent, uint32_t num_elements);
 
-#if defined(__linux__) || defined(__APPLE__) || defined(__GNU__)
+#if defined(__linux__) || defined(__APPLE__) || defined(__GNU__) || defined(__ironclad__)
 uint32_t ffPackagesGetNix(FFstrbuf* baseDir, const char* dirname);
 #endif
 #ifndef _WIN32
diff --git fastfetch-clean/src/detection/swap/swap_nosupport.c fastfetch-workdir/src/detection/swap/swap_nosupport.c
index ebc29b4..5a9b7e2 100644
--- fastfetch-clean/src/detection/swap/swap_nosupport.c
+++ fastfetch-workdir/src/detection/swap/swap_nosupport.c
@@ -1,5 +1,5 @@
 #include "swap.h"
 
-const char* ffDetectSwap(FFSwapResult* swap) {
+const char* ffDetectSwap(FFlist* swap) {
     return "Not supported on this platform";
 }
diff --git fastfetch-clean/src/detection/terminalsize/terminalsize_linux.c fastfetch-workdir/src/detection/terminalsize/terminalsize_linux.c
index e06f958..db050ee 100644
--- fastfetch-clean/src/detection/terminalsize/terminalsize_linux.c
+++ fastfetch-workdir/src/detection/terminalsize/terminalsize_linux.c
@@ -5,7 +5,7 @@
 #include <fcntl.h>
 #include <unistd.h>
 
-#ifdef __sun
+#if defined(__sun) || defined(__ironclad__)
     #include <sys/termios.h>
 #endif
 
diff --git fastfetch-clean/src/detection/uptime/uptime_linux.c fastfetch-workdir/src/detection/uptime/uptime_linux.c
index 75c7768..9739636 100644
--- fastfetch-clean/src/detection/uptime/uptime_linux.c
+++ fastfetch-workdir/src/detection/uptime/uptime_linux.c
@@ -5,7 +5,7 @@
 #include <inttypes.h>
 
 const char* ffDetectUptime(FFUptimeResult* result) {
-#ifndef __ANDROID__ // cat: /proc/uptime: Permission denied
+#if !defined(__ANDROID__) && !defined(__ironclad__) // cat: /proc/uptime: Permission denied
 
     // #620
     char buf[64];
@@ -23,12 +23,21 @@ const char* ffDetectUptime(FFUptimeResult* result) {
     }
 
 #endif
-#ifndef __GNU__
+#if !defined(__GNU__) && !defined(__ironclad__)
     struct timespec uptime;
     if (clock_gettime(CLOCK_BOOTTIME, &uptime) != 0) {
         return "clock_gettime(CLOCK_BOOTTIME) failed";
     }
 
+    result->uptime = (uint64_t) uptime.tv_sec * 1000 + (uint64_t) uptime.tv_nsec / 1000000;
+    result->bootTime = ffTimeGetNow() - result->uptime;
+    return NULL;
+#elif defined(__ironclad__)
+    struct timespec uptime;
+    if (clock_gettime(CLOCK_MONOTONIC, &uptime) != 0) {
+        return "clock_gettime(CLOCK_MONOTONIC) failed";
+    }
+
     result->uptime = (uint64_t) uptime.tv_sec * 1000 + (uint64_t) uptime.tv_nsec / 1000000;
     result->bootTime = ffTimeGetNow() - result->uptime;
     return NULL;
diff --git fastfetch-clean/src/detection/zpool/zpool.c fastfetch-workdir/src/detection/zpool/zpool.c
index ce7abb3..9a66e5f 100644
--- fastfetch-clean/src/detection/zpool/zpool.c
+++ fastfetch-workdir/src/detection/zpool/zpool.c
@@ -1,6 +1,6 @@
 #include "zpool.h"
 
-#if FF_HAVE_LIBZFS
+#if FF_HAVE_LIBZFS && !defined(__ironclad__)
 
     #include "common/kmod.h"
 
diff --git fastfetch-clean/src/options/general.h fastfetch-workdir/src/options/general.h
index b58d6c9..8a615d0 100644
--- fastfetch-clean/src/options/general.h
+++ fastfetch-workdir/src/options/general.h
@@ -15,7 +15,7 @@ typedef struct FFOptionsGeneral {
     FFstrbuf playerName;
 
 // Module options that cannot be put in module option structure
-#if defined(__linux__) || defined(__FreeBSD__) || defined(__sun) || defined(__OpenBSD__) || defined(__NetBSD__) || defined(__HAIKU__) || defined(__GNU__)
+#if defined(__linux__) || defined(__FreeBSD__) || defined(__sun) || defined(__OpenBSD__) || defined(__NetBSD__) || defined(__HAIKU__) || defined(__GNU__) || defined(__ironclad__)
     FFDsForceDrmType dsForceDrm;
 #endif
 } FFOptionsGeneral;
