diff --git libogg-clean/CMakeLists.txt libogg-workdir/CMakeLists.txt
index 9776109..f8bc244 100644
--- libogg-clean/CMakeLists.txt
+++ libogg-workdir/CMakeLists.txt
@@ -175,7 +175,8 @@ if(INSTALL_DOCS)
         doc/skeleton.html
     )
     install(FILES ${OGG_DOCS} DESTINATION ${CMAKE_INSTALL_DOCDIR}/html)
-    install(DIRECTORY doc/libogg DESTINATION ${CMAKE_INSTALL_DOCDIR}/html PATTERN "*.am" EXCLUDE)
+    install(DIRECTORY doc/libogg DESTINATION ${CMAKE_INSTALL_DOCDIR}/html
+            PATTERN "*.am" EXCLUDE PATTERN "*.in" EXCLUDE)
 endif()
 
 if(BUILD_TESTING)
