diff --git metalog-clean/configure.ac metalog-workdir/configure.ac
index 07ba1c3..c3c66e9 100644
--- metalog-clean/configure.ac
+++ metalog-workdir/configure.ac
@@ -82,15 +82,15 @@ AM_WITH_DMALLOC
 
 WFLAGS=""
 AC_DEFUN([W_CHECK_CFLAG],[AX_CHECK_COMPILE_FLAG([$1],[WFLAGS="$WFLAGS $1"])])
-W_CHECK_CFLAG([-Wall])
-W_CHECK_CFLAG([-W])
-W_CHECK_CFLAG([-Wcast-qual])
-W_CHECK_CFLAG([-Wcast-align])
-W_CHECK_CFLAG([-Wmissing-noreturn])
-W_CHECK_CFLAG([-Wbad-function-cast])
-W_CHECK_CFLAG([-Wstrict-prototypes])
-W_CHECK_CFLAG([-Wwrite-strings])
-W_CHECK_CFLAG([-Wreturn-type])
+#W_CHECK_CFLAG([-Wall])
+#W_CHECK_CFLAG([-W])
+#W_CHECK_CFLAG([-Wcast-qual])
+#W_CHECK_CFLAG([-Wcast-align])
+#W_CHECK_CFLAG([-Wmissing-noreturn])
+#W_CHECK_CFLAG([-Wbad-function-cast])
+#W_CHECK_CFLAG([-Wstrict-prototypes])
+#W_CHECK_CFLAG([-Wwrite-strings])
+#W_CHECK_CFLAG([-Wreturn-type])
 AC_SUBST(WFLAGS, $WFLAGS)
 
 AC_ARG_WITH(unicode,
