diff --git bash-clean/lib/termcap/termcap.c bash-workdir/lib/termcap/termcap.c
index 2b47062..a18af16 100644
--- bash-clean/lib/termcap/termcap.c
+++ bash-workdir/lib/termcap/termcap.c
@@ -611,7 +611,7 @@ scan_file (char *str, int fd, struct buffer *bufp)
   bufp->ateof = 0;
   *bufp->ptr = '\0';
 
-  lseek (fd, 0L, 0);
+  lseek (fd, 0L, SEEK_SET);
 
   while (!bufp->ateof)
     {
diff --git bash-clean/shell.c bash-workdir/shell.c
index e59029c..653cca1 100644
--- bash-clean/shell.c
+++ bash-workdir/shell.c
@@ -1654,7 +1654,7 @@ open_shell_script (char *script_name)
 #endif
 
   /* Only do this with non-tty file descriptors we can seek on. */
-  if (fd_is_tty == 0 && (lseek (fd, 0L, 1) != -1))
+  if (fd_is_tty == 0 && (lseek (fd, 0L, SEEK_CUR) != -1))
     {
       /* Check to see if the `file' in `bash file' is a binary file
 	 according to the same tests done by execute_simple_command (),
@@ -1691,7 +1691,7 @@ open_shell_script (char *script_name)
 	  exit (EX_BINARY_FILE);
 	}
       /* Now rewind the file back to the beginning. */
-      lseek (fd, 0L, 0);
+      lseek (fd, 0L, SEEK_SET);
     }
 
   /* Open the script.  But try to move the file descriptor to a randomly
