diff --git coreutils-clean/lib/pselect.c coreutils-workdir/lib/pselect.c
index 78f0446..92928be 100644
--- coreutils-clean/lib/pselect.c
+++ coreutils-workdir/lib/pselect.c
@@ -25,6 +25,7 @@
 
 #include <errno.h>
 #include <signal.h>
+#include <stddef.h>
 
 /* Examine the size-NFDS file descriptor sets in RFDS, WFDS, and XFDS
    to see whether some of their descriptors are ready for reading,
diff --git coreutils-clean/src/system.h coreutils-workdir/src/system.h
index b8612ff..52817af 100644
--- coreutils-clean/src/system.h
+++ coreutils-workdir/src/system.h
@@ -31,6 +31,11 @@
 
 #include <unistd.h>
 
+/* Ironclad doesn't have lchown - define to 0 if not already defined.  */
+#ifndef HAVE_LCHOWN
+# define HAVE_LCHOWN 0
+#endif
+
 #include <limits.h>
 
 #include "pathmax.h"
