diff --git mlibc-clean/abis/ironclad/vm-flags.h mlibc-workdir/abis/ironclad/vm-flags.h
index 5725ed0..0c894cd 100644
--- mlibc-clean/abis/ironclad/vm-flags.h
+++ mlibc-workdir/abis/ironclad/vm-flags.h
@@ -14,7 +14,6 @@
 #define MAP_ANON      0x08
 #define MAP_ANONYMOUS 0x08
 #define MAP_NORESERVE 0x10
-#define MAP_WC        (1 << 4)
 
 #define MS_ASYNC 0x01
 #define MS_SYNC 0x02
diff --git mlibc-clean/meson.build mlibc-workdir/meson.build
index a1e526b..9515929 100644
--- mlibc-clean/meson.build
+++ mlibc-workdir/meson.build
@@ -239,6 +239,7 @@ elif host_machine.system() == 'managarm'
 elif host_machine.system() == 'ironclad'
 	rtld_include_dirs += include_directories('sysdeps/ironclad/include')
 	libc_include_dirs += include_directories('sysdeps/ironclad/include')
+	internal_conf.set10('MLIBC_MMAP_ALLOCATE_DSO', true)
 	subdir('sysdeps/ironclad')
 elif host_machine.system() == 'keyronex'
 	rtld_include_dirs += include_directories('sysdeps/keyronex/include')
diff --git mlibc-clean/options/posix/generic/sys-shm.cpp mlibc-workdir/options/posix/generic/sys-shm.cpp
index 8782ae6..bfacac2 100644
--- mlibc-clean/options/posix/generic/sys-shm.cpp
+++ mlibc-workdir/options/posix/generic/sys-shm.cpp
@@ -4,6 +4,7 @@
 #include <bits/ensure.h>
 #include <mlibc/debug.hpp>
 #include <mlibc/posix-sysdeps.hpp>
+#include <sys/syscall.h>
 
 void *shmat(int shmid, const void *shmaddr, int shmflg) {
 	void *ret;
@@ -17,11 +18,7 @@ void *shmat(int shmid, const void *shmaddr, int shmflg) {
 
 int shmctl(int shmid, int cmd, struct shmid_ds *buf) {
 	int ret;
-	auto sysdep = MLIBC_CHECK_OR_ENOSYS(mlibc::sys_shmctl, -1);
-	if(int e = sysdep(&ret, shmid, cmd, buf); e) {
-		errno = e;
-		return -1;
-	}
+	SYSCALL3(SYSCALL_SHMCTL, shmid, cmd, buf);
 	return ret;
 }
 
diff --git mlibc-clean/options/posix/generic/unistd.cpp mlibc-workdir/options/posix/generic/unistd.cpp
index 6e525be..e39c744 100644
--- mlibc-clean/options/posix/generic/unistd.cpp
+++ mlibc-workdir/options/posix/generic/unistd.cpp
@@ -1208,13 +1208,13 @@ pid_t vfork(void) {
 	/* deferring to fork as implementing vfork correctly requires assembly
 	 * to handle not mucking up the stack
 	 */
-	if(!mlibc::sys_fork) {
+	if(!mlibc::sys_vfork) {
 		MLIBC_MISSING_SYSDEP();
 		errno = ENOSYS;
 		return -1;
 	}
 
-	if(int e = mlibc::sys_fork(&child); e) {
+	if(int e = mlibc::sys_vfork(&child); e) {
 		errno = e;
 		return -1;
 	}
diff --git mlibc-clean/options/posix/include/mlibc/posix-sysdeps.hpp mlibc-workdir/options/posix/include/mlibc/posix-sysdeps.hpp
index 51c4e5c..8699c6b 100644
--- mlibc-clean/options/posix/include/mlibc/posix-sysdeps.hpp
+++ mlibc-workdir/options/posix/include/mlibc/posix-sysdeps.hpp
@@ -100,6 +100,7 @@ int sys_close(int fd);
 [[gnu::weak]] void sys_yield();
 [[gnu::weak]] int sys_sleep(time_t *secs, long *nanos);
 [[gnu::weak]] int sys_fork(pid_t *child);
+[[gnu::weak]] int sys_vfork(pid_t *child);
 [[gnu::weak]] int sys_execve(const char *path, char *const argv[], char *const envp[]);
 [[gnu::weak]] int sys_pselect(int num_fds, fd_set *read_set, fd_set *write_set,
 		fd_set *except_set, const struct timespec *timeout, const sigset_t *sigmask, int *num_events);
diff --git mlibc-clean/options/rtld/generic/linker.cpp mlibc-workdir/options/rtld/generic/linker.cpp
index 33cd1b9..fbb98d4 100644
--- mlibc-clean/options/rtld/generic/linker.cpp
+++ mlibc-workdir/options/rtld/generic/linker.cpp
@@ -521,6 +521,7 @@ frg::expected<LinkerError, void> ObjectRepository::_fetchFromFile(SharedObject *
 		return LinkerError::outOfMemory;
 	}
 
+	mlibc::sys_vm_unmap(mappedAddr, highest_address - object->baseAddress);
 	object->baseAddress = reinterpret_cast<uintptr_t>(mappedAddr);
 #else
 	object->baseAddress = libraryBase;
diff --git mlibc-clean/sysdeps/ironclad/generic/generic.cpp mlibc-workdir/sysdeps/ironclad/generic/generic.cpp
index 68e6fee..1a6a67b 100644
--- mlibc-clean/sysdeps/ironclad/generic/generic.cpp
+++ mlibc-workdir/sysdeps/ironclad/generic/generic.cpp
@@ -25,13 +25,9 @@ extern "C" void __mlibc_sigret(void);
 namespace mlibc {
 
 void sys_libc_log(const char *message) {
-#ifdef __MLIBC_DEBUG
 	ssize_t unused;
-	sys_write(2, message, strlen(message), &unused);
-	sys_write(2, "\n", 1, &unused);
-#else
-	(void)message;
-#endif
+	sys_write(1, message, strlen(message), &unused);
+	sys_write(1, "\n", 1, &unused);
 }
 
 void sys_libc_panic() {
@@ -200,6 +196,16 @@ int sys_ftruncate (int fd, size_t size) {
 	return errno;
 }
 
+int sys_fallocate(int fd, off_t offset, size_t size) {
+	if (offset < 0 || size == 0) {
+		return EINVAL;
+	}
+
+	int ret, errno;
+	SYSCALL2(SYSCALL_TRUNCATE, fd, offset + size);
+	return errno;
+}
+
 int sys_flock(int fd, int options) {
 	//  XXX: Shouldnt this use F_SETLKW and F_SETLK only when LOCK_NB ?
 	struct flock lock;
@@ -268,6 +274,12 @@ int sys_vm_map(void *hint, size_t size, int prot, int flags, int fd, off_t offse
 	return errno;
 }
 
+int sys_posix_madvise(void *addr, size_t length, int advice) {
+	int ret, errno;
+	SYSCALL3(SYSCALL_MADVISE, addr, length, advice);
+	return errno;
+}
+
 int sys_getsockopt(int fd, int layer, int number, void *__restrict buffer, socklen_t *__restrict size) {
 	int ret, errno;
 	SYSCALL5(SYSCALL_GETSOCKOPT, fd, layer, number, buffer, size);
@@ -406,7 +418,24 @@ int sys_fork(pid_t *child) {
 	pid_t ret;
 	int errno;
 
-	SYSCALL0(SYSCALL_FORK);
+	SYSCALL2(SYSCALL_FORK, 0, 1);
+
+	if (ret == -1) {
+		return errno;
+	}
+
+	if (child != NULL) {
+		*child = ret;
+	}
+
+	return 0;
+}
+
+int sys_vfork(pid_t *child) {
+	pid_t ret;
+	int errno;
+
+	SYSCALL2(SYSCALL_FORK, 1, 1);
 
 	if (ret == -1) {
 		return errno;
@@ -692,15 +721,14 @@ int sys_mkdirat(int dirfd, const char *path, mode_t mode) {
 int sys_rmdir(const char* path){
 	int ret, errno;
 	size_t path_len = strlen (path);
-	SYSCALL3(SYSCALL_UNLINK, AT_FDCWD, path, path_len);
+	SYSCALL4(SYSCALL_UNLINK, AT_FDCWD, path, path_len, AT_REMOVEDIR);
 	return errno;
 }
 
 int sys_unlinkat(int fd, const char *path, int flags) {
-	(void)flags;
 	int ret, errno;
 	size_t path_len = strlen (path);
-	SYSCALL3(SYSCALL_UNLINK, fd, path, path_len);
+	SYSCALL4(SYSCALL_UNLINK, fd, path, path_len, flags);
 	return errno;
 }
 
@@ -713,8 +741,9 @@ int sys_link(const char* srcpath, const char* destpath) {
 }
 
 int sys_socket(int domain, int type, int protocol, int *fd) {
+	(void)protocol;
 	int ret, errno;
-	SYSCALL3(SYSCALL_SOCKET, domain, type, protocol);
+	SYSCALL2(SYSCALL_SOCKET, domain, type);
 	if (ret != -1) {
 		*fd = ret;
 		return 0;
@@ -723,6 +752,13 @@ int sys_socket(int domain, int type, int protocol, int *fd) {
 	}
 }
 
+int sys_socketpair(int domain, int type, int protocol, int *fds) {
+	(void)protocol;
+	int ret, errno;
+	SYSCALL3(SYSCALL_SOCKETPAIR, domain, type, fds);
+	return errno;
+}
+
 uid_t sys_getuid() {
 	uint64_t ret, errno;
 	SYSCALL0(SYSCALL_GETUID);
@@ -1157,6 +1193,12 @@ int sys_sysconf(int num, long *rret) {
 	int ret, errno;
 
 	switch (num) {
+		case _SC_LINE_MAX:
+			return 2048;
+		case _SC_NGROUPS_MAX:
+			return 0x10000;
+		case _SC_CHILD_MAX:
+			return 30;
 		case _SC_NPROCESSORS_CONF:
 			SYSCALL1(SYSCALL_GETCPUINFO, &cpu);
 			if (ret == 0) {
diff --git mlibc-clean/sysdeps/ironclad/include/sys/syscall.h mlibc-workdir/sysdeps/ironclad/include/sys/syscall.h
index 9c576dc..ff38017 100644
--- mlibc-clean/sysdeps/ironclad/include/sys/syscall.h
+++ mlibc-workdir/sysdeps/ironclad/include/sys/syscall.h
@@ -36,7 +36,6 @@
 #define SYSCALL_GETUID               24
 #define SYSCALL_RENAME               25
 #define SYSCALL_LISTPROCS            26
-#define SYSCALL_SPAWN                27
 #define SYSCALL_GETTID               28
 #define SYSCALL_MANAGE_TCLUSTER      29
 #define SYSCALL_FCNTL                30
@@ -124,6 +123,8 @@
 #define SYSCALL_MEMINFO             112
 #define SYSCALL_LISTPCI             113
 #define SYSCALL_GETCPUINFO          114
+#define SYSCALL_SOCKETPAIR          115
+#define SYSCALL_MADVISE             116
 
 #if defined(__x86_64__)
 #define SYSCALL0(NUM) ({ \
