diff --git mlibc-clean/abis/ironclad/abi.h mlibc-workdir/abis/ironclad/abi.h
index d522358..cb4854d 100644
--- mlibc-clean/abis/ironclad/abi.h
+++ mlibc-workdir/abis/ironclad/abi.h
@@ -21,6 +21,7 @@
 #define __MLIBC_O_DSYNC 0x00800
 #define __MLIBC_O_RSYNC 0x01000
 #define __MLIBC_O_SYNC 0x02000
+#define __MLIBC_O_ASYNC 0x6969
 #define __MLIBC_O_CLOEXEC 0x04000
 #define __MLIBC_O_PATH 0x08000
 #define __MLIBC_O_LARGEFILE 0x10000
diff --git mlibc-clean/sysdeps/ironclad/generic/generic.cpp mlibc-workdir/sysdeps/ironclad/generic/generic.cpp
index 3b63ed3..2e67e45 100644
--- mlibc-clean/sysdeps/ironclad/generic/generic.cpp
+++ mlibc-workdir/sysdeps/ironclad/generic/generic.cpp
@@ -39,7 +39,7 @@ void sys_exit(int status) {
 
 int sys_tcb_set(void *pointer) {
 	int ret, errno;
-	SYSCALL1(SYSCALL_SET_TCB, pointer);
+	SYSCALL2(SYSCALL_ARCH_PRCTL, 1, pointer);
 	return errno;
 }
 
@@ -181,9 +181,12 @@ int sys_pselect(int num_fds, fd_set *read_set, fd_set *write_set, fd_set *except
 }
 
 int sys_isatty(int fd) {
-	(void)fd;
-	mlibc::infoLogger() << "mlibc: " << __func__ << " is a stub!\n" << frg::endlog;
-	return 0;
+	struct termios t;
+	if (sys_tcgetattr(fd, &t) == 0) {
+		return 0;
+	} else {
+		return ENOTTY;
+	}
 }
 
 uid_t sys_getuid() {
diff --git mlibc-clean/sysdeps/ironclad/include/sys/syscall.h mlibc-workdir/sysdeps/ironclad/include/sys/syscall.h
index 28d233e..6f85c50 100644
--- mlibc-clean/sysdeps/ironclad/include/sys/syscall.h
+++ mlibc-workdir/sysdeps/ironclad/include/sys/syscall.h
@@ -49,7 +49,7 @@
 })
 
 #define SYSCALL_EXIT               0
-#define SYSCALL_SET_TCB            1
+#define SYSCALL_ARCH_PRCTL         1
 #define SYSCALL_OPEN               2
 #define SYSCALL_CLOSE              3
 #define SYSCALL_READ               4
