diff --git mlibc-clean/abis/ironclad/abi.h mlibc-workdir/abis/ironclad/abi.h
index d522358..f1e1c81 100644
--- mlibc-clean/abis/ironclad/abi.h
+++ mlibc-workdir/abis/ironclad/abi.h
@@ -8,6 +8,7 @@
 #define __MLIBC_O_RDWR   3
 #define __MLIBC_O_SEARCH 4
 #define __MLIBC_O_EXEC   7
+#define __MLIBC_O_CLOEXEC 16
 
 // This flags are stubs.
 #define __MLIBC_O_APPEND 0x00008
@@ -21,7 +22,8 @@
 #define __MLIBC_O_DSYNC 0x00800
 #define __MLIBC_O_RSYNC 0x01000
 #define __MLIBC_O_SYNC 0x02000
-#define __MLIBC_O_CLOEXEC 0x04000
+#define __MLIBC_O_ASYNC 0x6969
+
 #define __MLIBC_O_PATH 0x08000
 #define __MLIBC_O_LARGEFILE 0x10000
 #define __MLIBC_O_NOATIME 0x20000
diff --git mlibc-clean/options/posix/generic/unistd-stubs.cpp mlibc-workdir/options/posix/generic/unistd-stubs.cpp
index add6743..03dc437 100644
--- mlibc-clean/options/posix/generic/unistd-stubs.cpp
+++ mlibc-workdir/options/posix/generic/unistd-stubs.cpp
@@ -795,8 +795,8 @@ unsigned long sysconf(int number) {
 			return mlibc::page_size;
 		case _SC_OPEN_MAX:
 			// TODO: actually return a proper value for _SC_OPEN_MAX
-			mlibc::infoLogger() << "\e[31mmlibc: sysconf(_SC_OPEN_MAX) returns arbitrary value 256\e[39m" << frg::endlog;
-			return 256;
+			mlibc::infoLogger() << "\e[31mmlibc: sysconf(_SC_OPEN_MAX) returns arbitrary value 100\e[39m" << frg::endlog;
+			return 100;
 		case _SC_PHYS_PAGES:
 			// TODO: actually return a proper value for _SC_PHYS_PAGES
 			mlibc::infoLogger() << "\e[31mmlibc: sysconf(_SC_PHYS_PAGES) returns arbitrary value 1024\e[39m" << frg::endlog;
diff --git mlibc-clean/options/posix/include/sched.h mlibc-workdir/options/posix/include/sched.h
index c69ff2a..41dbe73 100644
--- mlibc-clean/options/posix/include/sched.h
+++ mlibc-workdir/options/posix/include/sched.h
@@ -9,29 +9,10 @@
 
 // MISSING: POSIX [PS], [SS] and [TSP] options
 
-#define CLONE_VM 0x00000100
-#define CLONE_FS 0x00000200
-#define CLONE_FILES	0x00000400
-#define CLONE_SIGHAND 0x00000800
-#define CLONE_PTRACE 0x00002000
-#define CLONE_VFORK 0x00004000
-#define CLONE_PARENT 0x00008000
-#define CLONE_THREAD 0x00010000
-#define CLONE_NEWNS 0x00020000
-#define CLONE_SYSVSEM 0x00040000
-#define CLONE_SETTLS 0x00080000
-#define CLONE_PARENT_SETTID 0x00100000
-#define CLONE_CHILD_CLEARTID 0x00200000
-#define CLONE_DETACHED 0x00400000
-#define CLONE_UNTRACED 0x00800000
-#define CLONE_CHILD_SETTID 0x01000000
-#define CLONE_NEWCGROUP 0x02000000
-#define CLONE_NEWUTS 0x04000000
-#define CLONE_NEWIPC 0x08000000
-#define CLONE_NEWUSER 0x10000000
-#define CLONE_NEWPID 0x20000000
-#define CLONE_NEWNET 0x40000000
-#define CLONE_IO 0x80000000
+#define CLONE_VM                0b00001
+#define CLONE_VFORK             0b00010
+#define CLONE_CLEAR_FILES_NOSTD 0b00100
+#define CLONE_CLEAR_FILES       0b01000
 
 #define CPU_SETSIZE 128
 #define CPU_ISSET __mlibc_cpu_isset
diff --git mlibc-clean/sysdeps/ironclad/generic/generic.cpp mlibc-workdir/sysdeps/ironclad/generic/generic.cpp
index 3b63ed3..02e3e3d 100644
--- mlibc-clean/sysdeps/ironclad/generic/generic.cpp
+++ mlibc-workdir/sysdeps/ironclad/generic/generic.cpp
@@ -10,6 +10,8 @@
 #include <string.h>
 #include <stdio.h>
 #include <sys/syscall.h>
+#include <sys/ironclad.h>
+#include <sched.h>
 
 namespace mlibc {
 
@@ -39,7 +41,7 @@ void sys_exit(int status) {
 
 int sys_tcb_set(void *pointer) {
 	int ret, errno;
-	SYSCALL1(SYSCALL_SET_TCB, pointer);
+	SYSCALL2(SYSCALL_ARCH_PRCTL, 1, pointer);
 	return errno;
 }
 
@@ -136,12 +138,10 @@ int sys_sigaction(int signum, const struct sigaction *act, struct sigaction *old
 }
 
 int sys_fcntl(int fd, int request, va_list args, int *result) {
-	(void)fd;
-	(void)request;
-	(void)args;
-	(void)result;
-	mlibc::infoLogger() << "fcntl() is a stub!" << frg::endlog;
-	return 0;
+	int ret, errno;
+	SYSCALL3(SYSCALL_FCNTL, fd, request, va_arg(args, uint64_t));
+	*result = ret;
+	return errno;
 }
 
 int sys_sigprocmask(int how, const sigset_t *__restrict set, sigset_t *__restrict retrieve) {
@@ -181,9 +181,12 @@ int sys_pselect(int num_fds, fd_set *read_set, fd_set *write_set, fd_set *except
 }
 
 int sys_isatty(int fd) {
-	(void)fd;
-	mlibc::infoLogger() << "mlibc: " << __func__ << " is a stub!\n" << frg::endlog;
-	return 0;
+	struct termios t;
+	if (sys_tcgetattr(fd, &t) == 0) {
+		return 0;
+	} else {
+		return ENOTTY;
+	}
 }
 
 uid_t sys_getuid() {
@@ -243,7 +246,9 @@ int sys_fork(pid_t *child) {
         return errno;
     }
 
-    *child = ret;
+    if (child != NULL) {
+        *child = ret;
+    }
     return 0;
 }
 
@@ -490,3 +495,27 @@ int sys_stat(fsfd_target fsfdt, int fd, const char *path, int flags, struct stat
 
 #endif
 } // namespace mlibc
+
+extern "C" {
+
+// Syscalls for ironclad.h
+int get_thread_sched(void) {
+	int ret, errno;
+	SYSCALL0(SYSCALL_GET_THREAD_SCHED);
+	return ret;
+}
+
+int set_thread_sched(int flags) {
+	int ret, errno;
+	SYSCALL1(SYSCALL_SET_THREAD_SCHED, flags);
+	return ret;
+}
+
+pid_t program_spawn(const char *path, char *const argv[], char *const envp[]) {
+	pid_t ret;
+	int errno;
+	SYSCALL3(SYSCALL_SPAWN, path, argv, envp);
+	return ret;
+}
+
+}
diff --git mlibc-clean/sysdeps/ironclad/include/sys/ironclad.h mlibc-workdir/sysdeps/ironclad/include/sys/ironclad.h
index 89a3851..e046507 100644
--- mlibc-clean/sysdeps/ironclad/include/sys/ironclad.h
+++ mlibc-workdir/sysdeps/ironclad/include/sys/ironclad.h
@@ -3,6 +3,7 @@
 
 #include <stdint.h>
 #include <stdbool.h>
+#include <sys/types.h>
 
 #ifdef __cplusplus
 extern "C" {
@@ -34,6 +35,18 @@ struct ironclad_mouse_data {
 	bool is_left;
 	bool is_right;
 };
+
+// Syscalls declared on the documentation, but without libc equivalents.
+// Implemented in generic.cpp
+#define THREAD_RT     0b0001
+#define THREAD_MONO   0b0010
+#define THREAD_MLOCK  0b0100
+#define THREAD_BANNED 0b1000
+
+int get_thread_sched(void);
+int set_thread_sched(int flags);
+pid_t program_spawn(const char *path, char *const argv[], char *const envp[]);
+
 #ifdef __cplusplus
 }
 #endif
diff --git mlibc-clean/sysdeps/ironclad/include/sys/syscall.h mlibc-workdir/sysdeps/ironclad/include/sys/syscall.h
index 28d233e..1b6c659 100644
--- mlibc-clean/sysdeps/ironclad/include/sys/syscall.h
+++ mlibc-workdir/sysdeps/ironclad/include/sys/syscall.h
@@ -49,7 +49,7 @@
 })
 
 #define SYSCALL_EXIT               0
-#define SYSCALL_SET_TCB            1
+#define SYSCALL_ARCH_PRCTL         1
 #define SYSCALL_OPEN               2
 #define SYSCALL_CLOSE              3
 #define SYSCALL_READ               4
@@ -76,5 +76,9 @@
 #define SYSCALL_DUP2              25
 #define SYSCALL_DUP3              26
 #define SYSCALL_ACCESS            27
+#define SYSCALL_GET_THREAD_SCHED  28
+#define SYSCALL_SET_THREAD_SCHED  29
+#define SYSCALL_FCNTL             30
+#define SYSCALL_SPAWN             31
 
 #endif // _SYSCALL_H
