diff --git zstd-clean/build/cmake/contrib/CMakeLists.txt zstd-workdir/build/cmake/contrib/CMakeLists.txt
index 8df2a17..5fbf3ce 100644
--- zstd-clean/build/cmake/contrib/CMakeLists.txt
+++ zstd-workdir/build/cmake/contrib/CMakeLists.txt
@@ -10,4 +10,5 @@
 project(contrib)
 
 add_subdirectory(pzstd)
-add_subdirectory(gen_html)
+# gen_html runs the binary it just built in order to generate
+# zstd_manual.html, which cannot work when cross-compiling.
diff --git zstd-clean/lib/common/portability_macros.h zstd-workdir/lib/common/portability_macros.h
index 8607341..0a33622 100644
--- zstd-clean/lib/common/portability_macros.h
+++ zstd-workdir/lib/common/portability_macros.h
@@ -114,7 +114,8 @@
  * 100% of code to be instrumented to work.
  */
 #if defined(__GNUC__)
-#  if defined(__linux__) || defined(__linux) || defined(__APPLE__) || defined(_WIN32)
+#  if defined(__linux__) || defined(__linux) || defined(__APPLE__) || defined(_WIN32) \
+   || defined(__ironclad__)
 #    if ZSTD_MEMORY_SANITIZER
 #      define ZSTD_ASM_SUPPORTED 0
 #    elif ZSTD_DATAFLOW_SANITIZER
diff --git zstd-clean/programs/fileio.c zstd-workdir/programs/fileio.c
index 0ecca40..7012b68 100644
--- zstd-clean/programs/fileio.c
+++ zstd-workdir/programs/fileio.c
@@ -173,6 +173,7 @@ static void clearHandler(void)
 #if !defined(BACKTRACE_ENABLE)
 /* automatic detector : backtrace enabled by default on linux+glibc and osx */
 #  if (defined(__linux__) && (defined(__GLIBC__) && !defined(__UCLIBC__))) \
+     || (defined(__ironclad__) && defined(__GLIBC__)) \
      || (defined(__APPLE__) && defined(__MACH__))
 #    define BACKTRACE_ENABLE 1
 #  else
diff --git zstd-clean/programs/util.c zstd-workdir/programs/util.c
index 065a358..e5f0210 100644
--- zstd-clean/programs/util.c
+++ zstd-workdir/programs/util.c
@@ -1598,7 +1598,8 @@ int UTIL_countCores(int logical)
     return numCores;
 }
 
-#elif defined(__NetBSD__) || defined(__OpenBSD__) || defined(__DragonFly__) || defined(__CYGWIN__)
+#elif defined(__NetBSD__) || defined(__OpenBSD__) || defined(__DragonFly__) || defined(__CYGWIN__) \
+ || defined(__ironclad__)
 
 /* Use POSIX sysconf
  * see: man 3 sysconf */
