diff --git qemu-clean/block/file-posix.c qemu-workdir/block/file-posix.c
index 3c985da..2ef19f4 100644
--- qemu-clean/block/file-posix.c
+++ qemu-workdir/block/file-posix.c
@@ -118,6 +118,8 @@
 #endif
 #endif
 
+#include <sys/ioctl.h>
+
 /* Approximate O_DIRECT with O_DSYNC if O_DIRECT isn't available */
 #ifndef O_DIRECT
 #define O_DIRECT O_DSYNC
diff --git qemu-clean/chardev/char-serial.c qemu-workdir/chardev/char-serial.c
index 0c73baf..9fc293b 100644
--- qemu-clean/chardev/char-serial.c
+++ qemu-workdir/chardev/char-serial.c
@@ -34,6 +34,9 @@
 #else
 #include <sys/ioctl.h>
 #include <termios.h>
+#ifndef CRTSCTS
+#define CRTSCTS 0
+#endif
 #include "chardev/char-fd.h"
 #endif
 
diff --git qemu-clean/configure qemu-workdir/configure
index 6ddfe4b..c8afaaf 100755
--- qemu-clean/configure
+++ qemu-workdir/configure
@@ -348,7 +348,9 @@ int main(void) { return 0; }
 EOF
 }
 
-if check_define __linux__ ; then
+if check_define __ironclad__ ; then
+  host_os=ironclad
+elif check_define __linux__ ; then
   host_os=linux
 elif check_define _WIN32 ; then
   host_os=windows
diff --git qemu-clean/include/gdbstub/syscalls.h qemu-workdir/include/gdbstub/syscalls.h
index a095591..629baf8 100644
--- qemu-clean/include/gdbstub/syscalls.h
+++ qemu-workdir/include/gdbstub/syscalls.h
@@ -43,9 +43,10 @@
 #define GDB_EUNKNOWN       9999
 
 /* For gdb file i/o remote protocol lseek whence. */
-#define GDB_SEEK_SET  0
-#define GDB_SEEK_CUR  1
-#define GDB_SEEK_END  2
+#include <unistd.h>
+#define GDB_SEEK_SET  SEEK_SET
+#define GDB_SEEK_CUR  SEEK_CUR
+#define GDB_SEEK_END  SEEK_END
 
 /* For gdb file i/o stat/fstat. */
 typedef uint32_t gdb_mode_t;
diff --git qemu-clean/meson.build qemu-workdir/meson.build
index 49a5baf..647970b 100644
--- qemu-clean/meson.build
+++ qemu-workdir/meson.build
@@ -910,7 +910,7 @@ if get_option('nitro').allowed() and host_os == 'linux'
 endif
 
 nvmm = not_found
-if host_os == 'netbsd'
+if host_os == 'netbsd' or host_os == 'ironclad'
   nvmm = cc.find_library('nvmm', required: get_option('nvmm'))
   if nvmm.found()
     accelerators += 'CONFIG_NVMM'
@@ -2489,6 +2489,7 @@ config_host_data.set('CONFIG_PASST', enable_passt)
 config_host_data.set('CONFIG_SLIRP', slirp.found())
 config_host_data.set('CONFIG_SNAPPY', snappy.found())
 config_host_data.set('CONFIG_SOLARIS', host_os == 'sunos')
+config_host_data.set('CONFIG_IRONCLAD', host_os == 'ironclad')
 if have_tcg
   config_host_data.set('CONFIG_TCG', 1)
   config_host_data.set('CONFIG_TCG_INTERPRETER', tcg_arch == 'tci')
diff --git qemu-clean/subprojects/dtc/libfdt/meson.build qemu-workdir/subprojects/dtc/libfdt/meson.build
index 0307ffb..1cde105 100644
--- qemu-clean/subprojects/dtc/libfdt/meson.build
+++ qemu-workdir/subprojects/dtc/libfdt/meson.build
@@ -21,7 +21,7 @@ libfdt = library(
   version: '1.6.0',
   link_args: ['-Wl,--no-undefined', version_script],
   link_depends: 'version.lds',
-  install: true,
+  install: false,
 )
 
 libfdt_inc = include_directories('.')
@@ -30,21 +30,3 @@ libfdt_dep = declare_dependency(
   include_directories: libfdt_inc,
   link_with: libfdt,
 )
-
-install_headers(
-  files(
-    'fdt.h',
-    'libfdt.h',
-    'libfdt_env.h',
-  )
-)
-
-pkgconfig = import('pkgconfig')
-
-pkgconfig.generate(
-  libraries: libfdt,
-  version: meson.project_version(),
-  filebase: 'libfdt',
-  name: 'libfdt',
-  description: 'Flat Device Tree manipulation',
-)
diff --git qemu-clean/util/cpuinfo-riscv.c qemu-workdir/util/cpuinfo-riscv.c
index 0291b72..df89863 100644
--- qemu-clean/util/cpuinfo-riscv.c
+++ qemu-workdir/util/cpuinfo-riscv.c
@@ -24,6 +24,8 @@ static void sigill_handler(int signo, siginfo_t *si, void *data)
 
 #ifdef __linux__
     uc->uc_mcontext.__gregs[REG_PC] += 4;
+#elif defined(__ironclad__)
+    // XXX: Actually implement this
 #elif defined(__OpenBSD__)
     uc->sc_sepc += 4;
 #else
