diff --git gcc-host-workdir/gcc/config/i386/ironclad-gnu.h gcc-host-workdir/gcc/config/i386/ironclad-gnu.h
new file mode 100644
index 0000000..aa69dd2
--- /dev/null
+++ gcc-host-workdir/gcc/config/i386/ironclad-gnu.h
@@ -0,0 +1,8 @@
+
+#define GNU_USER_LINK_EMULATION32 "elf_i386"
+#define GNU_USER_LINK_EMULATION64 "elf_x86_64"
+#define GNU_USER_LINK_EMULATIONX32 "elf32_x86_64"
+
+#define GLIBC_DYNAMIC_LINKER32 "/lib/ld-ironclad-i386.so.2"
+#define GLIBC_DYNAMIC_LINKER64 "/lib/ld-ironclad-x86-64.so.2"
+#define GLIBC_DYNAMIC_LINKERX32 "/lib/ld-ironclad-x32.so.2"
diff --git gcc-host-workdir/gcc/config/i386/t-ironclad64 gcc-host-workdir/gcc/config/i386/t-ironclad64
new file mode 100644
index 0000000..b3385f4
--- /dev/null
+++ gcc-host-workdir/gcc/config/i386/t-ironclad64
@@ -0,0 +1,26 @@
+# Copyright (C) 2002-2025 Free Software Foundation, Inc.
+#
+# This file is part of GCC.
+#
+# GCC is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3, or (at your option)
+# any later version.
+#
+# GCC is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with GCC; see the file COPYING3.  If not see
+# <http://www.gnu.org/licenses/>.
+
+# Ironclad uses /lib for all libraries (no /lib64).
+
+comma=,
+MULTILIB_OPTIONS    = $(subst $(comma),/,$(TM_MULTILIB_CONFIG))
+MULTILIB_DIRNAMES   = $(patsubst m%, %, $(subst /, ,$(MULTILIB_OPTIONS)))
+MULTILIB_OSDIRNAMES = m64=../lib$(call if_multiarch,:x86_64-ironclad-gnu)
+MULTILIB_OSDIRNAMES+= m32=$(if $(wildcard $(shell echo $(SYSTEM_HEADER_DIR))/../../usr/lib32),../lib32,../lib)$(call if_multiarch,:i386-ironclad-gnu)
+MULTILIB_OSDIRNAMES+= mx32=../libx32$(call if_multiarch,:x86_64-ironclad-gnux32)
diff --git gcc-host-workdir/gcc/config/ironclad-gnu.h gcc-host-workdir/gcc/config/ironclad-gnu.h
new file mode 100644
index 0000000..eada791
--- /dev/null
+++ gcc-host-workdir/gcc/config/ironclad-gnu.h
@@ -0,0 +1,15 @@
+
+#define GNU_USER_TARGET_OS_CPP_BUILTINS()		\
+  do						\
+    {						\
+	builtin_define ("__ironclad__");	\
+	builtin_define_std ("unix");		\
+	builtin_assert ("system=unix");		\
+	builtin_assert ("system=posix");	\
+    }						\
+  while (0)
+
+#define GNU_USER_DYNAMIC_LINKER        GLIBC_DYNAMIC_LINKER
+#define GNU_USER_DYNAMIC_LINKER32      GLIBC_DYNAMIC_LINKER32
+#define GNU_USER_DYNAMIC_LINKER64      GLIBC_DYNAMIC_LINKER64
+#define GNU_USER_DYNAMIC_LINKERX32     GLIBC_DYNAMIC_LINKERX32
diff --git gcc-host-workdir/gcc/config/riscv/ironclad.h gcc-host-workdir/gcc/config/riscv/ironclad.h
new file mode 100644
index 0000000..cb889e8
--- /dev/null
+++ gcc-host-workdir/gcc/config/riscv/ironclad.h
@@ -0,0 +1,45 @@
+#undef TARGET_IRONCLAD
+#define TARGET_IRONCLAD 1
+
+#undef LIB_SPEC
+#define LIB_SPEC "-lc"
+
+#define LD_EMUL_SUFFIX \
+  "%{mabi=lp64d:}" \
+  "%{mabi=lp64f:_lp64f}" \
+  "%{mabi=lp64:_lp64}" \
+  "%{mabi=ilp32d:}" \
+  "%{mabi=ilp32f:_ilp32f}" \
+  "%{mabi=ilp32:_ilp32}"
+
+#define GNU_USER_LINK_EMULATION "elf64lriscv"
+#define GNU_USER_DYNAMIC_LINKER "/usr/lib/ld-ironclad-riscv64.so.1"
+
+#define LINK_SPEC "\
+-melf" XLEN_SPEC DEFAULT_ENDIAN_SPEC "riscv" LD_EMUL_SUFFIX " \
+%{mno-relax:--no-relax} \
+%{mbig-endian:-EB} \
+%{mlittle-endian:-EL} \
+%{shared} \
+  %{!shared: \
+    %{!static: \
+      %{!static-pie: \
+	%{rdynamic:-export-dynamic} \
+	-dynamic-linker " GNU_USER_DYNAMIC_LINKER "}} \
+    %{static:-static} %{static-pie:-static -pie --no-dynamic-linker -z text}}"
+
+#undef STARTFILE_SPEC
+#define STARTFILE_SPEC "%{!shared: %{static-pie|pie|!no-pie:Scrt1.o%s;:crt1.o%s}} crti.o%s %{shared|static-pie|pie|!no-pie:crtbeginS.o%s;:crtbegin.o%s}"
+
+#undef ENDFILE_SPEC
+#define ENDFILE_SPEC "%{shared|static-pie|pie|!no-pie:crtendS.o%s;:crtend.o%s} crtn.o%s"
+
+#undef TARGET_OS_CPP_BUILTINS
+#define TARGET_OS_CPP_BUILTINS()         \
+  do {                                   \
+    builtin_define ("__ironclad__");     \
+    builtin_define ("__unix__");         \
+    builtin_assert ("system=ironclad");  \
+    builtin_assert ("system=unix");      \
+    builtin_assert ("system=posix");     \
+  } while (0);
diff --git gcc-host-clean/gcc/config.build gcc-host-workdir/gcc/config.build
index 7c80a0b..8fac7e4 100644
--- gcc-host-clean/gcc/config.build
+++ gcc-host-workdir/gcc/config.build
@@ -45,7 +45,7 @@
 build_xm_file=
 build_xm_defines=
 build_exeext=
-build_install_headers_dir=install-headers-tar
+build_install_headers_dir=install-headers-cp
 build_file_translate=
 
 # System-specific settings.
diff --git gcc-host-clean/gcc/config.gcc gcc-host-workdir/gcc/config.gcc
index 087aaa7..2461e7c 100644
--- gcc-host-clean/gcc/config.gcc
+++ gcc-host-workdir/gcc/config.gcc
@@ -885,7 +885,7 @@ case ${target} in
   rust_target_objs="${rust_target_objs} fuchsia-rust.o"
   target_has_targetrustm=yes
   ;;
-*-*-linux* | frv-*-*linux* | *-*-kfreebsd*-gnu | *-*-gnu* | *-*-kopensolaris*-gnu | *-*-uclinuxfdpiceabi)
+*-*-linux* | frv-*-*linux* | *-*-ironclad*-gnu | *-*-kfreebsd*-gnu | *-*-gnu* | *-*-kopensolaris*-gnu | *-*-uclinuxfdpiceabi)
   extra_options="$extra_options gnu-user.opt"
   gas=yes
   gnu_ld=yes
@@ -894,7 +894,7 @@ case ${target} in
   esac
   tmake_file="t-slibgcc"
   case $target in
-    *-*-linux* | frv-*-*linux* | *-*-kfreebsd*-gnu | *-*-kopensolaris*-gnu)
+    *-*-linux* | frv-*-*linux* | *-*-ironclad*-gnu | *-*-kfreebsd*-gnu | *-*-kopensolaris*-gnu)
       :;;
     *-*-gnu*)
       native_system_header_dir=/include
@@ -957,6 +957,8 @@ case ${target} in
       rust_target_objs="${rust_target_objs} linux-rust.o"
       target_has_targetrustm=yes
       ;;
+    *-*-ironclad*-gnu)
+      ;;
     *-*-kfreebsd*-gnu)
       d_target_objs="${d_target_objs} kfreebsd-d.o"
       target_has_targetdm=yes
@@ -2008,7 +2010,7 @@ x86_64-*-openbsd*)
 	gas=yes
 	gnu_ld=yes
 	;;
-i[34567]86-*-linux* | i[34567]86-*-kfreebsd*-gnu | i[34567]86-*-gnu* | i[34567]86-*-kopensolaris*-gnu)
+i[34567]86-*-linux* | i[34567]86-*-ironclad*-gnu | i[34567]86-*-kfreebsd*-gnu | i[34567]86-*-gnu* | i[34567]86-*-kopensolaris*-gnu)
 			# Intel 80386's running GNU/*
 			# with ELF format using glibc 2
 	tm_file="${tm_file} i386/unix.h i386/att.h elfos.h gnu-user.h glibc-stdint.h"
@@ -2056,6 +2058,9 @@ i[34567]86-*-linux* | i[34567]86-*-kfreebsd*-gnu | i[34567]86-*-gnu* | i[34567]8
 			tm_file="${tm_file} i386/gnu-user-common.h i386/gnu-user.h i386/linux-common.h i386/linux.h"
 		fi
 		;;
+	i[34567]86-*-ironclad*-gnu)
+		tm_file="${tm_file} i386/gnu-user-common.h i386/gnu-user.h ironclad-gnu.h i386/ironclad-gnu.h"
+		;;
 	i[34567]86-*-kfreebsd*-gnu)
 		tm_file="${tm_file} i386/gnu-user-common.h i386/gnu-user.h kfreebsd-gnu.h i386/kfreebsd-gnu.h"
 		;;
@@ -2067,7 +2072,7 @@ i[34567]86-*-linux* | i[34567]86-*-kfreebsd*-gnu | i[34567]86-*-gnu* | i[34567]8
 		;;
 	esac
 	;;
-x86_64-*-linux* | x86_64-*-kfreebsd*-gnu | x86_64-*-gnu*)
+x86_64-*-linux* | x86_64-*-ironclad*-gnu | x86_64-*-kfreebsd*-gnu | x86_64-*-gnu*)
 	tm_file="${tm_file} i386/unix.h i386/att.h elfos.h gnu-user.h glibc-stdint.h \
 		 i386/x86-64.h i386/gnu-user-common.h i386/gnu-user64.h"
 	case ${target} in
@@ -2075,6 +2080,9 @@ x86_64-*-linux* | x86_64-*-kfreebsd*-gnu | x86_64-*-gnu*)
 		tm_file="${tm_file} linux.h linux-android.h i386/linux-common.h i386/linux64.h"
 		extra_options="${extra_options} linux-android.opt"
 	  	;;
+	x86_64-*-ironclad*-gnu)
+		tm_file="${tm_file} ironclad-gnu.h i386/ironclad-gnu.h"
+		;;
 	x86_64-*-kfreebsd*-gnu)
 		tm_file="${tm_file} kfreebsd-gnu.h i386/kfreebsd-gnu64.h"
 		;;
@@ -3670,7 +3678,7 @@ esac
 case ${target} in
 *-*-linux*android*|*-*-linux*uclibc*|*-*-linux*musl*)
         ;;
-*-*-kfreebsd*-gnu | *-*-kopensolaris*-gnu)
+*-*-ironclad*-gnu | *-*-kfreebsd*-gnu | *-*-kopensolaris*-gnu)
         ;;
 *-*-linux* | *-*-gnu*)
 	case ${target} in
@@ -6106,6 +6114,9 @@ case ${target} in
 	visium-*-*)
 		target_cpu_default2="TARGET_CPU_$with_cpu"
 		;;
+	x86_64-*-ironclad*-gnu*)
+		tmake_file="$tmake_file i386/t-ironclad64"
+		;;
 	x86_64-*-gnu*)
 		tmake_file="$tmake_file i386/t-gnu64"
 		;;
diff --git gcc-host-clean/gcc/cp/module.cc gcc-host-workdir/gcc/cp/module.cc
index 1deadcc..9b72332 100644
--- gcc-host-clean/gcc/cp/module.cc
+++ gcc-host-workdir/gcc/cp/module.cc
@@ -1697,8 +1697,10 @@ elf_in::defrost (const char *name)
 	      set_error (errno);
 	  else
 	    {
+#ifndef __ironclad__
 	      if (madvise (mapping, hdr.pos, MADV_RANDOM))
 		goto fail;
+#endif
 
 	      /* These buffers are never NULL in this case.  */
 	      strtab.buffer = mapping + strtab.pos;
@@ -1808,8 +1810,10 @@ elf_in::begin (location_t loc)
     }
   /* We'll be hopping over this randomly.  Some systems declare the
      first parm as char *, and other declare it as void *.  */
+#ifndef __ironclad__
   if (madvise (reinterpret_cast <char *> (mapping), size, MADV_RANDOM))
     goto fail;
+#endif
 
   hdr.buffer = (char *)mapping;
 #else
diff --git gcc-host-clean/libcpp/Makefile.in gcc-host-workdir/libcpp/Makefile.in
index bb07e46..821fe80 100644
--- gcc-host-clean/libcpp/Makefile.in
+++ gcc-host-workdir/libcpp/Makefile.in
@@ -32,10 +32,10 @@ AUTOCONF = @AUTOCONF@
 AUTOHEADER = @AUTOHEADER@
 CATALOGS = $(patsubst %,po/%,@CATALOGS@)
 CC = @CC@
-CFLAGS = @CFLAGS@
+override CFLAGS := @CFLAGS@
 WARN_CFLAGS = @warn@ @c_warn@ @WARN_PEDANTIC@ @WERROR@
 CXX = @CXX@
-CXXFLAGS = @CXXFLAGS@
+override CXXFLAGS := @CXXFLAGS@
 WARN_CXXFLAGS = @warn@ @WARN_PEDANTIC@ @WERROR@
 CPP = @CPP@
 CPPFLAGS = @CPPFLAGS@
@@ -45,7 +45,7 @@ INCINTL = @INCINTL@
 INSTALL_DATA = @INSTALL_DATA@
 INSTALL_PROGRAM = @INSTALL_PROGRAM@
 INSTALL_SCRIPT = @INSTALL_SCRIPT@
-LDFLAGS = @LDFLAGS@
+override LDFLAGS := @LDFLAGS@
 LIBICONV = @LIBICONV@
 LIBINTL = @LIBINTL@
 PACKAGE = @PACKAGE@
diff --git gcc-host-clean/libgcc/config.host gcc-host-workdir/libgcc/config.host
index 6a88ee5..543c21e 100644
--- gcc-host-clean/libgcc/config.host
+++ gcc-host-workdir/libgcc/config.host
@@ -310,7 +310,7 @@ case ${host} in
   tmake_file="$tmake_file t-crtstuff-pic t-libgcc-pic t-eh-dw2-dip t-slibgcc t-slibgcc-fuchsia"
   extra_parts="crtbegin.o crtend.o"
   ;;
-*-*-linux* | frv-*-*linux* | *-*-kfreebsd*-gnu | *-*-gnu* | *-*-kopensolaris*-gnu | *-*-uclinuxfdpiceabi)
+*-*-linux* | frv-*-*linux* | *-*-ironclad*-gnu | *-*-kfreebsd*-gnu | *-*-gnu* | *-*-kopensolaris*-gnu | *-*-uclinuxfdpiceabi)
   tmake_file="$tmake_file t-crtstuff-pic t-libgcc-pic t-eh-dw2-dip t-slibgcc t-slibgcc-gld t-slibgcc-elf-ver t-linux"
   extra_parts="crtbegin.o crtbeginS.o crtbeginT.o crtend.o crtendS.o"
   if test x$enable_vtable_verify = xyes; then
@@ -800,7 +800,7 @@ i[34567]86-*-linux*)
 	md_unwind_header=i386/linux-unwind.h
 	tmake_file="${tmake_file} i386/t-heap-trampoline"
 	;;
-i[34567]86-*-kfreebsd*-gnu | i[34567]86-*-kopensolaris*-gnu)
+i[34567]86-*-ironclad*-gnu | i[34567]86-*-kfreebsd*-gnu | i[34567]86-*-kopensolaris*-gnu)
 	extra_parts="$extra_parts crtprec32.o crtprec64.o crtprec80.o crtfastmath.o"
 	tmake_file="${tmake_file} i386/t-crtpc t-crtfm i386/t-crtstuff t-dfprules"
 	tm_file="${tm_file} i386/elf-lib.h"
diff --git gcc-host-clean/libstdc++-v3/configure.host gcc-host-workdir/libstdc++-v3/configure.host
index 253e5a9..0d620f6 100644
--- gcc-host-clean/libstdc++-v3/configure.host
+++ gcc-host-workdir/libstdc++-v3/configure.host
@@ -268,7 +268,7 @@ case "${host_os}" in
   linux-musl*)
     os_include_dir="os/generic"
     ;;
-  gnu* | linux* | kfreebsd*-gnu | uclinux*)
+  gnu* | linux* | kfreebsd*-gnu | uclinux* | ironclad*)
     if [ "$uclibc" = "yes" ]; then
       os_include_dir="os/uclibc"
     elif [ "$bionic" = "yes" ]; then
diff --git gcc-host-clean/libstdc++-v3/include/bits/ios_base.h gcc-host-workdir/libstdc++-v3/include/bits/ios_base.h
index b94b2cd..1a5eaa5 100644
--- gcc-host-clean/libstdc++-v3/include/bits/ios_base.h
+++ gcc-host-workdir/libstdc++-v3/include/bits/ios_base.h
@@ -225,7 +225,7 @@ _GLIBCXX_BEGIN_NAMESPACE_VERSION
 
   enum _Ios_Seekdir
     {
-      _S_beg = 0,
+      _S_beg = 1,
       _S_cur = _GLIBCXX_STDIO_SEEK_CUR,
       _S_end = _GLIBCXX_STDIO_SEEK_END,
       _S_ios_seekdir_end = 1L << 16
