diff --git icu-clean/source/data/Makefile.in icu-workdir/source/data/Makefile.in
index 266ad7a..71f7da2 100644
--- icu-clean/source/data/Makefile.in
+++ icu-workdir/source/data/Makefile.in
@@ -240,6 +240,7 @@ ifneq ($(ICUDATA_SOURCE_IS_NATIVE_TARGET),YES)
 	@-$(RMV) $@
 	$(INVOKE) $(TOOLBINDIR)/icupkg -d $(BUILDDIR) --list -x \* $(ICUDATA_SOURCE_ARCHIVE) -o $@
 else
+	@mkdir -p "$$(dirname $@)"
 	@echo "$@" > $@
 endif
 endif
