diff --git librsvg-clean/configure.ac librsvg-workdir/configure.ac
index 568d62d..079d0cc 100644
--- librsvg-clean/configure.ac
+++ librsvg-workdir/configure.ac
@@ -166,7 +166,7 @@ if test "$enable_pixbuf_loader" = "yes"; then
   # First try the standalone gdk-pixbuf
   gdk_pixbuf_binary_version="`$PKG_CONFIG --variable=gdk_pixbuf_binary_version gdk-pixbuf-2.0`"
   gdk_pixbuf_binarydir="`$PKG_CONFIG --variable=gdk_pixbuf_binarydir gdk-pixbuf-2.0`"
-  gdk_pixbuf_moduledir=`$PKG_CONFIG --variable gdk_pixbuf_moduledir gdk-pixbuf-2.0`
+  gdk_pixbuf_moduledir=`$PKG_CONFIG --variable gdk_pixbuf_moduledir gdk-pixbuf-2.0 | sed "s|/sysroot||g"`
   gdk_pixbuf_cache_file=`$PKG_CONFIG --variable gdk_pixbuf_cache_file gdk-pixbuf-2.0`
 fi
 
diff --git librsvg-clean/rsvg-css.c librsvg-workdir/rsvg-css.c
index 5b359a2..db1416d 100644
--- librsvg-clean/rsvg-css.c
+++ librsvg-workdir/rsvg-css.c
@@ -863,7 +863,7 @@ rsvg_css_parse_xml_attribute_string (const char *attribute_string)
 
     memset (&handler, 0, sizeof (handler));
     xmlSAX2InitDefaultSAXHandler (&handler, 0);
-    handler.serror = rsvg_xml_noerror;
+    handler.serror = (void (*)(void *, const struct _xmlError *))rsvg_xml_noerror;
     parser = xmlCreatePushParserCtxt (&handler, NULL, tag, strlen (tag) + 1, NULL);
     parser->options |= XML_PARSE_NONET;
 
