diff --git xdg-user-dirs-clean/meson.build xdg-user-dirs-workdir/meson.build
index 7861bff..a453bce 100644
--- xdg-user-dirs-clean/meson.build
+++ xdg-user-dirs-workdir/meson.build
@@ -22,6 +22,7 @@ else
   libintl_dep = cc.find_library('intl', required: true)
 endif
 
+libiconv_dep = cc.find_library('iconv', required: true)
 systemd_dep = dependency('systemd', required: false)
 
 #
@@ -87,7 +88,7 @@ add_project_arguments([
 xdg_user_dirs_update = executable('xdg-user-dirs-update',
   'xdg-user-dirs-update.c',
   include_directories: config_inc,
-  dependencies: [libintl_dep],
+  dependencies: [libintl_dep, libiconv_dep],
   install: true
 )
 
