diff --git xorg-xlsfonts-clean/xlsfonts.c xorg-xlsfonts-workdir/xlsfonts.c
index b834637..752f7f1 100644
--- xorg-xlsfonts-clean/xlsfonts.c
+++ xorg-xlsfonts-workdir/xlsfonts.c
@@ -201,7 +201,7 @@ get_list(const char *pattern)
             info = XLoadQueryFont(dpy, pattern);
 
             if (info) {
-                fonts = &pattern;
+                fonts = (char **)&pattern;
                 available = 1;
                 XUnloadFont(dpy, info->fid);
             }
