diff --git mesa-clean/include/drm-uapi/drm.h mesa-workdir/include/drm-uapi/drm.h
index ceb0f9e..956da42 100644
--- mesa-clean/include/drm-uapi/drm.h
+++ mesa-workdir/include/drm-uapi/drm.h
@@ -45,11 +45,7 @@ typedef unsigned int drm_handle_t;
 
 #include <stdint.h>
 #include <sys/types.h>
-#if defined(__GNU__)
 #include <sys/ioctl.h>
-#else
-#include <sys/ioccom.h>
-#endif
 typedef int8_t   __s8;
 typedef uint8_t  __u8;
 typedef int16_t  __s16;
diff --git mesa-clean/include/renderdoc_app.h mesa-workdir/include/renderdoc_app.h
index b2d10c5..3236a9a 100644
--- mesa-clean/include/renderdoc_app.h
+++ mesa-workdir/include/renderdoc_app.h
@@ -35,7 +35,7 @@
 
 #if defined(WIN32) || defined(__WIN32__) || defined(_WIN32) || defined(_MSC_VER)
 #define RENDERDOC_CC __cdecl
-#elif defined(__linux__)
+#elif defined(__linux__) || defined(__ironclad__)
 #define RENDERDOC_CC
 #elif defined(__APPLE__)
 #define RENDERDOC_CC
diff --git mesa-clean/meson.build mesa-workdir/meson.build
index d6e3a8b..89a3221 100644
--- mesa-clean/meson.build
+++ mesa-workdir/meson.build
@@ -1249,7 +1249,7 @@ if cc.has_function('fmemopen')
 endif
 
 # TODO: this is very incomplete
-if ['linux', 'cygwin', 'gnu', 'freebsd', 'gnu/kfreebsd', 'haiku', 'android', 'managarm'].contains(host_machine.system())
+if ['linux', 'cygwin', 'gnu', 'freebsd', 'gnu/kfreebsd', 'haiku', 'android', 'managarm', 'ironclad'].contains(host_machine.system())
   pre_args += '-D_GNU_SOURCE'
 elif host_machine.system() == 'sunos'
   pre_args += '-D__EXTENSIONS__'
diff --git mesa-clean/src/gallium/drivers/llvmpipe/lp_texture.c mesa-workdir/src/gallium/drivers/llvmpipe/lp_texture.c
index d682021..893d1d8 100644
--- mesa-clean/src/gallium/drivers/llvmpipe/lp_texture.c
+++ mesa-workdir/src/gallium/drivers/llvmpipe/lp_texture.c
@@ -1841,7 +1841,7 @@ llvmpipe_resource_get_param(struct pipe_screen *screen,
    default:
       break;
    }
-   assert(0);
+   // assert(0);
    *value = 0;
    return false;
 }
diff --git mesa-clean/src/util/detect_os.h mesa-workdir/src/util/detect_os.h
index 7403724..99e14a2 100644
--- mesa-clean/src/util/detect_os.h
+++ mesa-workdir/src/util/detect_os.h
@@ -28,6 +28,12 @@
 #define DETECT_OS_ANDROID 1
 #endif
 
+#if defined(__ironclad__)
+#define DETECT_OS_IRONCLAD 1
+#define DETECT_OS_UNIX 1
+#define DETECT_OS_POSIX 1
+#endif
+
 #if defined(__FreeBSD__) || defined(__FreeBSD_kernel__)
 #define DETECT_OS_FREEBSD 1
 #define DETECT_OS_BSD 1
@@ -147,4 +153,8 @@
 #define DETECT_OS_MANAGARM 0
 #endif
 
+#ifndef DETECT_OS_IRONCLAD
+#define DETECT_OS_IRONCLAD 0
+#endif
+
 #endif /* DETECT_OS_H */
diff --git mesa-clean/src/util/os_misc.c mesa-workdir/src/util/os_misc.c
index 16dd51b..f1253c2 100644
--- mesa-clean/src/util/os_misc.c
+++ mesa-workdir/src/util/os_misc.c
@@ -60,7 +60,7 @@
 #  include <unistd.h>
 #  include <android/log.h>
 #  include <sys/system_properties.h>
-#elif DETECT_OS_LINUX || DETECT_OS_CYGWIN || DETECT_OS_SOLARIS || DETECT_OS_HURD || DETECT_OS_MANAGARM
+#elif DETECT_OS_LINUX || DETECT_OS_CYGWIN || DETECT_OS_SOLARIS || DETECT_OS_HURD || DETECT_OS_MANAGARM || DETECT_OS_IRONCLAD
 #  include <unistd.h>
 #elif DETECT_OS_OPENBSD || DETECT_OS_FREEBSD
 #  include <sys/resource.h>
diff --git mesa-clean/src/util/os_time.c mesa-workdir/src/util/os_time.c
index 71fc7e3..ba0fe5c 100644
--- mesa-clean/src/util/os_time.c
+++ mesa-workdir/src/util/os_time.c
@@ -80,7 +80,7 @@ os_time_nanosleep_until(int64_t deadline)
 void
 os_time_sleep(int64_t usecs)
 {
-#if DETECT_OS_LINUX || DETECT_OS_MANAGARM || DETECT_OS_FUCHSIA
+#if DETECT_OS_LINUX || DETECT_OS_MANAGARM || DETECT_OS_FUCHSIA || DETECT_OS_IRONCLAD
    struct timespec time;
    time.tv_sec = usecs / 1000000;
    time.tv_nsec = (usecs % 1000000) * 1000;
diff --git mesa-clean/src/util/u_thread.c mesa-workdir/src/util/u_thread.c
index d1cf085..fc18f6a 100644
--- mesa-clean/src/util/u_thread.c
+++ mesa-workdir/src/util/u_thread.c
@@ -97,7 +97,7 @@ int u_thread_create(thrd_t *thrd, int (*routine)(void *), void *param)
 void u_thread_setname( const char *name )
 {
 #if defined(HAVE_PTHREAD)
-#if DETECT_OS_LINUX || DETECT_OS_CYGWIN || DETECT_OS_SOLARIS || defined(__GLIBC__) || DETECT_OS_MANAGARM || DETECT_OS_FUCHSIA
+#if DETECT_OS_LINUX || DETECT_OS_CYGWIN || DETECT_OS_SOLARIS || defined(__GLIBC__) || DETECT_OS_MANAGARM || DETECT_OS_IRONCLAD || DETECT_OS_FUCHSIA
    int ret = pthread_setname_np(pthread_self(), name);
    if (ret == ERANGE) {
       char buf[16];
@@ -176,7 +176,7 @@ util_set_thread_affinity(thrd_t thread,
 int64_t
 util_thread_get_time_nano(thrd_t thread)
 {
-#if defined(HAVE_PTHREAD) && !defined(__APPLE__) && !defined(__HAIKU__) && !defined(__managarm__)
+#if defined(HAVE_PTHREAD) && !defined(__APPLE__) && !defined(__HAIKU__) && !defined(__managarm__) && !defined(__ironclad__)
    struct timespec ts;
    clockid_t cid;
 
