diff --git slim-clean/app.cpp slim-workdir/app.cpp
index d378793..f549e62 100644
--- slim-clean/app.cpp
+++ slim-workdir/app.cpp
@@ -10,6 +10,7 @@
  *  (at your option) any later version.
  */
 
+#include <crypt.h>
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <sys/wait.h>
@@ -1167,10 +1168,12 @@ int App::StartServer()
 		}
 	}
 
+	/*
 	if (!hasVtSet && daemonmode)
 	{
 		server[argc++] = (char*)"vt07";
 	}
+	*/
 	server[argc] = NULL;
 
 	switch (ServerPID) {
diff --git slim-clean/app.h slim-workdir/app.h
index 7453e34..f0108ed 100644
--- slim-clean/app.h
+++ slim-workdir/app.h
@@ -13,6 +13,7 @@
 #ifndef _APP_H_
 #define _APP_H_
 
+#include <string>
 #include <X11/Xlib.h>
 
 #ifdef USE_PAM
