diff -urN --no-dereference dieharder-clean/dieharder/Makefile.in dieharder-workdir/dieharder/Makefile.in
--- dieharder-clean/dieharder/Makefile.in	2022-12-21 14:04:02.000000000 +0100
+++ dieharder-workdir/dieharder/Makefile.in	2022-12-21 14:04:02.000000000 +0100
@@ -421,7 +421,7 @@
 # CC = gcc
 
 # Compile flags (use fairly standard -O3 as default)
-AM_CPPFLAGS = -I ${top_srcdir}/include $(DEFINES) -I ${includedir}
+AM_CPPFLAGS = -I ${top_srcdir}/include $(DEFINES)
 AM_CFLAGS = -O3
 
 # Load flags (optional)
diff -urN --no-dereference dieharder-clean/libdieharder/Makefile.in dieharder-workdir/libdieharder/Makefile.in
--- dieharder-clean/libdieharder/Makefile.in	2022-12-21 14:04:02.000000000 +0100
+++ dieharder-workdir/libdieharder/Makefile.in	2022-12-21 14:04:02.000000000 +0100
@@ -575,7 +575,7 @@
 # CC = gcc
 
 # Compile flags (use fairly standard -O3 as default)
-AM_CPPFLAGS = $(DEFINES) -I ${top_srcdir}/include -I ${includedir}
+AM_CPPFLAGS = $(DEFINES) -I ${top_srcdir}/include
 AM_CFLAGS = -O3 -Wall -pedantic -Wno-unused-variable 
 
 #========================================================================
