diff --git dieharder-clean/dieharder/Makefile.in dieharder-workdir/dieharder/Makefile.in
index a224393..e987a27 100644
--- dieharder-clean/dieharder/Makefile.in
+++ dieharder-workdir/dieharder/Makefile.in
@@ -421,7 +421,7 @@ DEFINES = -DVERSION=$(VERSION)
 # CC = gcc
 
 # Compile flags (use fairly standard -O3 as default)
-AM_CPPFLAGS = -I ${top_srcdir}/include $(DEFINES) -I ${includedir}
+AM_CPPFLAGS = -I ${top_srcdir}/include $(DEFINES)
 AM_CFLAGS = -O3
 
 # Load flags (optional)
diff --git dieharder-clean/libdieharder/Makefile.in dieharder-workdir/libdieharder/Makefile.in
index b437ae3..0ca5463 100644
--- dieharder-clean/libdieharder/Makefile.in
+++ dieharder-workdir/libdieharder/Makefile.in
@@ -575,7 +575,7 @@ DEFINES = -DVERSION=$(VERSION) -DLITTLE_ENDIAN=$(LITTLE_ENDIAN)
 # CC = gcc
 
 # Compile flags (use fairly standard -O3 as default)
-AM_CPPFLAGS = $(DEFINES) -I ${top_srcdir}/include -I ${includedir}
+AM_CPPFLAGS = $(DEFINES) -I ${top_srcdir}/include
 AM_CFLAGS = -O3 -Wall -pedantic -Wno-unused-variable 
 
 #========================================================================
