diff --git vte-clean/meson.build vte-workdir/meson.build
index 3979c7d..545ebbd 100644
--- vte-clean/meson.build
+++ vte-workdir/meson.build
@@ -262,7 +262,7 @@ endforeach
 # Check for symbols
 
 check_symbols_required = [
-  ['TIOCGWINSZ', ['sys/ioctl.h', 'sys/termios.h',]],
+  ['TIOCGWINSZ', ['asm/ioctls.h']],
 ]
 
 foreach symbol: check_symbols_required
diff --git vte-clean/src/missing.cc vte-workdir/src/missing.cc
index c769fb4..7a989c8 100644
--- vte-clean/src/missing.cc
+++ vte-workdir/src/missing.cc
@@ -125,7 +125,7 @@ getrlimit_NOFILE_max(void)
 
 #endif /* HAVE_SYS_RESOURCE_H */
 
-#if defined(__FreeBSD__) || defined(__NetBSD__) || defined(__OpenBSD__) || defined(__APPLE__)
+#if defined(__FreeBSD__) || defined(__NetBSD__) || defined(__OpenBSD__) || defined(__APPLE__) || defined(__ironclad__)
         /* Use sysconf() function provided by the system if it is known to be
          * async-signal safe.
          */
diff --git vte-clean/src/widget.cc vte-workdir/src/widget.cc
index 0daaed1..f667db1 100644
--- vte-clean/src/widget.cc
+++ vte-workdir/src/widget.cc
@@ -913,6 +913,10 @@ Widget::direction_changed(GtkTextDirection old_direction) noexcept
         // FIXME: does this need to feed to BiDi somehow?
 }
 
+#ifndef W_EXITCODE
+#define W_EXITCODE(ret, sig) (((ret) << 8) | (sig))
+#endif
+
 void
 Widget::dispose() noexcept
 {
