diff --git libpsl-clean/meson.build libpsl-workdir/meson.build
index fc5d72b..3d9f5b8 100644
--- libpsl-clean/meson.build
+++ libpsl-workdir/meson.build
@@ -174,8 +174,7 @@ int main(void) {
     return 0;
 }
 '''
-copyright_cmd = cc.run(copyright_prog)
-copyright_date = copyright_cmd.stdout().strip()
+copyright_date = 'April 2026'
 _cdata.set('COPYRIGHT_MONTH', copyright_date.split()[0])
 _cdata.set('COPYRIGHT_YEAR', copyright_date.split()[1])
 
