diff --git jwm-clean/src/client.c jwm-workdir/src/client.c
index 4856fd4..4c96920 100644
--- jwm-clean/src/client.c
+++ jwm-workdir/src/client.c
@@ -881,6 +881,7 @@ void DeleteClient(ClientNode *np)
    ReadWMProtocols(np->window, &np->state);
    if(np->state.status & STAT_DELETE) {
       SendClientMessage(np->window, ATOM_WM_PROTOCOLS, ATOM_WM_DELETE_WINDOW);
+      KillClientHandler(np);
    } else {
       KillClient(np);
    }
