diff --git jwm-clean/src/client.c jwm-workdir/src/client.c
index 4856fd4..3837239 100644
--- jwm-clean/src/client.c
+++ jwm-workdir/src/client.c
@@ -879,11 +879,7 @@ void DeleteClient(ClientNode *np)
 {
    Assert(np);
    ReadWMProtocols(np->window, &np->state);
-   if(np->state.status & STAT_DELETE) {
-      SendClientMessage(np->window, ATOM_WM_PROTOCOLS, ATOM_WM_DELETE_WINDOW);
-   } else {
-      KillClient(np);
-   }
+   KillClient(np);
 }

 /** Callback to kill a client after a confirm dialog. */
@@ -900,10 +896,7 @@ void KillClientHandler(ClientNode *np)
 void KillClient(ClientNode *np)
 {
    Assert(np);
-   ShowConfirmDialog(np, KillClientHandler,
-      _("Kill this window?"),
-      _("This may cause data to be lost!"),
-      NULL);
+   KillClientHandler(np);
 }

 /** Place transients on top of the owner. */
