diff -urN --no-dereference jwm-clean/src/client.c jwm-workdir/src/client.c
--- jwm-clean/src/client.c	2024-11-09 18:29:29.000000000 +0100
+++ jwm-workdir/src/client.c	2024-11-09 18:29:29.000000000 +0100
@@ -879,11 +879,7 @@
 {
    Assert(np);
    ReadWMProtocols(np->window, &np->state);
-   if(np->state.status & STAT_DELETE) {
-      SendClientMessage(np->window, ATOM_WM_PROTOCOLS, ATOM_WM_DELETE_WINDOW);
-   } else {
-      KillClient(np);
-   }
+   KillClient(np);
 }
 
 /** Callback to kill a client after a confirm dialog. */
@@ -900,10 +896,7 @@
 void KillClient(ClientNode *np)
 {
    Assert(np);
-   ShowConfirmDialog(np, KillClientHandler,
-      _("Kill this window?"),
-      _("This may cause data to be lost!"),
-      NULL);
+   KillClientHandler(np);
 }
 
 /** Place transients on top of the owner. */
