diff --git libxcb-clean/configure libxcb-workdir/configure
index 33f5d89..30b74b4 100755
--- libxcb-clean/configure
+++ libxcb-workdir/configure
@@ -21027,7 +21027,7 @@ printf "%s\n" "yes" >&6; }
 fi
 NEEDED="xau >= 0.99.2"
 case $host_os in
-linux*|darwin*|solaris*|dragonfly*|freebsd*|netbsd*) ;;
+linux*|darwin*|solaris*|dragonfly*|freebsd*|netbsd*|ironclad*) ;;
      *) NEEDED="$NEEDED pthread-stubs" ;;
 esac
 
