diff --git xfe-clean/configure.ac xfe-workdir/configure.ac
index 9393875..57a4636 100644
--- xfe-clean/configure.ac
+++ xfe-workdir/configure.ac
@@ -69,9 +69,9 @@ AC_FUNC_GETGROUPS
 AC_FUNC_GETMNTENT
 AC_FUNC_LSTAT
 AC_FUNC_LSTAT_FOLLOWS_SLASHED_SYMLINK
-AC_FUNC_MALLOC
+#AC_FUNC_MALLOC
 AC_FUNC_MKTIME
-AC_FUNC_REALLOC
+#AC_FUNC_REALLOC
 AC_FUNC_STAT
 AC_FUNC_UTIME_NULL
 AC_CHECK_FUNCS([endgrent endpwent gethostname getmntent gettimeofday lchown memset mkdir mkfifo putenv rmdir setlocale sqrt strchr strdup strerror strstr strtol strtoul strtoull utime])
@@ -87,25 +87,19 @@ AC_CHECK_LIB(FOX-1.6,fxfindfox,,AC_MSG_ERROR("libFOX-1.6 not found"))
 
 # Check for FOX 1.6 header files
 AC_LANG([C++])
-AC_CHECK_HEADER(fox-1.6/fx.h,,AC_MSG_ERROR("fox-1.6/fx.h not found"))
+#AC_CHECK_HEADER(fox-1.6/fx.h,,AC_MSG_ERROR("fox-1.6/fx.h not found"))
 
 
-# Check if fox-config exists
-AC_CHECK_PROGS(FOX_CONFIG,fox-config-1.6 fox-1.6-config fox-config)
-if test no"$FOX_CONFIG" = no ; then
-	AC_MSG_ERROR("fox-config not found")               
-fi
-
 
 # Include flags for the FOX library
-FOXCFLAGS=`$FOX_CONFIG --cflags`
+FOXCFLAGS=`$PKG_CONFIG --cflags fox`
 CXXFLAGS="${CXXFLAGS} $FOXCFLAGS"
 
 
 # Check if FOX was compiled with xft support
-TEST_XFT=`$FOX_CONFIG --libs | grep Xft`
+TEST_XFT=yes
 if test "x$TEST_XFT" != "x" ; then
-	
+
 	echo "checking whether FOX was compiled with Xft support... yes"
 
 	# Check for FreeType2
@@ -135,7 +129,7 @@ if test "x$TEST_XFT" != "x" ; then
 			CXXFLAGS="$CXXFLAGS -DHAVE_XFT_H"
 		], AC_MSG_ERROR("Xft not found"))
 	fi
-	AC_CHECK_HEADER(X11/Xft/Xft.h,,AC_MSG_ERROR("Xft.h not found"))
+	#AC_CHECK_HEADER(X11/Xft/Xft.h,,AC_MSG_ERROR("Xft.h not found"))
 
 else
 	echo "checking whether FOX was compiled with Xft support... no"
@@ -149,28 +143,28 @@ else
 	echo "   sudo make install"
 	echo "============================================================================================="
 	echo ""
-	AC_MSG_ERROR("missing Xft support in FOX")               
+	AC_MSG_ERROR("missing Xft support in FOX")
 fi
 
 
 # Check for Xlib headers
-AC_CHECK_HEADER(X11/Xlib.h,,AC_MSG_ERROR("Xlib.h not found"))
+#AC_CHECK_HEADER(X11/Xlib.h,,AC_MSG_ERROR("Xlib.h not found"))
 
 # Check for XRandR support
 AC_MSG_CHECKING(for xrandr extension)
 AC_ARG_WITH(xrandr,[  --with-xrandr           compile with XRandR support])
 AC_MSG_RESULT([$with_xrandr])
-if test "x$with_xrandr" != "xno"; then
-AC_CHECK_HEADERS(X11/extensions/Xrandr.h,CXXFLAGS="${CXXFLAGS} -DHAVE_XRANDR_H=1"; LIBS="${LIBS} -lXrandr")
-fi
+#if test "x$with_xrandr" != "xno"; then
+#AC_CHECK_HEADERS(X11/extensions/Xrandr.h,CXXFLAGS="${CXXFLAGS} -DHAVE_XRANDR_H=1"; LIBS="${LIBS} -lXrandr")
+#fi
 
 # Check for libPNG
 AC_CHECK_LIB(png, png_read_info,,AC_MSG_ERROR("libPNG not found"))
-AC_CHECK_HEADER(png.h,,AC_MSG_ERROR("png.h not found"))
+#AC_CHECK_HEADER(png.h,,AC_MSG_ERROR("png.h not found"))
 
 # Check for fontconfig
 AC_CHECK_LIB(fontconfig, FcInit,, AC_MSG_ERROR("fontconfig not found"))
-AC_CHECK_HEADER(fontconfig/fontconfig.h,,AC_MSG_ERROR("fontconfig.h not found"))
+#AC_CHECK_HEADER(fontconfig/fontconfig.h,,AC_MSG_ERROR("fontconfig.h not found"))
 
 # Check for startup notification support
 AC_MSG_CHECKING(for startup notification)
diff --git xfe-clean/src/ArchInputDialog.cpp xfe-workdir/src/ArchInputDialog.cpp
index 9e0a271..b9c5eec 100644
--- xfe-clean/src/ArchInputDialog.cpp
+++ xfe-workdir/src/ArchInputDialog.cpp
@@ -1,5 +1,5 @@
 // Input dialog for the add to archive command
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/BrowseInputDialog.cpp xfe-workdir/src/BrowseInputDialog.cpp
index 4c64e68..f60e3a4 100644
--- xfe-clean/src/BrowseInputDialog.cpp
+++ xfe-workdir/src/BrowseInputDialog.cpp
@@ -1,5 +1,5 @@
 // Input dialog with file browse icon
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/CommandWindow.cpp xfe-workdir/src/CommandWindow.cpp
index d9a9541..95b5fb2 100644
--- xfe-clean/src/CommandWindow.cpp
+++ xfe-workdir/src/CommandWindow.cpp
@@ -5,7 +5,7 @@
 // The object deletes itself when the close button is pressed
 // The command window can be a free-floating window or can be
 // a window which will always float over the owner window
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/DialogBox.cpp xfe-workdir/src/DialogBox.cpp
index e5fad00..c2c6b0f 100644
--- xfe-clean/src/DialogBox.cpp
+++ xfe-workdir/src/DialogBox.cpp
@@ -1,4 +1,5 @@
 // Dialog Box with additional toggle option
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/DirHistBox.cpp xfe-workdir/src/DirHistBox.cpp
index 1708b05..958efcb 100644
--- xfe-clean/src/DirHistBox.cpp
+++ xfe-workdir/src/DirHistBox.cpp
@@ -1,6 +1,6 @@
 // Display a history list box and allows the user to select a string
 // This is based on FXChoiceBox
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/DirList.cpp xfe-workdir/src/DirList.cpp
index c750a49..7d07579 100644
--- xfe-clean/src/DirList.cpp
+++ xfe-workdir/src/DirList.cpp
@@ -1,7 +1,7 @@
 // Directory list. Taken from the FOX library and slightly modified.
 // The compare(), compare_nolocale() and compare_locale() functions are adapted from a patch
 // submitted by Vladimir Támara Patiño
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/DirPanel.cpp xfe-workdir/src/DirPanel.cpp
index 5be5263..ecb219c 100644
--- xfe-clean/src/DirPanel.cpp
+++ xfe-workdir/src/DirPanel.cpp
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/ExecuteBox.cpp xfe-workdir/src/ExecuteBox.cpp
index 83c201f..3b403aa 100644
--- xfe-clean/src/ExecuteBox.cpp
+++ xfe-workdir/src/ExecuteBox.cpp
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/File.cpp xfe-workdir/src/File.cpp
index e42092a..a4d8a65 100644
--- xfe-clean/src/File.cpp
+++ xfe-workdir/src/File.cpp
@@ -1,5 +1,5 @@
 // File management class with progress dialog
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/FileDialog.cpp xfe-workdir/src/FileDialog.cpp
index 65f6085..b8bfd72 100644
--- xfe-clean/src/FileDialog.cpp
+++ xfe-workdir/src/FileDialog.cpp
@@ -1,5 +1,5 @@
 // File dialog. Taken from the FOX library and slightly modified.
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/FileDict.cpp xfe-workdir/src/FileDict.cpp
index 8a0f915..3e33bfd 100644
--- xfe-clean/src/FileDict.cpp
+++ xfe-workdir/src/FileDict.cpp
@@ -1,5 +1,5 @@
 // File association tables. Taken from the FOX library and slightly modified.
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/FileList.cpp xfe-workdir/src/FileList.cpp
index fb335e1..8639b8a 100644
--- xfe-clean/src/FileList.cpp
+++ xfe-workdir/src/FileList.cpp
@@ -2,7 +2,7 @@
 // The compare(), compare_nolocale() and compare_locale() functions are adapted from a patch
 // submitted by Vladimir Támara Patiño
 
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/FilePanel.cpp xfe-workdir/src/FilePanel.cpp
index 403b7aa..1eecd46 100644
--- xfe-clean/src/FilePanel.cpp
+++ xfe-workdir/src/FilePanel.cpp
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/FontDialog.cpp xfe-workdir/src/FontDialog.cpp
index 0fd78a1..9df1478 100644
--- xfe-clean/src/FontDialog.cpp
+++ xfe-workdir/src/FontDialog.cpp
@@ -1,5 +1,5 @@
 // Font dialog. Taken from the FOX library and only modified for translation purpose
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/HistInputDialog.cpp xfe-workdir/src/HistInputDialog.cpp
index 7faadcc..3df5161 100644
--- xfe-clean/src/HistInputDialog.cpp
+++ xfe-workdir/src/HistInputDialog.cpp
@@ -1,5 +1,5 @@
 // Input dialog with history list and an optional check box
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/IconList.cpp xfe-workdir/src/IconList.cpp
index 364fdf9..00ca802 100644
--- xfe-clean/src/IconList.cpp
+++ xfe-workdir/src/IconList.cpp
@@ -1,4 +1,5 @@
 // Icon list. Taken from the FOX library and slightly modified.
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/InputDialog.cpp xfe-workdir/src/InputDialog.cpp
index 27f4d77..fce88f5 100644
--- xfe-clean/src/InputDialog.cpp
+++ xfe-workdir/src/InputDialog.cpp
@@ -1,5 +1,5 @@
 // Simple input dialog (without history)
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/MessageBox.cpp xfe-workdir/src/MessageBox.cpp
index 768a62e..6516589 100644
--- xfe-clean/src/MessageBox.cpp
+++ xfe-workdir/src/MessageBox.cpp
@@ -1,6 +1,6 @@
 // Message box. Taken from the FOX library and slightly modified for translation purpose.
 // Also added a SU button
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/OverwriteBox.cpp xfe-workdir/src/OverwriteBox.cpp
index 2109677..7ec9f90 100644
--- xfe-clean/src/OverwriteBox.cpp
+++ xfe-workdir/src/OverwriteBox.cpp
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/Preferences.cpp xfe-workdir/src/Preferences.cpp
index df00073..c8b71ca 100644
--- xfe-clean/src/Preferences.cpp
+++ xfe-workdir/src/Preferences.cpp
@@ -1,5 +1,5 @@
 // Preferences dialog box
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/Properties.cpp xfe-workdir/src/Properties.cpp
index fe1fb6f..b2b9555 100644
--- xfe-clean/src/Properties.cpp
+++ xfe-workdir/src/Properties.cpp
@@ -1,5 +1,5 @@
 // Properties box
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/SearchPanel.cpp xfe-workdir/src/SearchPanel.cpp
index 815ec5b..0c766e7 100644
--- xfe-clean/src/SearchPanel.cpp
+++ xfe-workdir/src/SearchPanel.cpp
@@ -1,4 +1,5 @@
 // Search panel
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/SearchWindow.cpp xfe-workdir/src/SearchWindow.cpp
index c46637d..3dd6731 100644
--- xfe-clean/src/SearchWindow.cpp
+++ xfe-workdir/src/SearchWindow.cpp
@@ -1,4 +1,5 @@
 // Search files dialog and panel
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/TextWindow.cpp xfe-workdir/src/TextWindow.cpp
index 0cb00ab..c9a4160 100644
--- xfe-clean/src/TextWindow.cpp
+++ xfe-workdir/src/TextWindow.cpp
@@ -1,4 +1,5 @@
 // Text widget with a close button
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/WriteWindow.cpp xfe-workdir/src/WriteWindow.cpp
index a45d151..3ee2067 100644
--- xfe-clean/src/WriteWindow.cpp
+++ xfe-workdir/src/WriteWindow.cpp
@@ -1,6 +1,6 @@
 // This is adapted from 'adie', a text editor found
 // in the FOX library and written by Jeroen van der Zijp.
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/XFileExplorer.cpp xfe-workdir/src/XFileExplorer.cpp
index 7ef91f0..5c7ed17 100644
--- xfe-clean/src/XFileExplorer.cpp
+++ xfe-workdir/src/XFileExplorer.cpp
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/XFileImage.cpp xfe-workdir/src/XFileImage.cpp
index 01a045c..5d0f7d4 100644
--- xfe-clean/src/XFileImage.cpp
+++ xfe-workdir/src/XFileImage.cpp
@@ -1,6 +1,6 @@
 // This code is adapted from 'imageviewer', a demo image viewer found
 // in the FOX library and written by Jeroen van der Zijp.
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/XFilePackage.cpp xfe-workdir/src/XFilePackage.cpp
index 28cd76e..acf1eaf 100644
--- xfe-clean/src/XFilePackage.cpp
+++ xfe-workdir/src/XFilePackage.cpp
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/XFileWrite.cpp xfe-workdir/src/XFileWrite.cpp
index 870eeeb..9bed875 100644
--- xfe-clean/src/XFileWrite.cpp
+++ xfe-workdir/src/XFileWrite.cpp
@@ -1,6 +1,6 @@
 // This is adapted from 'adie', a demo text editor found
 // in the FOX library and written by Jeroen van der Zijp.
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/icons.cpp xfe-workdir/src/icons.cpp
index 7754362..f8b74eb 100644
--- xfe-clean/src/icons.cpp
+++ xfe-workdir/src/icons.cpp
@@ -1,5 +1,5 @@
 // Global icons for all applications
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/main.cpp xfe-workdir/src/main.cpp
index e8e58ea..46b5e0c 100644
--- xfe-clean/src/main.cpp
+++ xfe-workdir/src/main.cpp
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/startupnotification.cpp xfe-workdir/src/startupnotification.cpp
index 27c0b7a..d16df00 100644
--- xfe-clean/src/startupnotification.cpp
+++ xfe-workdir/src/startupnotification.cpp
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/xfeutils.cpp xfe-workdir/src/xfeutils.cpp
index ad7aeba..4572a12 100644
--- xfe-clean/src/xfeutils.cpp
+++ xfe-workdir/src/xfeutils.cpp
@@ -3,7 +3,7 @@
 // Also supports a timeout on the stat and lstat function (this is the reason
 // why some standard FOX function cannot be used and are rewritten here)
 
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/st/st.c xfe-workdir/st/st.c
index a80d8e9..4673220 100644
--- xfe-clean/st/st.c
+++ xfe-workdir/st/st.c
@@ -22,17 +22,22 @@
 #include <termios.h>
 #include <unistd.h>
 #include <wchar.h>
+#include <sys/ttydefaults.h>
 
 #include "st.h"
 #include "win.h"
 
+/*
 #if   defined(__linux)
+*/
  #include <pty.h>
+/*
 #elif defined(__OpenBSD__) || defined(__NetBSD__) || defined(__APPLE__)
  #include <util.h>
 #elif defined(__FreeBSD__) || defined(__DragonFly__)
  #include <libutil.h>
 #endif
+*/
 
 /* Arbitrary sizes */
 #define UTF_INVALID   0xFFFD
