diff --git xfe-clean/configure.ac xfe-workdir/configure.ac
index 03fd881..9744ce2 100644
--- xfe-clean/configure.ac
+++ xfe-workdir/configure.ac
@@ -69,9 +69,9 @@ AC_FUNC_GETGROUPS
 AC_FUNC_GETMNTENT
 AC_FUNC_LSTAT
 AC_FUNC_LSTAT_FOLLOWS_SLASHED_SYMLINK
-AC_FUNC_MALLOC
+#AC_FUNC_MALLOC
 AC_FUNC_MKTIME
-AC_FUNC_REALLOC
+#AC_FUNC_REALLOC
 AC_FUNC_STAT
 AC_FUNC_UTIME_NULL
 AC_CHECK_FUNCS([endgrent endpwent gethostname getmntent gettimeofday lchown memset mkdir mkfifo putenv rmdir setlocale sqrt strchr strdup strerror strstr strtol strtoul strtoull utime])
@@ -87,25 +87,19 @@ AC_CHECK_LIB(FOX-1.6,fxfindfox,,AC_MSG_ERROR("libFOX-1.6 not found"))
 
 # Check for FOX 1.6 header files
 AC_LANG([C++])
-AC_CHECK_HEADER(fox-1.6/fx.h,,AC_MSG_ERROR("fox-1.6/fx.h not found"))
+#AC_CHECK_HEADER(fox-1.6/fx.h,,AC_MSG_ERROR("fox-1.6/fx.h not found"))
 
 
-# Check if fox-config exists
-AC_CHECK_PROGS(FOX_CONFIG,fox-config-1.6 fox-1.6-config fox-config)
-if test no"$FOX_CONFIG" = no ; then
-	AC_MSG_ERROR("fox-config not found")               
-fi
-
 
 # Include flags for the FOX library
-FOXCFLAGS=`$FOX_CONFIG --cflags`
+FOXCFLAGS=`$PKG_CONFIG --cflags fox`
 CXXFLAGS="${CXXFLAGS} $FOXCFLAGS"
 
 
 # Check if FOX was compiled with xft support
-TEST_XFT=`$FOX_CONFIG --libs | grep Xft`
+TEST_XFT=yes
 if test "x$TEST_XFT" != "x" ; then
-	
+
 	echo "checking whether FOX was compiled with Xft support... yes"
 
 	# Check for FreeType2
@@ -135,7 +129,7 @@ if test "x$TEST_XFT" != "x" ; then
 			CXXFLAGS="$CXXFLAGS -DHAVE_XFT_H"
 		], AC_MSG_ERROR("Xft not found"))
 	fi
-	AC_CHECK_HEADER(X11/Xft/Xft.h,,AC_MSG_ERROR("Xft.h not found"))
+	#AC_CHECK_HEADER(X11/Xft/Xft.h,,AC_MSG_ERROR("Xft.h not found"))
 
 else
 	echo "checking whether FOX was compiled with Xft support... no"
@@ -149,28 +143,28 @@ else
 	echo "   sudo make install"
 	echo "============================================================================================="
 	echo ""
-	AC_MSG_ERROR("missing Xft support in FOX")               
+	AC_MSG_ERROR("missing Xft support in FOX")
 fi
 
 
 # Check for Xlib headers
-AC_CHECK_HEADER(X11/Xlib.h,,AC_MSG_ERROR("Xlib.h not found"))
+#AC_CHECK_HEADER(X11/Xlib.h,,AC_MSG_ERROR("Xlib.h not found"))
 
 # Check for XRandR support
 AC_MSG_CHECKING(for xrandr extension)
 AC_ARG_WITH(xrandr,[  --with-xrandr           compile with XRandR support])
 AC_MSG_RESULT([$with_xrandr])
-if test "x$with_xrandr" != "xno"; then
-AC_CHECK_HEADERS(X11/extensions/Xrandr.h,CXXFLAGS="${CXXFLAGS} -DHAVE_XRANDR_H=1"; LIBS="${LIBS} -lXrandr")
-fi
+#if test "x$with_xrandr" != "xno"; then
+#AC_CHECK_HEADERS(X11/extensions/Xrandr.h,CXXFLAGS="${CXXFLAGS} -DHAVE_XRANDR_H=1"; LIBS="${LIBS} -lXrandr")
+#fi
 
 # Check for libPNG
 AC_CHECK_LIB(png, png_read_info,,AC_MSG_ERROR("libPNG not found"))
-AC_CHECK_HEADER(png.h,,AC_MSG_ERROR("png.h not found"))
+#AC_CHECK_HEADER(png.h,,AC_MSG_ERROR("png.h not found"))
 
 # Check for fontconfig
 AC_CHECK_LIB(fontconfig, FcInit,, AC_MSG_ERROR("fontconfig not found"))
-AC_CHECK_HEADER(fontconfig/fontconfig.h,,AC_MSG_ERROR("fontconfig.h not found"))
+#AC_CHECK_HEADER(fontconfig/fontconfig.h,,AC_MSG_ERROR("fontconfig.h not found"))
 
 # Check for startup notification support
 AC_MSG_CHECKING(for startup notification)
diff --git xfe-clean/src/ArchInputDialog.cpp xfe-workdir/src/ArchInputDialog.cpp
index 5d2490e..222bc6c 100644
--- xfe-clean/src/ArchInputDialog.cpp
+++ xfe-workdir/src/ArchInputDialog.cpp
@@ -1,5 +1,5 @@
 // Input dialog for the add to archive command
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/BrowseInputDialog.cpp xfe-workdir/src/BrowseInputDialog.cpp
index 346776f..861158f 100644
--- xfe-clean/src/BrowseInputDialog.cpp
+++ xfe-workdir/src/BrowseInputDialog.cpp
@@ -1,5 +1,5 @@
 // Input dialog with file browse icon
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/CommandWindow.cpp xfe-workdir/src/CommandWindow.cpp
index 7f6ae02..c29c8b1 100644
--- xfe-clean/src/CommandWindow.cpp
+++ xfe-workdir/src/CommandWindow.cpp
@@ -5,7 +5,7 @@
 // The object deletes itself when the close button is pressed
 // The command window can be a free-floating window or can be
 // a window which will always float over the owner window
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/DialogBox.cpp xfe-workdir/src/DialogBox.cpp
index a58d9c0..87e93bb 100644
--- xfe-clean/src/DialogBox.cpp
+++ xfe-workdir/src/DialogBox.cpp
@@ -1,4 +1,5 @@
 // Dialog Box with additional toggle option
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/DirHistBox.cpp xfe-workdir/src/DirHistBox.cpp
index 6e0c24f..6e3739c 100644
--- xfe-clean/src/DirHistBox.cpp
+++ xfe-workdir/src/DirHistBox.cpp
@@ -1,6 +1,6 @@
 // Display a history list box and allows the user to select a string
 // This is based on FXChoiceBox
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/DirList.cpp xfe-workdir/src/DirList.cpp
index 8113c52..bc9d7ed 100644
--- xfe-clean/src/DirList.cpp
+++ xfe-workdir/src/DirList.cpp
@@ -1,7 +1,7 @@
 // Directory list. Taken from the FOX library and slightly modified.
 // The compare(), compare_nolocale() and compare_locale() functions are adapted from a patch
 // submitted by Vladimir Támara Patiño
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/DirPanel.cpp xfe-workdir/src/DirPanel.cpp
index 3fbac85..10561cd 100644
--- xfe-clean/src/DirPanel.cpp
+++ xfe-workdir/src/DirPanel.cpp
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/ExecuteBox.cpp xfe-workdir/src/ExecuteBox.cpp
index 10fa3f0..d730c33 100644
--- xfe-clean/src/ExecuteBox.cpp
+++ xfe-workdir/src/ExecuteBox.cpp
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/File.cpp xfe-workdir/src/File.cpp
index 118415a..5a5f070 100644
--- xfe-clean/src/File.cpp
+++ xfe-workdir/src/File.cpp
@@ -1,5 +1,5 @@
 // File management class with progress dialog
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/FileDialog.cpp xfe-workdir/src/FileDialog.cpp
index 4cb7466..d734129 100644
--- xfe-clean/src/FileDialog.cpp
+++ xfe-workdir/src/FileDialog.cpp
@@ -1,5 +1,5 @@
 // File dialog. Taken from the FOX library and slightly modified.
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/FileDict.cpp xfe-workdir/src/FileDict.cpp
index 1a76655..37f1bca 100644
--- xfe-clean/src/FileDict.cpp
+++ xfe-workdir/src/FileDict.cpp
@@ -1,5 +1,5 @@
 // File association tables. Taken from the FOX library and slightly modified.
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/FileList.cpp xfe-workdir/src/FileList.cpp
index d3f514c..5e7b408 100644
--- xfe-clean/src/FileList.cpp
+++ xfe-workdir/src/FileList.cpp
@@ -2,7 +2,7 @@
 // The compare(), compare_nolocale() and compare_locale() functions are adapted from a patch
 // submitted by Vladimir Támara Patiño
 
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/FilePanel.cpp xfe-workdir/src/FilePanel.cpp
index 8e6d158..92f73c5 100644
--- xfe-clean/src/FilePanel.cpp
+++ xfe-workdir/src/FilePanel.cpp
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/FontDialog.cpp xfe-workdir/src/FontDialog.cpp
index 7fbfe06..95baf56 100644
--- xfe-clean/src/FontDialog.cpp
+++ xfe-workdir/src/FontDialog.cpp
@@ -1,5 +1,5 @@
 // Font dialog. Taken from the FOX library and only modified for translation purpose
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/HistInputDialog.cpp xfe-workdir/src/HistInputDialog.cpp
index c5b8d39..4c4e564 100644
--- xfe-clean/src/HistInputDialog.cpp
+++ xfe-workdir/src/HistInputDialog.cpp
@@ -1,5 +1,5 @@
 // Input dialog with history list and an optional check box
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/IconList.cpp xfe-workdir/src/IconList.cpp
index 3846914..1242618 100644
--- xfe-clean/src/IconList.cpp
+++ xfe-workdir/src/IconList.cpp
@@ -1,4 +1,5 @@
 // Icon list. Taken from the FOX library and slightly modified.
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/InputDialog.cpp xfe-workdir/src/InputDialog.cpp
index 1999aff..744261f 100644
--- xfe-clean/src/InputDialog.cpp
+++ xfe-workdir/src/InputDialog.cpp
@@ -1,5 +1,5 @@
 // Simple input dialog (without history)
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/MessageBox.cpp xfe-workdir/src/MessageBox.cpp
index 7bf48d3..ccf00fb 100644
--- xfe-clean/src/MessageBox.cpp
+++ xfe-workdir/src/MessageBox.cpp
@@ -1,6 +1,6 @@
 // Message box. Taken from the FOX library and slightly modified for translation purpose.
 // Also added a SU button
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/OverwriteBox.cpp xfe-workdir/src/OverwriteBox.cpp
index d352957..9c5e3f4 100644
--- xfe-clean/src/OverwriteBox.cpp
+++ xfe-workdir/src/OverwriteBox.cpp
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/Preferences.cpp xfe-workdir/src/Preferences.cpp
index c7c49ca..2dec50a 100644
--- xfe-clean/src/Preferences.cpp
+++ xfe-workdir/src/Preferences.cpp
@@ -1,5 +1,5 @@
 // Preferences dialog box
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/Properties.cpp xfe-workdir/src/Properties.cpp
index d17d0d8..92c1e35 100644
--- xfe-clean/src/Properties.cpp
+++ xfe-workdir/src/Properties.cpp
@@ -1,5 +1,5 @@
 // Properties box
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/SearchPanel.cpp xfe-workdir/src/SearchPanel.cpp
index 6cc65a2..56d5151 100644
--- xfe-clean/src/SearchPanel.cpp
+++ xfe-workdir/src/SearchPanel.cpp
@@ -1,4 +1,5 @@
 // Search panel
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/SearchWindow.cpp xfe-workdir/src/SearchWindow.cpp
index f7fa356..5347e01 100644
--- xfe-clean/src/SearchWindow.cpp
+++ xfe-workdir/src/SearchWindow.cpp
@@ -1,4 +1,5 @@
 // Search files dialog and panel
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/TextWindow.cpp xfe-workdir/src/TextWindow.cpp
index 33f2f84..85b5d99 100644
--- xfe-clean/src/TextWindow.cpp
+++ xfe-workdir/src/TextWindow.cpp
@@ -1,4 +1,5 @@
 // Text widget with a close button
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/WriteWindow.cpp xfe-workdir/src/WriteWindow.cpp
index 28b5219..0bb1135 100644
--- xfe-clean/src/WriteWindow.cpp
+++ xfe-workdir/src/WriteWindow.cpp
@@ -1,6 +1,6 @@
 // This is adapted from 'adie', a text editor found
 // in the FOX library and written by Jeroen van der Zijp.
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/XFileExplorer.cpp xfe-workdir/src/XFileExplorer.cpp
index 54b867f..bf7a89a 100644
--- xfe-clean/src/XFileExplorer.cpp
+++ xfe-workdir/src/XFileExplorer.cpp
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/XFileImage.cpp xfe-workdir/src/XFileImage.cpp
index af06bb2..e287860 100644
--- xfe-clean/src/XFileImage.cpp
+++ xfe-workdir/src/XFileImage.cpp
@@ -1,6 +1,6 @@
 // This code is adapted from 'imageviewer', a demo image viewer found
 // in the FOX library and written by Jeroen van der Zijp.
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/XFilePackage.cpp xfe-workdir/src/XFilePackage.cpp
index 99a288f..2b768c3 100644
--- xfe-clean/src/XFilePackage.cpp
+++ xfe-workdir/src/XFilePackage.cpp
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/XFileWrite.cpp xfe-workdir/src/XFileWrite.cpp
index 9d666a1..244aa97 100644
--- xfe-clean/src/XFileWrite.cpp
+++ xfe-workdir/src/XFileWrite.cpp
@@ -1,6 +1,6 @@
 // This is adapted from 'adie', a demo text editor found
 // in the FOX library and written by Jeroen van der Zijp.
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/icons.cpp xfe-workdir/src/icons.cpp
index d473a08..a32a01e 100644
--- xfe-clean/src/icons.cpp
+++ xfe-workdir/src/icons.cpp
@@ -1,5 +1,5 @@
 // Global icons for all applications
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/main.cpp xfe-workdir/src/main.cpp
index 5996aca..f78a43c 100644
--- xfe-clean/src/main.cpp
+++ xfe-workdir/src/main.cpp
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
@@ -389,6 +390,7 @@ int main(int argc, char* argv[])
     textdomain(PACKAGE);
 #endif
 
+    FXuint pkg_format;
 #if defined(linux)
     // For package query on Linux systems, try to guess if the default package format is deb or rpm:
     //   - if dpkg exists then the system uses deb packages
@@ -398,7 +400,6 @@ int main(int argc, char* argv[])
     cmd = "dpkg --version";
     FXString str = xf_getcommandoutput(cmd);
 
-    FXuint pkg_format;
     if (str.find("Debian") != -1)
     {
         pkg_format = DEB_PKG; // deb based system
@@ -417,6 +418,8 @@ int main(int argc, char* argv[])
             pkg_format = OTHER_PKG; // other (unsupported) package system
         }
     }
+#else
+    pkg_format = OTHER_PKG;
 #endif
 
     // Parse basic arguments
diff --git xfe-clean/src/startupnotification.cpp xfe-workdir/src/startupnotification.cpp
index 6a8a6b1..93ec81a 100644
--- xfe-clean/src/startupnotification.cpp
+++ xfe-workdir/src/startupnotification.cpp
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/src/xfedefs.h xfe-workdir/src/xfedefs.h
index 2c30d42..2158104 100644
--- xfe-clean/src/xfedefs.h
+++ xfe-workdir/src/xfedefs.h
@@ -366,9 +366,9 @@ enum
 #define MTAB_PATH    "/proc/mounts"
 #endif
 
+#endif
+
 // Package format
 #define DEB_PKG      0
 #define RPM_PKG      1
 #define OTHER_PKG    2
-
-#endif
diff --git xfe-clean/src/xfeutils.cpp xfe-workdir/src/xfeutils.cpp
index e4f070f..284b1e6 100644
--- xfe-clean/src/xfeutils.cpp
+++ xfe-workdir/src/xfeutils.cpp
@@ -3,7 +3,7 @@
 // Also supports a timeout on the stat and lstat function (this is the reason
 // why some standard FOX function cannot be used and are rewritten here)
 
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff --git xfe-clean/st/st.c xfe-workdir/st/st.c
index a80d8e9..72fefc5 100644
--- xfe-clean/st/st.c
+++ xfe-workdir/st/st.c
@@ -22,17 +22,22 @@
 #include <termios.h>
 #include <unistd.h>
 #include <wchar.h>
+#include <sys/ttydefaults.h>
 
 #include "st.h"
 #include "win.h"
 
+/*
 #if   defined(__linux)
+*/
  #include <pty.h>
+/*
 #elif defined(__OpenBSD__) || defined(__NetBSD__) || defined(__APPLE__)
  #include <util.h>
 #elif defined(__FreeBSD__) || defined(__DragonFly__)
  #include <libutil.h>
 #endif
+*/
 
 /* Arbitrary sizes */
 #define UTF_INVALID   0xFFFD
