diff -urN --no-dereference xfe-clean/configure xfe-workdir/configure
--- xfe-clean/configure	2025-03-15 12:25:48.000000000 +0100
+++ xfe-workdir/configure	2025-04-26 23:14:56.143336025 +0200
@@ -12529,66 +12529,13 @@
 
 
 
-# Check if fox-config exists
-for ac_prog in fox-config-1.6 fox-1.6-config fox-config
-do
-  # Extract the first word of "$ac_prog", so it can be a program name with args.
-set dummy $ac_prog; ac_word=$2
-{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for $ac_word" >&5
-printf %s "checking for $ac_word... " >&6; }
-if test ${ac_cv_prog_FOX_CONFIG+y}
-then :
-  printf %s "(cached) " >&6
-else $as_nop
-  if test -n "$FOX_CONFIG"; then
-  ac_cv_prog_FOX_CONFIG="$FOX_CONFIG" # Let the user override the test.
-else
-as_save_IFS=$IFS; IFS=$PATH_SEPARATOR
-for as_dir in $PATH
-do
-  IFS=$as_save_IFS
-  case $as_dir in #(((
-    '') as_dir=./ ;;
-    */) ;;
-    *) as_dir=$as_dir/ ;;
-  esac
-    for ac_exec_ext in '' $ac_executable_extensions; do
-  if as_fn_executable_p "$as_dir$ac_word$ac_exec_ext"; then
-    ac_cv_prog_FOX_CONFIG="$ac_prog"
-    printf "%s\n" "$as_me:${as_lineno-$LINENO}: found $as_dir$ac_word$ac_exec_ext" >&5
-    break 2
-  fi
-done
-  done
-IFS=$as_save_IFS
-
-fi
-fi
-FOX_CONFIG=$ac_cv_prog_FOX_CONFIG
-if test -n "$FOX_CONFIG"; then
-  { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: $FOX_CONFIG" >&5
-printf "%s\n" "$FOX_CONFIG" >&6; }
-else
-  { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: no" >&5
-printf "%s\n" "no" >&6; }
-fi
-
-
-  test -n "$FOX_CONFIG" && break
-done
-
-if test no"$FOX_CONFIG" = no ; then
-	as_fn_error $? "\"fox-config not found\"" "$LINENO" 5
-fi
-
-
 # Include flags for the FOX library
-FOXCFLAGS=`$FOX_CONFIG --cflags`
+FOXCFLAGS=`$PKG_CONFIG --cflags fox`
 CXXFLAGS="${CXXFLAGS} $FOXCFLAGS"
 
 
 # Check if FOX was compiled with xft support
-TEST_XFT=`$FOX_CONFIG --libs | grep Xft`
+TEST_XFT=yes
 if test "x$TEST_XFT" != "x" ; then
 
 	echo "checking whether FOX was compiled with Xft support... yes"
diff -urN --no-dereference xfe-clean/src/ArchInputDialog.cpp xfe-workdir/src/ArchInputDialog.cpp
--- xfe-clean/src/ArchInputDialog.cpp	2025-01-24 09:43:26.000000000 +0100
+++ xfe-workdir/src/ArchInputDialog.cpp	2025-04-26 23:11:14.898057065 +0200
@@ -1,5 +1,5 @@
 // Input dialog for the add to archive command
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/BrowseInputDialog.cpp xfe-workdir/src/BrowseInputDialog.cpp
--- xfe-clean/src/BrowseInputDialog.cpp	2025-01-24 09:43:26.000000000 +0100
+++ xfe-workdir/src/BrowseInputDialog.cpp	2025-04-26 23:11:14.898112139 +0200
@@ -1,5 +1,5 @@
 // Input dialog with file browse icon
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/CommandWindow.cpp xfe-workdir/src/CommandWindow.cpp
--- xfe-clean/src/CommandWindow.cpp	2025-01-23 16:06:34.000000000 +0100
+++ xfe-workdir/src/CommandWindow.cpp	2025-04-26 23:11:14.898151537 +0200
@@ -5,7 +5,7 @@
 // The object deletes itself when the close button is pressed
 // The command window can be a free-floating window or can be
 // a window which will always float over the owner window
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/DialogBox.cpp xfe-workdir/src/DialogBox.cpp
--- xfe-clean/src/DialogBox.cpp	2025-01-23 16:06:34.000000000 +0100
+++ xfe-workdir/src/DialogBox.cpp	2025-04-26 23:11:14.898187193 +0200
@@ -1,4 +1,5 @@
 // Dialog Box with additional toggle option
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/DirHistBox.cpp xfe-workdir/src/DirHistBox.cpp
--- xfe-clean/src/DirHistBox.cpp	2025-01-23 16:06:34.000000000 +0100
+++ xfe-workdir/src/DirHistBox.cpp	2025-04-26 23:11:14.898217616 +0200
@@ -1,6 +1,6 @@
 // Display a history list box and allows the user to select a string
 // This is based on FXChoiceBox
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/DirList.cpp xfe-workdir/src/DirList.cpp
--- xfe-clean/src/DirList.cpp	2025-01-24 17:41:20.000000000 +0100
+++ xfe-workdir/src/DirList.cpp	2025-04-26 23:11:14.898309833 +0200
@@ -1,7 +1,7 @@
 // Directory list. Taken from the FOX library and slightly modified.
 // The compare(), compare_nolocale() and compare_locale() functions are adapted from a patch
 // submitted by Vladimir Támara Patiño
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/DirPanel.cpp xfe-workdir/src/DirPanel.cpp
--- xfe-clean/src/DirPanel.cpp	2025-01-25 11:14:13.000000000 +0100
+++ xfe-workdir/src/DirPanel.cpp	2025-04-26 23:11:14.898510338 +0200
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/ExecuteBox.cpp xfe-workdir/src/ExecuteBox.cpp
--- xfe-clean/src/ExecuteBox.cpp	2025-01-23 16:06:34.000000000 +0100
+++ xfe-workdir/src/ExecuteBox.cpp	2025-04-26 23:11:14.898650040 +0200
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/File.cpp xfe-workdir/src/File.cpp
--- xfe-clean/src/File.cpp	2025-01-24 09:42:30.000000000 +0100
+++ xfe-workdir/src/File.cpp	2025-04-26 23:11:14.898723012 +0200
@@ -1,5 +1,5 @@
 // File management class with progress dialog
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/FileDialog.cpp xfe-workdir/src/FileDialog.cpp
--- xfe-clean/src/FileDialog.cpp	2025-03-15 11:30:05.000000000 +0100
+++ xfe-workdir/src/FileDialog.cpp	2025-04-26 23:11:14.898834256 +0200
@@ -1,5 +1,5 @@
 // File dialog. Taken from the FOX library and slightly modified.
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/FileDict.cpp xfe-workdir/src/FileDict.cpp
--- xfe-clean/src/FileDict.cpp	2025-01-24 09:46:29.000000000 +0100
+++ xfe-workdir/src/FileDict.cpp	2025-04-26 23:11:14.898903767 +0200
@@ -1,5 +1,5 @@
 // File association tables. Taken from the FOX library and slightly modified.
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/FileList.cpp xfe-workdir/src/FileList.cpp
--- xfe-clean/src/FileList.cpp	2025-03-15 11:37:58.000000000 +0100
+++ xfe-workdir/src/FileList.cpp	2025-04-26 23:11:14.899091483 +0200
@@ -2,7 +2,7 @@
 // The compare(), compare_nolocale() and compare_locale() functions are adapted from a patch
 // submitted by Vladimir Támara Patiño
 
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/FilePanel.cpp xfe-workdir/src/FilePanel.cpp
--- xfe-clean/src/FilePanel.cpp	2025-01-25 11:14:13.000000000 +0100
+++ xfe-workdir/src/FilePanel.cpp	2025-04-26 23:11:14.899430277 +0200
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/FontDialog.cpp xfe-workdir/src/FontDialog.cpp
--- xfe-clean/src/FontDialog.cpp	2025-01-25 11:10:13.000000000 +0100
+++ xfe-workdir/src/FontDialog.cpp	2025-04-26 23:11:14.899621618 +0200
@@ -1,5 +1,5 @@
 // Font dialog. Taken from the FOX library and only modified for translation purpose
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/HistInputDialog.cpp xfe-workdir/src/HistInputDialog.cpp
--- xfe-clean/src/HistInputDialog.cpp	2025-01-25 11:10:13.000000000 +0100
+++ xfe-workdir/src/HistInputDialog.cpp	2025-04-26 23:11:14.899662123 +0200
@@ -1,5 +1,5 @@
 // Input dialog with history list and an optional check box
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/IconList.cpp xfe-workdir/src/IconList.cpp
--- xfe-clean/src/IconList.cpp	2025-03-15 11:46:00.000000000 +0100
+++ xfe-workdir/src/IconList.cpp	2025-04-26 23:11:14.899790228 +0200
@@ -1,4 +1,5 @@
 // Icon list. Taken from the FOX library and slightly modified.
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/InputDialog.cpp xfe-workdir/src/InputDialog.cpp
--- xfe-clean/src/InputDialog.cpp	2025-01-23 16:06:34.000000000 +0100
+++ xfe-workdir/src/InputDialog.cpp	2025-04-26 23:11:14.899892579 +0200
@@ -1,5 +1,5 @@
 // Simple input dialog (without history)
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/MessageBox.cpp xfe-workdir/src/MessageBox.cpp
--- xfe-clean/src/MessageBox.cpp	2025-01-24 09:43:26.000000000 +0100
+++ xfe-workdir/src/MessageBox.cpp	2025-04-26 23:11:14.899925161 +0200
@@ -1,6 +1,6 @@
 // Message box. Taken from the FOX library and slightly modified for translation purpose.
 // Also added a SU button
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/OverwriteBox.cpp xfe-workdir/src/OverwriteBox.cpp
--- xfe-clean/src/OverwriteBox.cpp	2025-01-24 09:42:30.000000000 +0100
+++ xfe-workdir/src/OverwriteBox.cpp	2025-04-26 23:11:14.899959386 +0200
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/Preferences.cpp xfe-workdir/src/Preferences.cpp
--- xfe-clean/src/Preferences.cpp	2025-03-15 11:30:05.000000000 +0100
+++ xfe-workdir/src/Preferences.cpp	2025-04-26 23:11:14.900054510 +0200
@@ -1,5 +1,5 @@
 // Preferences dialog box
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/Properties.cpp xfe-workdir/src/Properties.cpp
--- xfe-clean/src/Properties.cpp	2025-01-25 11:10:13.000000000 +0100
+++ xfe-workdir/src/Properties.cpp	2025-04-26 23:11:14.900198116 +0200
@@ -1,5 +1,5 @@
 // Properties box
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/SearchPanel.cpp xfe-workdir/src/SearchPanel.cpp
--- xfe-clean/src/SearchPanel.cpp	2025-01-25 11:14:13.000000000 +0100
+++ xfe-workdir/src/SearchPanel.cpp	2025-04-26 23:11:14.900389201 +0200
@@ -1,4 +1,5 @@
 // Search panel
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/SearchWindow.cpp xfe-workdir/src/SearchWindow.cpp
--- xfe-clean/src/SearchWindow.cpp	2025-03-15 11:30:05.000000000 +0100
+++ xfe-workdir/src/SearchWindow.cpp	2025-04-26 23:11:14.900544657 +0200
@@ -1,4 +1,5 @@
 // Search files dialog and panel
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/TextWindow.cpp xfe-workdir/src/TextWindow.cpp
--- xfe-clean/src/TextWindow.cpp	2025-01-23 16:06:34.000000000 +0100
+++ xfe-workdir/src/TextWindow.cpp	2025-04-26 23:11:14.900598512 +0200
@@ -1,4 +1,5 @@
 // Text widget with a close button
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/WriteWindow.cpp xfe-workdir/src/WriteWindow.cpp
--- xfe-clean/src/WriteWindow.cpp	2025-01-24 09:44:02.000000000 +0100
+++ xfe-workdir/src/WriteWindow.cpp	2025-04-26 23:11:14.900673236 +0200
@@ -1,6 +1,6 @@
 // This is adapted from 'adie', a text editor found
 // in the FOX library and written by Jeroen van der Zijp.
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/XFileExplorer.cpp xfe-workdir/src/XFileExplorer.cpp
--- xfe-clean/src/XFileExplorer.cpp	2025-03-15 11:30:05.000000000 +0100
+++ xfe-workdir/src/XFileExplorer.cpp	2025-04-26 23:11:14.900865283 +0200
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/XFileImage.cpp xfe-workdir/src/XFileImage.cpp
--- xfe-clean/src/XFileImage.cpp	2025-03-15 11:30:05.000000000 +0100
+++ xfe-workdir/src/XFileImage.cpp	2025-04-26 23:11:14.901062007 +0200
@@ -1,6 +1,6 @@
 // This code is adapted from 'imageviewer', a demo image viewer found
 // in the FOX library and written by Jeroen van der Zijp.
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/XFilePackage.cpp xfe-workdir/src/XFilePackage.cpp
--- xfe-clean/src/XFilePackage.cpp	2025-01-25 11:14:13.000000000 +0100
+++ xfe-workdir/src/XFilePackage.cpp	2025-04-26 23:11:14.901170771 +0200
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/XFileWrite.cpp xfe-workdir/src/XFileWrite.cpp
--- xfe-clean/src/XFileWrite.cpp	2025-01-23 16:06:34.000000000 +0100
+++ xfe-workdir/src/XFileWrite.cpp	2025-04-26 23:11:14.901222737 +0200
@@ -1,6 +1,6 @@
 // This is adapted from 'adie', a demo text editor found
 // in the FOX library and written by Jeroen van der Zijp.
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/icons.cpp xfe-workdir/src/icons.cpp
--- xfe-clean/src/icons.cpp	2025-01-24 09:42:30.000000000 +0100
+++ xfe-workdir/src/icons.cpp	2025-04-26 23:11:14.901261770 +0200
@@ -1,5 +1,5 @@
 // Global icons for all applications
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/main.cpp xfe-workdir/src/main.cpp
--- xfe-clean/src/main.cpp	2025-01-24 09:42:30.000000000 +0100
+++ xfe-workdir/src/main.cpp	2025-04-26 23:11:14.901309168 +0200
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
@@ -389,6 +390,7 @@
     textdomain(PACKAGE);
 #endif
 
+    FXuint pkg_format;
 #if defined(linux)
     // For package query on Linux systems, try to guess if the default package format is deb or rpm:
     //   - if dpkg exists then the system uses deb packages
@@ -398,7 +400,6 @@
     cmd = "dpkg --version";
     FXString str = xf_getcommandoutput(cmd);
 
-    FXuint pkg_format;
     if (str.find("Debian") != -1)
     {
         pkg_format = DEB_PKG; // deb based system
@@ -417,6 +418,8 @@
             pkg_format = OTHER_PKG; // other (unsupported) package system
         }
     }
+#else
+    pkg_format = OTHER_PKG;
 #endif
 
     // Parse basic arguments
diff -urN --no-dereference xfe-clean/src/startupnotification.cpp xfe-workdir/src/startupnotification.cpp
--- xfe-clean/src/startupnotification.cpp	2025-01-24 09:35:02.000000000 +0100
+++ xfe-workdir/src/startupnotification.cpp	2025-04-26 23:11:14.901350355 +0200
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/src/xfedefs.h xfe-workdir/src/xfedefs.h
--- xfe-clean/src/xfedefs.h	2025-01-24 16:19:54.000000000 +0100
+++ xfe-workdir/src/xfedefs.h	2025-04-26 23:11:14.901382881 +0200
@@ -366,9 +366,9 @@
 #define MTAB_PATH    "/proc/mounts"
 #endif
 
+#endif
+
 // Package format
 #define DEB_PKG      0
 #define RPM_PKG      1
 #define OTHER_PKG    2
-
-#endif
diff -urN --no-dereference xfe-clean/src/xfeutils.cpp xfe-workdir/src/xfeutils.cpp
--- xfe-clean/src/xfeutils.cpp	2025-01-24 09:44:02.000000000 +0100
+++ xfe-workdir/src/xfeutils.cpp	2025-04-26 23:11:14.901445906 +0200
@@ -3,7 +3,7 @@
 // Also supports a timeout on the stat and lstat function (this is the reason
 // why some standard FOX function cannot be used and are rewritten here)
 
-
+#include <stdlib.h>
 #include "config.h"
 #include "i18n.h"
 
diff -urN --no-dereference xfe-clean/st/st.c xfe-workdir/st/st.c
--- xfe-clean/st/st.c	2025-01-20 18:23:14.000000000 +0100
+++ xfe-workdir/st/st.c	2025-04-26 23:11:14.901556112 +0200
@@ -22,17 +22,22 @@
 #include <termios.h>
 #include <unistd.h>
 #include <wchar.h>
+#include <sys/ttydefaults.h>
 
 #include "st.h"
 #include "win.h"
 
+/*
 #if   defined(__linux)
+*/
  #include <pty.h>
+/*
 #elif defined(__OpenBSD__) || defined(__NetBSD__) || defined(__APPLE__)
  #include <util.h>
 #elif defined(__FreeBSD__) || defined(__DragonFly__)
  #include <libutil.h>
 #endif
+*/
 
 /* Arbitrary sizes */
 #define UTF_INVALID   0xFFFD
