diff --git mailutils-clean/libmailutils/base/hostname.c mailutils-workdir/libmailutils/base/hostname.c
index 859184a..467c4fc 100644
--- mailutils-clean/libmailutils/base/hostname.c
+++ mailutils-workdir/libmailutils/base/hostname.c
@@ -28,6 +28,8 @@
 #include <netdb.h>
 #include <mailutils/alloc.h>
 
+#include <sys/socket.h>
+
 #ifndef MAXHOSTNAMELEN
 # define MAXHOSTNAMELEN 64
 #endif
diff --git mailutils-clean/libmailutils/string/cstrcasestr.c mailutils-workdir/libmailutils/string/cstrcasestr.c
index f3f1a54..dac7be2 100644
--- mailutils-clean/libmailutils/string/cstrcasestr.c
+++ mailutils-workdir/libmailutils/string/cstrcasestr.c
@@ -19,6 +19,8 @@
 # include <config.h>
 #endif
 
+#include <stddef.h>
+
 #include <unistd.h>
 #include <mailutils/cctype.h>
 
diff --git mailutils-clean/mh/mh_init.c mailutils-workdir/mh/mh_init.c
index 1db971a..1ffc94a 100644
--- mailutils-clean/mh/mh_init.c
+++ mailutils-workdir/mh/mh_init.c
@@ -30,6 +30,8 @@
 #include <fnmatch.h>
 #include <sys/ioctl.h>
 
+#include <termios.h>
+
 void
 mh_init (void)
 {
