diff -urN --no-dereference mailutils-clean/libmailutils/base/hostname.c mailutils-workdir/libmailutils/base/hostname.c
--- mailutils-clean/libmailutils/base/hostname.c	2025-01-02 14:36:29.000000000 +0100
+++ mailutils-workdir/libmailutils/base/hostname.c	2025-01-02 14:36:29.000000000 +0100
@@ -28,6 +28,8 @@
 #include <netdb.h>
 #include <mailutils/alloc.h>
 
+#include <sys/socket.h>
+
 #ifndef MAXHOSTNAMELEN
 # define MAXHOSTNAMELEN 64
 #endif
diff -urN --no-dereference mailutils-clean/libmailutils/string/cstrcasestr.c mailutils-workdir/libmailutils/string/cstrcasestr.c
--- mailutils-clean/libmailutils/string/cstrcasestr.c	2025-01-02 14:36:33.000000000 +0100
+++ mailutils-workdir/libmailutils/string/cstrcasestr.c	2025-01-02 14:36:33.000000000 +0100
@@ -19,6 +19,8 @@
 # include <config.h>
 #endif
 
+#include <stddef.h>
+
 #include <unistd.h>
 #include <mailutils/cctype.h>
 
diff -urN --no-dereference mailutils-clean/mh/mh_init.c mailutils-workdir/mh/mh_init.c
--- mailutils-clean/mh/mh_init.c	2025-01-02 14:36:42.000000000 +0100
+++ mailutils-workdir/mh/mh_init.c	2025-01-02 14:36:42.000000000 +0100
@@ -30,6 +30,8 @@
 #include <fnmatch.h>
 #include <sys/ioctl.h>
 
+#include <termios.h>
+
 void
 mh_init (void)
 {
