diff --git python-clean/Lib/site.py python-workdir/Lib/site.py
index aedf363..b56b236 100644
--- python-clean/Lib/site.py
+++ python-workdir/Lib/site.py
@@ -409,6 +409,10 @@ def getsitepackages(prefixes=None):
                                     f"{implementation}{ver[0]}.{ver[1]}{abi_thread}",
                                     "site-packages")
                 sitepackages.append(path)
+                path = os.path.join(prefix, libdir,
+                                    f"{implementation}{ver[0]}.{ver[1]}{abi_thread}",
+                                    "dist-packages")
+                sitepackages.append(path)
         else:
             sitepackages.append(prefix)
             sitepackages.append(os.path.join(prefix, "Lib", "site-packages"))
diff --git python-clean/Modules/timemodule.c python-workdir/Modules/timemodule.c
index 49626d3..8fb0403 100644
--- python-clean/Modules/timemodule.c
+++ python-workdir/Modules/timemodule.c
@@ -185,7 +185,7 @@ py_clock(time_module_state *state, PyTime_t *tp, _Py_clock_info_t *info)
 static int
 time_clockid_converter(PyObject *obj, clockid_t *p)
 {
-#ifdef _AIX
+#if defined(_AIX) || defined(__ironclad__)
     long long clk_id = PyLong_AsLongLong(obj);
 #elif defined(__DragonFly__)
     long clk_id = PyLong_AsLong(obj);
diff --git python-clean/Objects/mimalloc/prim/unix/prim.c python-workdir/Objects/mimalloc/prim/unix/prim.c
index 4d3fb65..55c7139 100644
--- python-clean/Objects/mimalloc/prim/unix/prim.c
+++ python-workdir/Objects/mimalloc/prim/unix/prim.c
@@ -29,6 +29,10 @@ terms of the MIT license. A copy of the license can be found in the file
 #include <unistd.h>    // sysconf
 #include <fcntl.h>     // open, close, read, access
 
+#if defined(__ironclad__)
+  #define MADV_DONTNEED 1
+#endif
+
 #if defined(__linux__)
   #include <features.h>
   #include <fcntl.h>
@@ -156,6 +160,9 @@ int _mi_prim_free(void* addr, size_t size ) {
 static int unix_madvise(void* addr, size_t size, int advice) {
   #if defined(__sun)
   return madvise((caddr_t)addr, size, advice);  // Solaris needs cast (issue #520)
+  #elif defined(__ironclad__)
+  (void)addr; (void)size; (void)advice;
+  return 0;
   #else
   return madvise(addr, size, advice);
   #endif
diff --git python-clean/configure python-workdir/configure
index 901a26c..b14a858 100755
--- python-clean/configure
+++ python-workdir/configure
@@ -4067,6 +4067,9 @@ then
 	*-*-wasi*)
 	    ac_sys_system=WASI
 	    ;;
+	*-*-mlibc*)
+	    ac_sys_system=Mlibc
+	    ;;
 	*)
 		# for now, limit cross builds to known configurations
 		MACHDEP="unknown"
@@ -4491,7 +4494,7 @@ fi
 
 if test "$cross_compiling" = yes; then
 	case "$host" in
-	*-*-linux*)
+	*-*-linux*|*-*-mlibc*)
 		case "$host_cpu" in
 		arm*)
 			_host_ident=arm
@@ -7587,7 +7590,7 @@ printf "%s\n" "#define Py_ENABLE_SHARED 1" >>confdefs.h
         PY3LIBRARY=libpython3.so
       fi
       ;;
-    Linux*|GNU*|NetBSD*|FreeBSD*|DragonFly*|OpenBSD*|VxWorks*)
+    Linux*|GNU*|NetBSD*|FreeBSD*|DragonFly*|OpenBSD*|VxWorks*|Mlibc*)
       LDLIBRARY='libpython$(LDVERSION).so'
       BLDLIBRARY='-L. -lpython$(LDVERSION)'
       RUNSHARED=LD_LIBRARY_PATH=`pwd`${LD_LIBRARY_PATH:+:${LD_LIBRARY_PATH}}
@@ -12983,7 +12986,7 @@ then
 	Emscripten*|WASI*)
 		LDSHARED='$(CC) -shared'
 		LDCXXSHARED='$(CXX) -shared';;
-	Linux*|GNU*|QNX*|VxWorks*|Haiku*)
+	Linux*|GNU*|QNX*|VxWorks*|Haiku*|Mlibc*)
 		LDSHARED='$(CC) -shared'
 		LDCXXSHARED='$(CXX) -shared';;
 	FreeBSD*)
@@ -13068,7 +13071,7 @@ then
 		 then CCSHARED="-fPIC";
 		 else CCSHARED="+z";
 		 fi;;
-	Linux*|GNU*) CCSHARED="-fPIC";;
+	Linux*|GNU*|Mlibc*) CCSHARED="-fPIC";;
 	Emscripten*|WASI*)
 		if test "x$enable_wasm_dynamic_linking" = xyes
 then :
@@ -13106,7 +13109,7 @@ then
 	    LINKFORSHARED="-Wl,-E -Wl,+s";;
 #	    LINKFORSHARED="-Wl,-E -Wl,+s -Wl,+b\$(BINLIBDEST)/lib-dynload";;
 	Linux-android*) LINKFORSHARED="-pie -Xlinker -export-dynamic";;
-	Linux*|GNU*) LINKFORSHARED="-Xlinker -export-dynamic";;
+	Linux*|GNU*|Mlibc*) LINKFORSHARED="-Xlinker -export-dynamic";;
 	# -u libsys_s pulls in all symbols in libsys
 	Darwin/*|iOS/*)
 		LINKFORSHARED="$extra_undefs -framework CoreFoundation"
