diff --git git-clean/config.mak.uname git-workdir/config.mak.uname
index 1691c6a..a2c5b90 100644
--- git-clean/config.mak.uname
+++ git-workdir/config.mak.uname
@@ -4,11 +4,11 @@
 # Microsoft's Safe Exception Handling in libraries (such as zlib).
 # Typically required for VS2013+/32-bit compilation on Vista+ versions.
 
-uname_S := $(shell sh -c 'uname -s 2>/dev/null || echo not')
-uname_M := $(shell sh -c 'uname -m 2>/dev/null || echo not')
-uname_O := $(shell sh -c 'uname -o 2>/dev/null || echo not')
-uname_R := $(shell sh -c 'uname -r 2>/dev/null || echo not')
-uname_V := $(shell sh -c 'uname -v 2>/dev/null || echo not')
+uname_S := Ironclad
+uname_M := x86_64
+uname_O := Ironclad
+uname_R := 0.0.0
+uname_V := 0.0.0
 
 ifneq ($(findstring MINGW,$(uname_S)),)
 	uname_S := MINGW
