diff --git atril-clean/backend/comics/meson.build atril-workdir/backend/comics/meson.build
index a990388..9733285 100644
--- atril-clean/backend/comics/meson.build
+++ atril-workdir/backend/comics/meson.build
@@ -34,16 +34,3 @@ shared_module(
 
 
 
-test_name = 'test-ev-archive'
-
-test_sources = files(
-  'ev-archive.c',
-  'test-ev-archive.c',
-)
-
-executable(
-  test_name,
-  test_sources,
-  include_directories: include_dirs,
-  dependencies: comics_deps,
-)
diff --git atril-clean/meson.build atril-workdir/meson.build
index 8a02b91..5988610 100644
--- atril-clean/meson.build
+++ atril-workdir/meson.build
@@ -348,7 +348,6 @@ summary = [
   'desktopdir = @0@'.format(desktopdir),
   'schema_dir = @0@'.format(schema_dir),
   'ENABLE_DBUS = @0@'.format(get_option('enable_dbus')),
-  'MathJax directory = @0@'.format(mathjax_directory),
   '',
   'Backends enabled:',
 ]
