diff --git atril-clean/backend/comics/meson.build atril-workdir/backend/comics/meson.build
index a990388..9733285 100644
--- atril-clean/backend/comics/meson.build
+++ atril-workdir/backend/comics/meson.build
@@ -34,16 +34,3 @@ shared_module(
 
 
 
-test_name = 'test-ev-archive'
-
-test_sources = files(
-  'ev-archive.c',
-  'test-ev-archive.c',
-)
-
-executable(
-  test_name,
-  test_sources,
-  include_directories: include_dirs,
-  dependencies: comics_deps,
-)
diff --git atril-clean/meson.build atril-workdir/meson.build
index 779e08a..fce157b 100644
--- atril-clean/meson.build
+++ atril-workdir/meson.build
@@ -345,7 +345,6 @@ summary = [
   'desktopdir = @0@'.format(desktopdir),
   'schema_dir = @0@'.format(schema_dir),
   'ENABLE_DBUS = @0@'.format(get_option('enable_dbus')),
-  'MathJax directory = @0@'.format(mathjax_directory),
   '',
   'Backends enabled:',
 ]
diff --git atril-clean/properties/meson.build atril-workdir/properties/meson.build
index 247abe6..49ccf78 100644
--- atril-clean/properties/meson.build
+++ atril-workdir/properties/meson.build
@@ -28,7 +28,7 @@ if gtk_x11.found() and caja_extension.found()
         include_directories: include_dirs,
         dependencies: [gtk_x11, caja_extension, libdocument_dep, libevproperties_dep],
         install: true,
-        install_dir: caja_extension.get_variable(pkgconfig: 'extensiondir')
+        install_dir: join_paths(prefix, libdir, 'caja', 'extensions-2.0')
     )
 
     caja_ext_conf = configuration_data()
