diff --git shadow-clean/lib/chowntty.c shadow-workdir/lib/chowntty.c
index 1f99d2d..d2d7808 100644
--- shadow-clean/lib/chowntty.c
+++ shadow-workdir/lib/chowntty.c
@@ -50,7 +50,7 @@ void chown_tty (const struct passwd *info)
 	 * the group as determined above.
 	 */
 
-	if (   (fchown (STDIN_FILENO, info->pw_uid, gid) != 0)
+	/*if (   (fchown (STDIN_FILENO, info->pw_uid, gid) != 0)
 	    || (fchmod (STDIN_FILENO, getdef_num ("TTYPERM", 0600)) != 0)) {
 		int err = errno;
 		FILE *shadow_logfd = log_get_logfd();
@@ -65,7 +65,7 @@ void chown_tty (const struct passwd *info)
 			closelog ();
 			exit (EXIT_FAILURE);
 		}
-	}
+	}*/
 #ifdef __linux__
 	/*
 	 * Please don't add code to chown /dev/vcs* to the user logging in -
diff --git shadow-clean/lib/commonio.c shadow-workdir/lib/commonio.c
index 2c6a1a2..d5957be 100644
--- shadow-clean/lib/commonio.c
+++ shadow-workdir/lib/commonio.c
@@ -387,38 +387,6 @@ int commonio_lock (struct commonio_db *db)
 {
 	int i;
 
-#ifdef HAVE_LCKPWDF
-	/*
-	 * Only if the system libc has a real lckpwdf() - the one from
-	 * lockpw.c calls us and would cause infinite recursion!
-	 * It is also not used with the prefix option.
-	 */
-	if (!db->setname) {
-		/*
-		 * Call lckpwdf() on the first lock.
-		 * If it succeeds, call *_lock() only once
-		 * (no retries, it should always succeed).
-		 */
-		if (0 == lock_count) {
-			if (lckpwdf () == -1) {
-				if (geteuid () != 0) {
-					(void) fprintf (shadow_logfd,
-					                "%s: Permission denied.\n",
-					                shadow_progname);
-				}
-				return 0;	/* failure */
-			}
-		}
-
-		if (commonio_lock_nowait (db, true) != 0) {
-			return 1;	/* success */
-		}
-
-		ulckpwdf ();
-		return 0;		/* failure */
-	}
-#endif				/* !HAVE_LCKPWDF */
-
 	/*
 	 * lckpwdf() not used - do it the old way.
 	 */
