diff --git gtk+-3-clean/gtk/a11y/gtkaccessibility.c gtk+-3-workdir/gtk/a11y/gtkaccessibility.c
index 7f0e520..3bf670c 100644
--- gtk+-3-clean/gtk/a11y/gtkaccessibility.c
+++ gtk+-3-workdir/gtk/a11y/gtkaccessibility.c
@@ -974,6 +974,10 @@ do_window_event_initialization (void)
 void
 _gtk_accessibility_init (void)
 {
+// We don't care about a11y at this point in time and dbus isn't working, so let's just not start that.
+#ifdef __vinix__
+  return;
+#else
   if (initialized)
     return;
 
@@ -993,4 +997,5 @@ _gtk_accessibility_init (void)
 #endif
 
   atk_misc_instance = g_object_new (GTK_TYPE_MISC_IMPL, NULL);
+#endif
 }
