diff --git which-clean/getopt.h which-workdir/getopt.h
index f080053..6d2520f 100644
--- which-clean/getopt.h
+++ which-workdir/getopt.h
@@ -96,7 +96,7 @@ struct option {
 #define optional_argument 2
 
 #if defined(__STDC__) && __STDC__
-#ifdef __GNU_LIBRARY__
+#if defined(__GNU_LIBRARY__) || defined(__mlibc__)
 /* Many other libraries have conflicting prototypes for getopt, with
    differences in the consts, in stdlib.h.  To avoid compilation
    errors, only prototype getopt for the GNU C library.  */
