diff --git ncurses-clean/c++/Makefile.in ncurses-workdir/c++/Makefile.in
index 502b3e7..16e7fa4 100644
--- ncurses-clean/c++/Makefile.in
+++ ncurses-workdir/c++/Makefile.in
@@ -97,6 +97,7 @@ CTAGS		= @CTAGS@
 ETAGS		= @ETAGS@
 
 CC		= ${CXX}
+CFLAGS		= ${CXXFLAGS}
 CCFLAGS		= $(CPPFLAGS) $(CXXFLAGS)
 
 CFLAGS_LIBTOOL	= $(CCFLAGS)
@@ -147,7 +148,7 @@ LINK_SHARED	= $(LINK_FLAGS)
 # flags for test-programs
 TEST_LIBS	= @TEST_LIBS@
 TEST_ARGS	= @LDFLAGS_STATIC@ @TEST_ARGS@ @LDFLAGS_SHARED@
-TEST_LDFLAGS	= $(TEST_ARGS) @LDFLAGS@ \
+TEST_LDFLAGS	= @LDFLAGS@ $(TEST_ARGS) \
 	@LD_MODEL@ $(TEST_LIBS) @LIBS@ @LOCAL_LDFLAGS@ $(CXXLIBS)
 
 LDFLAGS_LIBTOOL	= $(TEST_LDFLAGS) $(CFLAGS_LIBTOOL)
@@ -156,10 +157,10 @@ LDFLAGS_DEBUG	= $(TEST_LDFLAGS) $(CFLAGS_DEBUG)
 LDFLAGS_PROFILE	= $(TEST_LDFLAGS) $(CFLAGS_PROFILE)
 LDFLAGS_SHARED	= $(TEST_LDFLAGS) $(CFLAGS_SHARED) @LD_SHARED_OPTS@
 
-LDFLAGS_DEFAULT	= $(LINK_@DFT_UPR_MODEL@) $(LDFLAGS_@DFT_UPR_MODEL@)
+LDFLAGS_DEFAULT	= $(LDFLAGS_@DFT_UPR_MODEL@) $(LINK_@DFT_UPR_MODEL@)
 
 # flags for library built by this makefile
-LDFLAGS		= $(TEST_ARGS) @LDFLAGS@ \
+LDFLAGS		= @LDFLAGS@ $(TEST_ARGS) \
 	@LIBS@ $(CXXLIBS)
 
 AUTO_SRC	= \
diff --git ncurses-clean/configure ncurses-workdir/configure
index 48ad381..6be5c4b 100755
--- ncurses-clean/configure
+++ ncurses-workdir/configure
@@ -3827,7 +3827,7 @@ echo $ECHO_N "checking if $CXX works... $ECHO_C" >&6
 
 		save_CPPFLAGS="$CPPFLAGS"
 		eval cf_includedir=${includedir}
-		CPPFLAGS="$CPPFLAGS -I${cf_includedir}"
+		CPPFLAGS="$CPPFLAGS"
 
 cat >"conftest.$ac_ext" <<_ACEOF
 #line 3833 "configure"
@@ -6764,7 +6764,7 @@ echo $ECHO_N "checking for an rpath option... $ECHO_C" >&6
 			LD_RPATH_OPT="-rpath "
 		fi
 		;;
-	(linux*|gnu*|k*bsd*-gnu|freebsd*)
+	(linux*|gnu*|k*bsd*-gnu|freebsd*|ironclad*)
 		LD_RPATH_OPT="-Wl,-rpath,"
 		;;
 	(openbsd[2-9].*|mirbsd*)
@@ -7123,7 +7123,7 @@ echo "${ECHO_T}$cf_cv_ldflags_search_paths_first" >&6
 		fi
 		cf_cv_rm_so_locs=yes
 		;;
-	(linux*|gnu*|k*bsd*-gnu)
+	(linux*|gnu*|k*bsd*-gnu|ironclad*)
 		if test "$DFT_LWR_MODEL" = "shared" && test -n "$LD_RPATH_OPT" ; then
 			LOCAL_LDFLAGS="${LD_RPATH_OPT}\$(LOCAL_LIBDIR)"
 			LOCAL_LDFLAGS2="$LOCAL_LDFLAGS"
@@ -7247,6 +7247,10 @@ CF_EOF
 		LINK_PROGS="$SHELL ${rel_builddir}/mk_prog.sh"
 		LINK_TESTS="$SHELL ${rel_builddir}/mk_prog.sh"
 		;;
+	(ironclad*)
+		CC_SHARED_OPTS='-fPIC'
+		MK_SHARED_LIB='${CC} -shared -o $@'
+		;;
 	(mingw*)
 		cf_cv_shlib_version=mingw
 		cf_cv_shlib_version_infix=mingw
@@ -9807,7 +9811,7 @@ echo "${ECHO_T}$with_xterm_new" >&6
 WHICH_XTERM=$with_xterm_new
 
 case $host_os in
-(linux*gnu*|uclinux*|gnu*|mint*|k*bsd*-gnu|cygwin|msys|mingw*|linux*uclibc|linux*musl|openbsd*|darwin*)
+(linux*gnu*|uclinux*|gnu*|mint*|k*bsd*-gnu|cygwin|msys|mingw*|linux*uclibc|linux*musl|openbsd*|darwin*|ironclad*)
 	want_xterm_kbs=DEL
 	;;
 (*)
@@ -10785,7 +10789,7 @@ case "$host_os" in
 	cf_xopen_source="-D_SGI_SOURCE"
 	cf_XOPEN_SOURCE=
 	;;
-(linux*gnu*|uclinux*|gnu*|mint*|k*bsd*-gnu|cygwin|msys|mingw*|linux*uclibc)
+(linux*gnu*|uclinux*|gnu*|mint*|k*bsd*-gnu|cygwin|msys|mingw*|linux*uclibc|ironclad*)
 
 cf_gnu_xopen_source=$cf_XOPEN_SOURCE
 
diff --git ncurses-clean/ncurses/tty/tty_update.c ncurses-workdir/ncurses/tty/tty_update.c
index 93dd597..75464a9 100644
--- ncurses-clean/ncurses/tty/tty_update.c
+++ ncurses-workdir/ncurses/tty/tty_update.c
@@ -84,6 +84,7 @@
 #endif
 
 #include <ctype.h>
+#include <wctype.h>
 
 #if USE_WIDEC_SUPPORT
 #if HAVE_WCTYPE_H
diff --git ncurses-clean/progs/Makefile.in ncurses-workdir/progs/Makefile.in
index 0e1fc01..f3245f0 100644
--- ncurses-clean/progs/Makefile.in
+++ ncurses-workdir/progs/Makefile.in
@@ -80,7 +80,7 @@ LIBTOOL_INSTALL	= @LIB_INSTALL@
 LIBTOOL_UNINSTALL = @LIB_UNINSTALL@
 
 INSTALL		= @INSTALL@
-INSTALL_PROG	= @INSTALL_PROGRAM@ @INSTALL_OPT_S@
+INSTALL_PROG	= @INSTALL_PROGRAM@
 transform	= @program_transform_name@
 
 AWK		= @AWK@
