diff --git tr7-clean/Makefile tr7-workdir/Makefile
index d0308a7..13a76c7 100644
--- tr7-clean/Makefile
+++ tr7-workdir/Makefile
@@ -34,7 +34,7 @@ DFLAGS = -g
 FFLAGS =
 EFLAGS =
 MFLAGS = '-DVERSION="${VERSION}"'
-LDFLAGS = -Wl,--gc-sections,-z,relro,-z,now
+override LDFLAGS := -Wl,--gc-sections,-z,relro,-z,now $(LDFLAGS)
 # Detect darwin linker, which is not compatible with gnu linker
 ifneq ($(findstring PROJECT,$(shell $(LD) -v 2>&1)),)
 	LDFLAGS = -Wl,-dead_strip,-pie
@@ -53,10 +53,10 @@ BINDIR = $(PREFIX)/bin
 LIBDIR = $(PREFIX)/lib
 INCDIR = $(PREFIX)/include
 DATADIR = $(PREFIX)/share
-TR7LIBPATH = $(PREFIX)/share/tr7/tr7libs:$(PWD)/tr7libs
+TR7LIBPATH = $(PREFIX)/share/tr7/tr7libs
 
 # default targets
-ALL = $(LIBSTA) $(TR7I) tags
+ALL = $(LIBSO) $(TR7I)
 
 # uncomment below to compile shared library
 #ALL += $(LIBSO)
@@ -69,7 +69,7 @@ ALL = $(LIBSTA) $(TR7I) tags
 #FFLAGS += -DDUMP_LAMBDAS=1
 
 # set CFLAGS now
-CFLAGS := $(CFLAGS) $(SFLAGS) $(GFLAGS) $(WFLAGS) $(OFLAGS) $(DFLAGS) $(FFLAGS) $(EFLAGS) $(MFLAGS)
+override CFLAGS := $(SFLAGS) $(GFLAGS) $(WFLAGS) $(OFLAGS) $(DFLAGS) $(FFLAGS) $(EFLAGS) $(MFLAGS) $(CFLAGS)
 
 # rules
 all: $(ALL)
@@ -81,10 +81,10 @@ libtr7.so: $(LIBSO)
 	$(LN) -sf $(LIBSO) libtr7.so
 
 $(LIBSO): tr7.dyn.o
-	$(CC) -shared -o $(LIBSO) tr7.dyn.o $(SYSLIBS) -Wl,-soname,$(LIBSO)
+	$(CC) $(CFLAGS) -shared $(LDFLAGS) -o $(LIBSO) tr7.dyn.o $(SYSLIBS) -Wl,-soname,$(LIBSO)
 
 $(TR7I): examples/tr7i.c tr7.c tr7.h
-	$(CC) $(CFLAGS) -I. -o $(TR7I) examples/tr7i.c tr7.c $(LDFLAGS) $(SYSLIBS) -DDEFLIBPATH='"$(TR7LIBPATH)"'
+	$(CC) $(CFLAGS) $(LDFLAGS) -I. -o $(TR7I) examples/tr7i.c tr7.c $(SYSLIBS) -DDEFLIBPATH='"$(TR7LIBPATH)"'
 
 tr7.sta.o: tr7.c tr7.h
 	$(CC) $(CFLAGS) -c -o tr7.sta.o  tr7.c -ffunction-sections
@@ -119,7 +119,7 @@ updtref:
 # C demo examples
 
 demo-c: examples/demo-c.c $(LIBSTA)
-	$(CC) $(CFLAGS) -I. -o demo-c examples/demo-c.c $(LIBSTA) $(LDFLAGS) $(SYSLIBS)
+	$(CC) $(CFLAGS) $(LDFLAGS) -I. -o demo-c examples/demo-c.c $(LIBSTA) $(SYSLIBS)
 
 #---------------------
 # install
