diff --git xxhash-clean/Makefile xxhash-workdir/Makefile
index 5b7b8ff..4bf6446 100644
--- xxhash-clean/Makefile
+++ xxhash-workdir/Makefile
@@ -539,7 +539,7 @@ lint-unicode:
 # =========================================================
 # make install is validated only for the following targets
 # =========================================================
-ifneq (,$(filter Linux Darwin GNU/kFreeBSD GNU Haiku OpenBSD FreeBSD NetBSD DragonFly SunOS CYGWIN% , $(UNAME)))
+ifneq (,$(filter Linux Darwin GNU/kFreeBSD GNU Haiku Ironclad OpenBSD FreeBSD NetBSD DragonFly SunOS CYGWIN% , $(UNAME)))

 DESTDIR     ?=
 # directory variables: GNU conventions prefer lowercase
