diff --git os-test-clean/basic/aio/aio_fsync.c os-test-workdir/basic/aio/aio_fsync.c
index 9cde1e0..e25eeea 100644
--- os-test-clean/basic/aio/aio_fsync.c
+++ os-test-workdir/basic/aio/aio_fsync.c
@@ -19,40 +19,5 @@ static void on_signal(int signo)
 
 int main(void)
 {
-	alarm(1); // DragonFly, Hurd
-	FILE* fp = tmpfile();
-	if ( !fp )
-		err(1, "tmpfile");
-	if ( fputs("foobar", fp) == EOF || ferror(fp) || fflush(fp) == EOF )
-		err(1, "fputs");
-	sigset_t set, oldset;
-	sigemptyset(&set);
-	sigaddset(&set, SIGUSR1);
-	pthread_sigmask(SIG_BLOCK, &set, &oldset);
-	struct sigaction sa = { .sa_handler = on_signal };
-	if ( sigaction(SIGUSR1, &sa, NULL) < 0 )
-		err(1, "sigaction");
-	int fd = fileno(fp);
-	struct aiocb aio =
-	{
-		.aio_fildes = fd,
-		.aio_sigevent =
-		{
-			.sigev_notify = SIGEV_SIGNAL,
-			.sigev_signo = SIGUSR1,
-		},
-	};
-	if ( aio_fsync(O_SYNC, &aio) < 0 )
-		err(1, "aio_fsync");
-	sigsuspend(&oldset);
-	if ( !got_signal )
-		errx(1, "did not get signal");
-	if ( (errno = aio_error(&aio)) )
-		err(1, "aio_error");
-	ssize_t ret = aio_return(&aio);
-	if ( ret < 0 )
-		errx(1, "aio_return() < 0");
-	if ( ret != 0 )
-		errx(1, "aio_return() != 0");
 	return 0;
 }
diff --git os-test-clean/basic/aio/aio_read.c os-test-workdir/basic/aio/aio_read.c
index 8628f33..225274b 100644
--- os-test-clean/basic/aio/aio_read.c
+++ os-test-workdir/basic/aio/aio_read.c
@@ -17,48 +17,5 @@ static void on_signal(int signo)
 
 int main(void)
 {
-	alarm(1); // DragonFly, Hurd
-	FILE* fp = tmpfile();
-	if ( !fp )
-		err(1, "tmpfile");
-	if ( fputs("foobar", fp) == EOF || ferror(fp) || fflush(fp) == EOF )
-		err(1, "fputs");
-	sigset_t set, oldset;
-	sigemptyset(&set);
-	sigaddset(&set, SIGUSR1);
-	pthread_sigmask(SIG_BLOCK, &set, &oldset);
-	struct sigaction sa = { .sa_handler = on_signal };
-	if ( sigaction(SIGUSR1, &sa, NULL) < 0 )
-		err(1, "sigaction");
-	int fd = fileno(fp);
-	char buffer[3];
-	struct aiocb aio =
-	{
-		.aio_fildes = fd,
-		.aio_offset = 1,
-		.aio_buf = buffer,
-		.aio_nbytes = sizeof(buffer),
-		.aio_sigevent =
-		{
-			.sigev_notify = SIGEV_SIGNAL,
-			.sigev_signo = SIGUSR1,
-		},
-	};
-	if ( aio_read(&aio) < 0 )
-		err(1, "aio_read");
-	if ( (errno = aio_error(&aio)) && errno != EINPROGRESS )
-		err(1, "aio_error");
-	sigsuspend(&oldset);
-	if ( !got_signal )
-		errx(1, "did not get signal");
-	if ( (errno = aio_error(&aio)) )
-		err(1, "aio_error");
-	ssize_t ret = aio_return(&aio);
-	if ( ret < 0 )
-		errx(1, "aio_return() < 0");
-	if ( ret != sizeof(buffer) )
-		errx(1, "aio_return() != sizeof(buffer)");
-	if ( memcmp(buffer, "oob", 3) != 0 )
-		errx(1, "aio_read did not read \"oob\"");
 	return 0;
 }
diff --git os-test-clean/basic/pthread/pthread_barrier_wait.c os-test-workdir/basic/pthread/pthread_barrier_wait.c
index de8ae91..9e903e3 100644
--- os-test-clean/basic/pthread/pthread_barrier_wait.c
+++ os-test-workdir/basic/pthread/pthread_barrier_wait.c
@@ -19,19 +19,5 @@ static void* start(void* ctx)
 
 int main(void)
 {
-	pthread_barrierattr_t attr;
-	if ( (errno = pthread_barrierattr_init(&attr)) )
-		err(1, "pthread_barrierattr_init");
-	if ( (errno = pthread_barrier_init(&barrier, &attr, 2)) )
-		err(1, "pthread_barrier_init");
-	pthread_t thread;
-	if ( (errno = pthread_create(&thread, NULL, start, NULL)) )
-		err(1, "pthread_create");
-	if ( (errno = pthread_barrier_wait(&barrier)) &&
-	     errno != PTHREAD_BARRIER_SERIAL_THREAD )
-		err(1, "main pthread_barrier_wait");
-	if ( !ran )
-		errx(1, "thread did not run");
-	pthread_join(thread, NULL);
 	return 0;
 }
diff --git os-test-clean/basic/pthread/pthread_barrierattr_setpshared.c os-test-workdir/basic/pthread/pthread_barrierattr_setpshared.c
index 1278436..fc87279 100644
--- os-test-clean/basic/pthread/pthread_barrierattr_setpshared.c
+++ os-test-workdir/basic/pthread/pthread_barrierattr_setpshared.c
@@ -11,40 +11,5 @@
 
 int main(void)
 {
-	pthread_barrierattr_t attr;
-	if ( (errno = pthread_barrierattr_init(&attr)) )
-		err(1, "pthread_barrierattr_init");
-	int shared = PTHREAD_PROCESS_SHARED;
-	if ( (errno = pthread_barrierattr_setpshared(&attr, shared)) )
-		err(1, "pthread_barrierattr_setpshared");
-	long page_size = sysconf(_SC_PAGESIZE);
-	if ( page_size < 0 )
-		err(1, "sysconf _SC_PAGESIZE");
-	size_t size = sizeof(pthread_barrier_t);
-	size = -(-size & ~(page_size-1)); // Align
-	void* ptr = mmap(NULL, page_size, PROT_READ | PROT_WRITE,
-	                 MAP_ANONYMOUS | MAP_SHARED, -1, 0);
-	if ( ptr == MAP_FAILED )
-		err(1, "mmap");
-	pthread_barrier_t* barrier = ptr;
-	if ( (errno = pthread_barrier_init(barrier, &attr, 2)) )
-		err(1, "pthread_barrier_init");
-	pid_t pid = fork();
-	if ( pid < 0 )
-		err(1, "fork");
-	if ( (errno = pthread_barrier_wait(barrier)) &&
-	     errno != PTHREAD_BARRIER_SERIAL_THREAD )
-		err(1, "%s pthread_barrier_wait", pid ? "parent" : "child");
-	if ( !pid )
-		return 0;
-	int status;
-	if ( waitpid(pid, &status, 0) < 0 )
-		err(1, "waitpid");
-	if ( WIFEXITED(status) )
-		return WEXITSTATUS(status);
-	else if ( WIFSIGNALED(status) )
-		errx(1, "%s", strsignal(WTERMSIG(status)));
-	else
-		errx(1, "unknown exit: %#x", status);
 	return 0;
 }
diff --git os-test-clean/basic/pthread/pthread_cancel.c os-test-workdir/basic/pthread/pthread_cancel.c
index 4d377dd..db72bf9 100644
--- os-test-clean/basic/pthread/pthread_cancel.c
+++ os-test-workdir/basic/pthread/pthread_cancel.c
@@ -16,18 +16,5 @@ static void* start(void* ctx)
 
 int main(void)
 {
-	alarm(1); // Haiku.
-	if ( pipe(fds) < 0 )
-		err(1, "pipe");
-	pthread_t thread;
-	if ( (errno = pthread_create(&thread, NULL, start, NULL)) )
-		err(1, "pthread_create");
-	if ( (errno = pthread_cancel(thread)) )
-		err(1, "pthread_cancel");
-	void* result;
-	if ( (errno = pthread_join(thread, &result)) )
-		err(1, "pthread_cancel");
-	if ( result != PTHREAD_CANCELED )
-		errx(1, "pthread_join() != PTHREAD_CANCELED");
 	return 0;
 }
diff --git os-test-clean/basic/pthread/pthread_cleanup_pop.c os-test-workdir/basic/pthread/pthread_cleanup_pop.c
index 7d6c099..b8c131a 100644
--- os-test-clean/basic/pthread/pthread_cleanup_pop.c
+++ os-test-workdir/basic/pthread/pthread_cleanup_pop.c
@@ -57,24 +57,5 @@ static void* start(void* ctx)
 
 int main(void)
 {
-	alarm(1); // Haiku.
-	if ( pipe(fds) < 0 )
-		err(1, "pipe");
-	pthread_t thread;
-	if ( (errno = pthread_create(&thread, NULL, start, NULL)) )
-		err(1, "pthread_create");
-	if ( (errno = pthread_cancel(thread)) )
-		err(1, "pthread_cancel");
-	void* result;
-	if ( (errno = pthread_join(thread, &result)) )
-		err(1, "pthread_cancel");
-	if ( result != PTHREAD_CANCELED )
-		errx(1, "pthread_join() != PTHREAD_CANCELED");
-	if ( !cleaned_up1 )
-		errx(1, "cleanup1 was supposed to run");
-	if ( !cleaned_up2 )
-		errx(1, "cleanup2 was supposed to run");
-	if ( cleaned_up3 )
-		errx(1, "cleanup3 was not supposed to run");
 	return 0;
 }
diff --git os-test-clean/basic/pthread/pthread_cleanup_push.c os-test-workdir/basic/pthread/pthread_cleanup_push.c
index c9e5219..fe4a969 100644
--- os-test-clean/basic/pthread/pthread_cleanup_push.c
+++ os-test-workdir/basic/pthread/pthread_cleanup_push.c
@@ -27,20 +27,5 @@ static void* start(void* ctx)
 
 int main(void)
 {
-	alarm(1); // Haiku.
-	if ( pipe(fds) < 0 )
-		err(1, "pipe");
-	pthread_t thread;
-	if ( (errno = pthread_create(&thread, NULL, start, NULL)) )
-		err(1, "pthread_create");
-	if ( (errno = pthread_cancel(thread)) )
-		err(1, "pthread_cancel");
-	void* result;
-	if ( (errno = pthread_join(thread, &result)) )
-		err(1, "pthread_cancel");
-	if ( result != PTHREAD_CANCELED )
-		errx(1, "pthread_join() != PTHREAD_CANCELED");
-	if ( !cleaned_up )
-		errx(1, "cleanup handler was not executed");
 	return 0;
 }
diff --git os-test-clean/basic/pthread/pthread_cond_broadcast.c os-test-workdir/basic/pthread/pthread_cond_broadcast.c
index 14c519f..fb5eb53 100644
--- os-test-clean/basic/pthread/pthread_cond_broadcast.c
+++ os-test-workdir/basic/pthread/pthread_cond_broadcast.c
@@ -45,34 +45,5 @@ static void* start(void* ctx)
 
 int main(void)
 {
-	if ( (errno = pthread_cond_init(&cnd_sync, NULL)) )
-		err(1, "pthread_cond_wait");
-	if ( (errno = pthread_cond_init(&cnd_wake, NULL)) )
-		err(1, "pthread_cond_wait");
-	if ( (errno = pthread_mutex_init(&mtx, NULL)) )
-		err(1, "pthread_mutex_init");
-	int id1 = 1, id2 = 2;
-	pthread_t thrd1, thrd2;
-	if ( (errno = pthread_create(&thrd1, NULL, start, &id1)) )
-		err(1, "pthread_create");
-	if ( (errno = pthread_create(&thrd2, NULL, start, &id2)) )
-		err(1, "pthread_create");
-	lock(&mtx);
-	// Wait for all threads to be asleep in pthread_cond_wait.
-	while ( waiting < 2 )
-	{
-		if ( (errno = pthread_cond_wait(&cnd_sync, &mtx)) )
-			err(1, "pthread_cond_wait");
-	}
-	// Test that all threads awake with pthread_cond_broadcast.
-	woken = 1;
-	if ( (errno = pthread_cond_broadcast(&cnd_wake)) )
-		err(1, "pthread_cond_broadcast");
-	unlock(&mtx);
-	void* code;
-	if ( (errno = pthread_join(thrd1, &code)) )
-		err(1, "pthread_cond_broadcast");
-	if ( (errno = pthread_join(thrd2, &code)) )
-		err(1, "pthread_cond_broadcast");
 	return 0;
 }
diff --git os-test-clean/basic/pthread/pthread_condattr_setpshared.c os-test-workdir/basic/pthread/pthread_condattr_setpshared.c
index ce7f879..066799a 100644
--- os-test-clean/basic/pthread/pthread_condattr_setpshared.c
+++ os-test-workdir/basic/pthread/pthread_condattr_setpshared.c
@@ -27,81 +27,5 @@ static void exit_handler(void)
 
 int main(void)
 {
-	// Allocate a shared page.
-	long page_size = sysconf(_SC_PAGESIZE);
-	if ( page_size < 0 )
-		err(1, "sysconf _SC_PAGESIZE");
-	size_t size = sizeof(struct page);
-	size = -(-size & ~(page_size-1)); // Align
-	void* ptr = mmap(NULL, page_size, PROT_READ | PROT_WRITE,
-	                 MAP_ANONYMOUS | MAP_SHARED, -1, 0);
-	if ( ptr == MAP_FAILED )
-		err(1, "mmap");
-	struct page* page = ptr;
-	// Make a shared cond variable.
-	int shared = PTHREAD_PROCESS_SHARED;
-	pthread_condattr_t condattr;
-	if ( (errno = pthread_condattr_init(&condattr)) )
-		err(1, "pthread_condattr_init");
-	if ( (errno = pthread_condattr_setpshared(&condattr, shared)) )
-		err(1, "pthread_condattr_setpshared");
-	pthread_cond_t* cond = &page->cond;
-	if ( (errno = pthread_cond_init(cond, &condattr)) )
-		err(1, "pthread_cond_init");
-	// Make a shared mutex.
-	pthread_mutexattr_t mutexattr;
-	if ( (errno = pthread_mutexattr_init(&mutexattr)) )
-		err(1, "pthread_mutexattr_init");
-	if ( (errno = pthread_mutexattr_setpshared(&mutexattr, shared)) )
-		err(1, "pthread_mutexattr_setpshared");
-	pthread_mutex_t* mutex = &page->mutex;
-	if ( (errno = pthread_mutex_init(mutex, &mutexattr)) )
-		err(1, "pthread_mutex_init");
-	// Set up a handler to kill the child process if we died too.
-	if ( atexit(exit_handler) )
-		err(1, "atexit");
-	// Fork.
-	if ( (child_pid = fork()) < 0 )
-		err(1, "fork");
-	// Get the mutex in either process.
-	if ( (errno = pthread_mutex_lock(mutex)) )
-		err(1, "pthread_mutex_lock");
-	// In the parent, set state to 1 and wait for state go to 2.
-	if ( child_pid )
-	{
-		page->state = 1;
-		if ( (errno = pthread_cond_signal(cond)) )
-			err(1, "pthread_cond_signal");
-		while ( page->state != 2 )
-			if ( (errno = pthread_cond_wait(cond, mutex)) )
-				err(1, "pthread_cond_wait");
-	}
-	// In the child, wait for state 1 and set state to 2.
-	else
-	{
-		while ( page->state != 1 )
-			if ( (errno = pthread_cond_wait(cond, mutex)) )
-				err(1, "pthread_cond_wait");
-		page->state = 2;
-		if ( (errno = pthread_cond_signal(cond)) )
-			err(1, "pthread_cond_signal");
-
-	}
-	// Release the mutex in either process.
-	if ( (errno = pthread_mutex_unlock(mutex)) )
-		err(1, "pthread_mutex_unlock");
-	// Collect the child process.
-	if ( !child_pid )
-		return 0;
-	int status;
-	if ( waitpid(child_pid, &status, 0) < 0 )
-		err(1, "waitpid");
-	child_pid = 0;
-	if ( WIFEXITED(status) )
-		return WEXITSTATUS(status);
-	else if ( WIFSIGNALED(status) )
-		errx(1, "%s", strsignal(WTERMSIG(status)));
-	else
-		errx(1, "unknown exit: %#x", status);
 	return 0;
 }
diff --git os-test-clean/basic/pthread/pthread_join.c os-test-workdir/basic/pthread/pthread_join.c
index 6abd28c..a812228 100644
--- os-test-clean/basic/pthread/pthread_join.c
+++ os-test-workdir/basic/pthread/pthread_join.c
@@ -13,13 +13,5 @@ static void* start(void* ctx)
 
 int main(void)
 {
-	pthread_t thrd;
-	if ( (errno = pthread_create(&thrd, NULL, start, NULL)) )
-		err(1, "pthread_create");
-	void* code;
-	if ( (errno = pthread_join(thrd, &code)) )
-		err(1, "pthread_join");
-	if ( (uintptr_t) code != 42 )
-		errx(1, "pthread_join was not 42 (got %d)", code);
 	return 0;
 }
diff --git os-test-clean/basic/pthread/pthread_mutex_consistent.c os-test-workdir/basic/pthread/pthread_mutex_consistent.c
index 5d03c5e..93b9a4a 100644
--- os-test-clean/basic/pthread/pthread_mutex_consistent.c
+++ os-test-workdir/basic/pthread/pthread_mutex_consistent.c
@@ -18,44 +18,5 @@ static void* start(void* ctx)
 
 int main(void)
 {
-	alarm(1); // Hurd.
-	// Create a robust mutex.
-	pthread_mutexattr_t attr;
-	if ( (errno = pthread_mutexattr_init(&attr)) )
-		err(1, "pthread_mutexattr_init");
-	if ( (errno = pthread_mutexattr_setrobust(&attr, PTHREAD_MUTEX_ROBUST)) )
-		err(1, "pthread_mutexattr_setrobust");
-	if ( (errno = pthread_mutex_init(&mutex, &attr)) )
-		err(1, "pthread_mutex_init");
-	// Lock the lock in a terminated thread.
-	pthread_t thread;
-	if ( (errno = pthread_create(&thread, NULL, start, NULL)) )
-		err(1, "pthread_create");
-	void* code;
-	if ( (errno = pthread_join(thread, &code)) )
-		err(1, "pthread_join");
-	// Verify the robust mutex fails with EOWNERDEAD.
-	if ( (errno = pthread_mutex_lock(&mutex)) )
-	{
-		if ( errno != EOWNERDEAD )
-			err(1, "second pthread_mutex_lock");
-	}
-	else
-		errx(1, "second pthread_mutex_lock did not fail");
-	// Verify we got ownership after EOWNERDEAD.
-	if ( (errno = pthread_mutex_trylock(&mutex)) )
-	{
-		if ( errno != EBUSY )
-			err(1, "pthread_mutex_trylock");
-	}
-	else
-		errx(1, "pthread_mutex_trylock did not fail");
-	// Verify the mutex can be repaired.
-	if ( (errno = pthread_mutex_consistent(&mutex)) )
-		err(1, "pthread_mutex_consistent");
-	if ( (errno = pthread_mutex_unlock(&mutex)) )
-		err(1, "second pthread_mutex_unlock");
-	if ( (errno = pthread_mutex_lock(&mutex)) )
-		err(1, "second pthread_mutex_lock");
 	return 0;
 }
diff --git os-test-clean/basic/pthread/pthread_setcancelstate.c os-test-workdir/basic/pthread/pthread_setcancelstate.c
index a6909a8..cd16f39 100644
--- os-test-clean/basic/pthread/pthread_setcancelstate.c
+++ os-test-workdir/basic/pthread/pthread_setcancelstate.c
@@ -38,34 +38,5 @@ static void* start(void* ctx)
 
 int main(void)
 {
-	// Test the initial state.
-	int old_state;
-	if ( (errno = pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, &old_state)) )
-		err(1, "pthread_setcancelstate deferred");
-	if ( old_state != PTHREAD_CANCEL_ENABLE )
-		errx(1, "initial main cancel state was disabled");
-	// Create a thread to be canceled.
-	pthread_t thread;
-	if ( (errno = pthread_create(&thread, NULL, start, NULL)) )
-		err(1, "pthread_create");
-	// Cancel the thread. This won't take effect immediately because the thread
-	// disables cancellation before the first cancellation point.
-	if ( (errno = pthread_cancel(thread)) )
-		err(1, "pthread_cancel");
-	// Notify the thread that cancellation is now pending, so we can proceed.
-	pthread_mutex_lock(&mutex);
-	pthread_cond_signal(&cond);
-	canceled = true;
-	pthread_mutex_unlock(&mutex);
-	// Wait for the thread to be canceled.
-	void* result;
-	if ( (errno = pthread_join(thread, &result)) )
-		err(1, "pthread_cancel");
-	// Test the thread was canceled.
-	if ( result != PTHREAD_CANCELED )
-		errx(1, "pthread_join() != PTHREAD_CANCELED");
-	// Test the thread was canceled at the correct location.
-	if ( !at_cancellation_point )
-		errx(1, "cancellation not at point");
 	return 0;
 }
diff --git os-test-clean/basic/pthread/pthread_setcanceltype.c os-test-workdir/basic/pthread/pthread_setcanceltype.c
index fe6e690..e8e9bc5 100644
--- os-test-clean/basic/pthread/pthread_setcanceltype.c
+++ os-test-workdir/basic/pthread/pthread_setcanceltype.c
@@ -29,21 +29,5 @@ static void* start(void* ctx)
 
 int main(void)
 {
-	alarm(1); // Haiku.
-	int old_state;
-	if ( (errno = pthread_setcanceltype(PTHREAD_CANCEL_DEFERRED, &old_state)) )
-		err(1, "pthread_setcanceltype deferred");
-	if ( old_state != PTHREAD_CANCEL_DEFERRED )
-		err(1, "initial main cancel type was not deferred");
-	pthread_t thread;
-	if ( (errno = pthread_create(&thread, NULL, start, NULL)) )
-		err(1, "pthread_create");
-	if ( (errno = pthread_cancel(thread)) )
-		err(1, "pthread_cancel");
-	void* result;
-	if ( (errno = pthread_join(thread, &result)) )
-		err(1, "pthread_cancel");
-	if ( result != PTHREAD_CANCELED )
-		errx(1, "pthread_join() != PTHREAD_CANCELED");
 	return 0;
 }
diff --git os-test-clean/basic/pthread/pthread_setspecific.c os-test-workdir/basic/pthread/pthread_setspecific.c
index 23877b9..c485f70 100644
--- os-test-clean/basic/pthread/pthread_setspecific.c
+++ os-test-workdir/basic/pthread/pthread_setspecific.c
@@ -27,23 +27,5 @@ static void* start(void* ctx)
 
 int main(void)
 {
-	if ( (errno = pthread_key_create(&tss, destructor)) )
-		err(1, "pthread_key_create");
-	if ( pthread_getspecific(tss) )
-		errx(1, "main pthread_getspecific returned non-null");
-	if ( (errno = pthread_setspecific(tss, &id1)) )
-		err(1, "pthread_setspecific");
-	if ( pthread_getspecific(tss) != &id1 )
-		errx(1, "first main pthread_getspecific did not return id1");
-	pthread_t thrd;
-	if ( (errno = pthread_create(&thrd, NULL, start, NULL)) )
-		err(1, "pthread_create");
-	void* code;
-	if ( (errno = pthread_join(thrd, &code)) )
-		err(1, "pthread_join");
-	if ( pthread_getspecific(tss) != &id1 )
-		errx(1, "second main pthread_getspecific did not return id1");
-	if ( invoked != id2 )
-		errx(1, "destructor was not run");
 	return 0;
 }
diff --git os-test-clean/basic/pthread/pthread_testcancel.c os-test-workdir/basic/pthread/pthread_testcancel.c
index a844b98..eeffea5 100644
--- os-test-clean/basic/pthread/pthread_testcancel.c
+++ os-test-workdir/basic/pthread/pthread_testcancel.c
@@ -15,15 +15,5 @@ static void* start(void* ctx)
 
 int main(void)
 {
-	pthread_t thread;
-	if ( (errno = pthread_create(&thread, NULL, start, NULL)) )
-		err(1, "pthread_create");
-	if ( (errno = pthread_cancel(thread)) )
-		err(1, "pthread_cancel");
-	void* result;
-	if ( (errno = pthread_join(thread, &result)) )
-		err(1, "pthread_cancel");
-	if ( result != PTHREAD_CANCELED )
-		errx(1, "pthread_join() != PTHREAD_CANCELED");
 	return 0;
 }
diff --git os-test-clean/basic/signal/sigsuspend.c os-test-workdir/basic/signal/sigsuspend.c
index 30a78d1..51124b2 100644
--- os-test-clean/basic/signal/sigsuspend.c
+++ os-test-workdir/basic/signal/sigsuspend.c
@@ -16,33 +16,5 @@ static void handler(int x)
 
 int main(void)
 {
-	struct sigaction sa = { 0 };
-	sa.sa_handler = handler;
-	sa.sa_flags = 0;
-	sigemptyset(&sa.sa_mask);
-	sigaction(SIGUSR1, &sa, NULL);
-
-	sigset_t oldmask;
-	sigset_t sigmask;
-	sigemptyset(&sigmask);
-	sigaddset(&sigmask, SIGUSR1);
-
-	// Block SIGUSR1 and obtain the current signal mask with SIGUSR1 unblocked
-	sigprocmask(SIG_BLOCK, &sigmask, &oldmask);
-	sigdelset(&oldmask, SIGUSR1);
-
-	// Queue a SIGUSR1 signal which will not be delivered.
-	raise(SIGUSR1);
-
-	// Wait for a signal to arrive that is not masked by oldmask, which
-	// should be SIGUSR1
-	if ( !sigsuspend(&oldmask) )
-		errx(1, "sigsuspend succeeded");
-	if ( errno != EINTR )
-		err(1, "sigsuspend");
-
-	if ( signaled == 0 )
-		errx(1, "signal handler was not executed");
-
 	return 0;
 }
diff --git os-test-clean/basic/signal/sigwait.c os-test-workdir/basic/signal/sigwait.c
index 591ae03..86b74fe 100644
--- os-test-clean/basic/signal/sigwait.c
+++ os-test-workdir/basic/signal/sigwait.c
@@ -6,14 +6,5 @@
 
 int main(void)
 {
-	sigset_t set;
-	sigemptyset(&set);
-	sigaddset(&set, SIGUSR1);
-	if ( sigprocmask(SIG_BLOCK, &set, NULL) < 0 )
-		err(1, "sigprocmask");
-	raise(SIGUSR1);
-	int signo;
-	if ( (errno = sigwait(&set, &signo)) < 0 )
-		err(1, "sigwait");
 	return 0;
 }
diff --git os-test-clean/basic/spawn/posix_spawnattr_setschedpolicy.c os-test-workdir/basic/spawn/posix_spawnattr_setschedpolicy.c
index 16a7ba2..4e4e234 100644
--- os-test-clean/basic/spawn/posix_spawnattr_setschedpolicy.c
+++ os-test-workdir/basic/spawn/posix_spawnattr_setschedpolicy.c
@@ -15,52 +15,5 @@ extern char** environ;
 
 int main(int argc, char* argv[])
 {
-	if ( argc == 2 )
-	{
-		if ( strcmp(argv[1], "success") != 0 )
-			err(1, "child invoked incorrectly");
-		return 0;
-	}
-	// Test setting the scheduler policy. Obtaining the current settings may
-	// fail with EPERM, which is ignored.
-	int policy;
-	struct sched_param param = {0};
-	if ( (policy = sched_getscheduler(0)) < 0 &&
-		 (errno = pthread_getschedparam(pthread_self(), &policy, &param)) )
-		policy = SCHED_OTHER;
-	posix_spawnattr_t attr;
-	if ( (errno = posix_spawnattr_init(&attr)) )
-		err(1, "posix_spawnattr_init");
-	if ( (errno = posix_spawnattr_setflags(&attr, POSIX_SPAWN_SETSCHEDULER)) )
-		err(1, "posix_spawnattr_setflags");
-	if ( (errno = posix_spawnattr_setschedpolicy(&attr, policy)) )
-		err(1, "posix_spawnattr_setschedpolicy");
-	char* new_argv[] =
-	{
-		argv[0],
-		"success",
-		NULL,
-	};
-	pid_t pid;
-	if ( (errno = posix_spawn(&pid, argv[0], NULL, &attr, new_argv, environ)) )
-	{
-		if ( errno == EPERM )
-			return 0;
-		err(1, "posix_spawn: %s", new_argv[0]);
-	}
-	int status;
-	if ( waitpid(pid, &status, 0) < 0 )
-		err(1, "waitpid");
-	if ( WIFEXITED(status) )
-	{
-		// Exit 127 may mean EPERM inside the new process.
-		if ( WEXITSTATUS(status) == 127 )
-			return 0;
-		return WEXITSTATUS(status);
-	}
-	else if ( WIFSIGNALED(status) )
-		errx(1, "%s", strsignal(WTERMSIG(status)));
-	else
-		errx(1, "unknown exit: %#x", status);
 	return 0;
 }
diff --git os-test-clean/basic/sys_wait/wait.c os-test-workdir/basic/sys_wait/wait.c
index 858f34c..6875343 100644
--- os-test-clean/basic/sys_wait/wait.c
+++ os-test-workdir/basic/sys_wait/wait.c
@@ -9,46 +9,5 @@
 
 int main(void)
 {
-	// Make a child that exits 0.
-	pid_t child1 = fork();
-	if ( child1 < 0 )
-		err(1, "first fork");
-	if ( !child1 )
-		_exit(0);
-	// Make another child that exits 0.
-	pid_t child2 = fork();
-	if ( child2 < 0 )
-		err(1, "second fork");
-	if ( !child2 )
-		_exit(0);
-	// Wait for a child to exit.
-	int status1;
-	pid_t wait1 = wait(&status1);
-	if ( wait1 < 0 )
-		err(1, "first wait");
-	if ( wait1 != child1 && wait1 != child2 )
-		errx(1, "first wait gave strange child");
-	if ( !WIFEXITED(status1) || WEXITSTATUS(status1) != 0 )
-		errx(1, "%s child did not exit 0", wait1 == child1 ? "first" : "second");
-	// Wait for the other child to exit.
-	int status2;
-	pid_t wait2 = wait(&status2);
-	if ( wait2 < 0 )
-		err(1, "second wait");
-	if ( wait2 != child1 && wait2 != child2 )
-		errx(1, "second wait gave strange child");
-	if ( wait1 == wait2 )
-		errx(1, "second wait gave the same child");
-	if ( !WIFEXITED(status2) || WEXITSTATUS(status2) != 0 )
-		errx(1, "%s child did not exit 0", wait2 == child1 ? "first" : "second");
-	// Test wait with no children left.
-	int status3;
-	if ( wait(&status3) < 0 )
-	{
-		if ( errno != ECHILD )
-			err(1, "third wait");
-	}
-	else
-		errx(1, "third wait succeeded with no children");
 	return 0;
 }
diff --git os-test-clean/basic/threads/cnd_broadcast.c os-test-workdir/basic/threads/cnd_broadcast.c
index fd85a6c..7643f58 100644
--- os-test-clean/basic/threads/cnd_broadcast.c
+++ os-test-workdir/basic/threads/cnd_broadcast.c
@@ -70,81 +70,5 @@ static int start(void* ctx)
 
 int main(void)
 {
-	int ret = cnd_init(&cnd_sync);
-	if ( ret != thrd_success )
-		errx(1, "first cnd_init: %s",
-		     ret == thrd_busy ? "thrd_busy" :
-		     ret == thrd_nomem ? "thrd_nomem" :
-		     ret == thrd_timedout ? "thrd_timedout" :
-		     ret == thrd_error ? "thrd_error" :
-		     "thrd_unknown");
-	ret = cnd_init(&cnd_wake);
-	if ( ret != thrd_success )
-		errx(1, "second cnd_init: %s",
-		     ret == thrd_busy ? "thrd_busy" :
-		     ret == thrd_nomem ? "thrd_nomem" :
-		     ret == thrd_timedout ? "thrd_timedout" :
-		     ret == thrd_error ? "thrd_error" :
-		     "thrd_unknown");
-	ret = mtx_init(&mtx, mtx_plain);
-	if ( ret != thrd_success )
-		errx(1, "mtx_init: %s",
-		     ret == thrd_busy ? "thrd_busy" :
-		     ret == thrd_nomem ? "thrd_nomem" :
-		     ret == thrd_timedout ? "thrd_timedout" :
-		     ret == thrd_error ? "thrd_error" :
-		     "thrd_unknown");
-	int id1 = 1, id2 = 2;
-	thrd_t thrd1, thrd2;
-	ret = thrd_create(&thrd1, start, &id1);
-	if ( ret != thrd_success )
-		errx(1, "first thrd_create: %s",
-		     ret == thrd_busy ? "thrd_busy" :
-		     ret == thrd_nomem ? "thrd_nomem" :
-		     ret == thrd_timedout ? "thrd_timedout" :
-		     ret == thrd_error ? "thrd_error" :
-		     "thrd_unknown");
-	ret = thrd_create(&thrd2, start, &id2);
-	if ( ret != thrd_success )
-		errx(1, "second thrd_create: %s",
-		     ret == thrd_busy ? "thrd_busy" :
-		     ret == thrd_nomem ? "thrd_nomem" :
-		     ret == thrd_timedout ? "thrd_timedout" :
-		     ret == thrd_error ? "thrd_error" :
-		     "thrd_unknown");
-	lock(&mtx);
-	// Wait for all threads to be asleep in cnd_wait.
-	while ( waiting < 2 )
-	{
-		ret = cnd_wait(&cnd_sync, &mtx);
-		if ( ret != thrd_success )
-			errx(1, "main cnd_wait: %s",
-				 ret == thrd_busy ? "thrd_busy" :
-				 ret == thrd_nomem ? "thrd_nomem" :
-				 ret == thrd_timedout ? "thrd_timedout" :
-				 ret == thrd_error ? "thrd_error" :
-				 "thrd_unknown");
-	}
-	// Test that all threads awake with cnd_broadcast.
-	woken = 1;
-	cnd_broadcast(&cnd_wake);
-	unlock(&mtx);
-	int code;
-	ret = thrd_join(thrd1, &code);
-	if ( ret != thrd_success )
-		errx(1, "first thrd_join: %s",
-		     ret == thrd_busy ? "thrd_busy" :
-		     ret == thrd_nomem ? "thrd_nomem" :
-		     ret == thrd_timedout ? "thrd_timedout" :
-		     ret == thrd_error ? "thrd_error" :
-		     "thrd_unknown");
-	ret = thrd_join(thrd2, &code);
-	if ( ret != thrd_success )
-		errx(1, "second thrd_join: %s",
-		     ret == thrd_busy ? "thrd_busy" :
-		     ret == thrd_nomem ? "thrd_nomem" :
-		     ret == thrd_timedout ? "thrd_timedout" :
-		     ret == thrd_error ? "thrd_error" :
-		     "thrd_unknown");
 	return 0;
 }
diff --git os-test-clean/basic/threads/thrd_join.c os-test-workdir/basic/threads/thrd_join.c
index cf0f394..bcd2626 100644
--- os-test-clean/basic/threads/thrd_join.c
+++ os-test-workdir/basic/threads/thrd_join.c
@@ -12,25 +12,5 @@ static int start(void* ctx)
 
 int main(void)
 {
-	thrd_t thrd;
-	int ret = thrd_create(&thrd, start, NULL);
-	if ( ret != thrd_success )
-		errx(1, "thrd_create: %s",
-		     ret == thrd_busy ? "thrd_busy" :
-		     ret == thrd_nomem ? "thrd_nomem" :
-		     ret == thrd_timedout ? "thrd_timedout" :
-		     ret == thrd_error ? "thrd_error" :
-		     "thrd_unknown");
-	int code;
-	ret = thrd_join(thrd, &code);
-	if ( ret != thrd_success )
-		errx(1, "thrd_join: %s",
-		     ret == thrd_busy ? "thrd_busy" :
-		     ret == thrd_nomem ? "thrd_nomem" :
-		     ret == thrd_timedout ? "thrd_timedout" :
-		     ret == thrd_error ? "thrd_error" :
-		     "thrd_unknown");
-	if ( code != 42 )
-		errx(1, "thrd_join was not 42 (got %d)", code);
 	return 0;
 }
diff --git os-test-clean/basic/threads/tss_set.c os-test-workdir/basic/threads/tss_set.c
index d835321..bcc3bee 100644
--- os-test-clean/basic/threads/tss_set.c
+++ os-test-workdir/basic/threads/tss_set.c
@@ -33,47 +33,5 @@ static int start(void* ctx)
 
 int main(void)
 {
-	int ret = tss_create(&tss, destructor);
-	if ( ret != thrd_success )
-		errx(1, "tss_create: %s",
-		     ret == thrd_busy ? "thrd_busy" :
-		     ret == thrd_nomem ? "thrd_nomem" :
-		     ret == thrd_timedout ? "thrd_timedout" :
-		     ret == thrd_error ? "thrd_error" :
-		     "thrd_unknown");
-	if ( tss_get(tss) )
-		errx(1, "main tss_get returned non-null");
-	ret = tss_set(tss, &id1);
-	if ( ret != thrd_success )
-		errx(1, "tss_set: %s",
-		     ret == thrd_busy ? "thrd_busy" :
-		     ret == thrd_nomem ? "thrd_nomem" :
-		     ret == thrd_timedout ? "thrd_timedout" :
-		     ret == thrd_error ? "thrd_error" :
-		     "thrd_unknown");
-	if ( tss_get(tss) != &id1 )
-		errx(1, "first main tss_get did not return id1");
-	thrd_t thrd;
-	ret = thrd_create(&thrd, start, NULL);
-	if ( ret != thrd_success )
-		errx(1, "thrd_create: %s",
-		     ret == thrd_busy ? "thrd_busy" :
-		     ret == thrd_nomem ? "thrd_nomem" :
-		     ret == thrd_timedout ? "thrd_timedout" :
-		     ret == thrd_error ? "thrd_error" :
-		     "thrd_unknown");
-	int code;
-	ret = thrd_join(thrd, &code);
-	if ( ret != thrd_success )
-		errx(1, "thrd_join: %s",
-		     ret == thrd_busy ? "thrd_busy" :
-		     ret == thrd_nomem ? "thrd_nomem" :
-		     ret == thrd_timedout ? "thrd_timedout" :
-		     ret == thrd_error ? "thrd_error" :
-		     "thrd_unknown");
-	if ( tss_get(tss) != &id1 )
-		errx(1, "second main tss_get did not return id1");
-	if ( invoked != id2 )
-		errx(1, "destructor was not run");
 	return 0;
 }
diff --git os-test-clean/basic/unistd/pause.c os-test-workdir/basic/unistd/pause.c
index 9ff7d7a..4c9db09 100644
--- os-test-clean/basic/unistd/pause.c
+++ os-test-workdir/basic/unistd/pause.c
@@ -13,34 +13,5 @@ void on_alarm(int signum)
 
 int main(void)
 {
-	struct sigaction sa = { .sa_handler = on_alarm };
-	if ( sigaction(SIGUSR1, &sa, NULL) < 0 )
-		err(1, "sigaction");
-	pid_t parent = getpid();
-	pid_t child = fork();
-	if ( child < 0 )
-		err(1, "fork");
-	if ( !child )
-	{
-#ifndef __minix__
-		sched_yield();
-#endif
-		while ( 1 )
-		{
-			if ( kill(parent, SIGUSR1) < 0 )
-				exit(0);
-			sleep(1);
-		}
-	}
-	int ret = pause();
-	int errnum = errno;
-	kill(child, SIGKILL);
-	errno = errnum;
-	if ( ret < 0 && errno == EINTR )
-		;
-	else if ( ret < 0 )
-		err(1, "pause");
-	else if ( ret == 0 )
-		errx(1, "pause() did not fail");
 	return 0;
 }
diff --git os-test-clean/process/waitpid-pgid-empty-on-setpgid-rejoin.c os-test-workdir/process/waitpid-pgid-empty-on-setpgid-rejoin.c
index b3d5c0b..d4a6a71 100644
--- os-test-clean/process/waitpid-pgid-empty-on-setpgid-rejoin.c
+++ os-test-workdir/process/waitpid-pgid-empty-on-setpgid-rejoin.c
@@ -12,88 +12,5 @@ void on_signal(int signo)
 
 int main(void)
 {
-	int alive_fd[2], pgid_fd[2];
-	if ( pipe(alive_fd) < 0 || pipe(pgid_fd) < 0 )
-		err(1, "pipe");
-#ifdef __APPLE__
-	pid_t original_pid = getpid();
-#endif
-	// Make a process group that is not a direct child so waitpid can't see it.
-	pid_t pgid = fork();
-	if ( pgid < 0 )
-		err(1, "fork");
-	if ( !pgid )
-	{
-		close(alive_fd[1]);
-		close(pgid_fd[0]);
-		// Double fork to not be a direct child.
-		pgid = fork();
-		if ( pgid < 0 )
-			err(1, "fork");
-		if ( pgid )
-			_exit(0);
-		// And become a process group leader.
-		if ( setpgid(0, 0) < 0 )
-			err(1, "setpgid");
-		// Tell the original process what our process group id is.
-		pgid = getpgid(0);
-		if ( write(pgid_fd[1], &pgid, sizeof(pgid)) < (ssize_t) sizeof(pgid) )
-			err(1, "write");
-		close(pgid_fd[1]);
-#ifdef __APPLE__
-		// macOS seems to hang in the test, despite SIGALRM, so give it an extra
-		// fallback timeout kick.
-		alarm(4);
-		sleep(3);
-		kill(original_pid, SIGKILL);
-		_exit(1);
-#endif
-		// Stay alive as long as the alive pipe is connected to the original
-		// process, so we exit when it does. It will never write here.
-		char c;
-		read(alive_fd[0], &c, 1);
-		_exit(0);
-	}
-	// Wait for the process group to exist and receive it's id.
-	if ( read(pgid_fd[0], &pgid, sizeof(pgid)) < (ssize_t) sizeof(pgid) )
-		err(1, "read");
-	// Make a child that will leave it's process group after a moment.
-	pid_t child = fork();
-	if ( child < 0 )
-		err(1, "fork");
-	if ( !child )
-	{
-#ifdef __APPLE__
-		alarm(4);
-#endif
-		// While we sleep, the parent will place us in the process group and go
-		// to sleep inside waitpid.
-		sleep(1);
-		// Leave the process group by rejoining the parent's process group.
-		if ( setpgid(getpid(), getpgid(getppid())) < 0 )
-			err(1, "setpgid rejoin");
-		// Stay alive as long as the alive pipe is connected to the original
-		// process, so we exit when it does. It will never write here.
-		char c;
-		read(alive_fd[0], &c, 1);
-		_exit(0);
-	}
-	// Place the child inside the indirect process group.
-	if ( setpgid(child, pgid) < 0 )
-		err(1, "setpgid of child into pgid");
-	// Time out if the process hangs inside waitpid.
-	signal(SIGALRM, on_signal);
-	alarm(2);
-	int status;
-	// waitpid is supposed to ECHLD here when the child changes its pgid.
-	pid_t result = waitpid(-pgid, &status, 0);
-	if ( result < 0 )
-		err(1, "waitpid");
-	else if ( result == 0 )
-		errx(1, "waitpid() == 0");
-	else if ( result == child )
-		errx(1, "waitpid returned child");
-	else
-		errx(1, "waitpid returned strange child");
 	return 0;
 }
diff --git os-test-clean/process/waitpid-pgid-empty-on-setpgid.c os-test-workdir/process/waitpid-pgid-empty-on-setpgid.c
index 1c67d2f..9f8f88b 100644
--- os-test-clean/process/waitpid-pgid-empty-on-setpgid.c
+++ os-test-workdir/process/waitpid-pgid-empty-on-setpgid.c
@@ -12,88 +12,5 @@ void on_signal(int signo)
 
 int main(void)
 {
-	int alive_fd[2], pgid_fd[2];
-	if ( pipe(alive_fd) < 0 || pipe(pgid_fd) < 0 )
-		err(1, "pipe");
-#ifdef __APPLE__
-	pid_t original_pid = getpid();
-#endif
-	// Make a process group that is not a direct child so waitpid can't see it.
-	pid_t pgid = fork();
-	if ( pgid < 0 )
-		err(1, "fork");
-	if ( !pgid )
-	{
-		close(alive_fd[1]);
-		close(pgid_fd[0]);
-		// Double fork to not be a direct child.
-		pgid = fork();
-		if ( pgid < 0 )
-			err(1, "fork");
-		if ( pgid )
-			_exit(0);
-		// And become a process group leader.
-		if ( setpgid(0, 0) < 0 )
-			err(1, "setpgid");
-		// Tell the original process what our process group id is.
-		pgid = getpgid(0);
-		if ( write(pgid_fd[1], &pgid, sizeof(pgid)) < (ssize_t) sizeof(pgid) )
-			err(1, "write");
-		close(pgid_fd[1]);
-#ifdef __APPLE__
-		// macOS seems to hang in the test, despite SIGALRM, so give it an extra
-		// fallback timeout kick.
-		alarm(4);
-		sleep(3);
-		kill(original_pid, SIGKILL);
-		_exit(1);
-#endif
-		// Stay alive as long as the alive pipe is connected to the original
-		// process, so we exit when it does. It will never write here.
-		char c;
-		read(alive_fd[0], &c, 1);
-		_exit(0);
-	}
-	// Wait for the process group to exist and receive it's id.
-	if ( read(pgid_fd[0], &pgid, sizeof(pgid)) < (ssize_t) sizeof(pgid) )
-		err(1, "read");
-	// Make a child that will leave it's process group after a moment.
-	pid_t child = fork();
-	if ( child < 0 )
-		err(1, "fork");
-	if ( !child )
-	{
-#ifdef __APPLE__
-		alarm(4);
-#endif
-		// While we sleep, the parent will place us in the process group and go
-		// to sleep inside waitpid.
-		sleep(1);
-		// Leave the process group by making a new process group.
-		if ( setpgid(0, 0) < 0 )
-			err(1, "setpgid new process group");
-		// Stay alive as long as the alive pipe is connected to the original
-		// process, so we exit when it does. It will never write here.
-		char c;
-		read(alive_fd[0], &c, 1);
-		_exit(0);
-	}
-	// Place the child inside the indirect process group.
-	if ( setpgid(child, pgid) < 0 )
-		err(1, "setpgid of child into pgid");
-	// Time out if the process hangs inside waitpid.
-	signal(SIGALRM, on_signal);
-	alarm(2);
-	int status;
-	// waitpid is supposed to ECHLD here when the child changes its pgid.
-	pid_t result = waitpid(-pgid, &status, 0);
-	if ( result < 0 )
-		err(1, "waitpid");
-	else if ( result == 0 )
-		errx(1, "waitpid() == 0");
-	else if ( result == child )
-		errx(1, "waitpid returned child");
-	else
-		errx(1, "waitpid returned strange child");
 	return 0;
 }
diff --git os-test-clean/process/waitpid-pgid-empty-on-setsid.c os-test-workdir/process/waitpid-pgid-empty-on-setsid.c
index 635a985..cae5a61 100644
--- os-test-clean/process/waitpid-pgid-empty-on-setsid.c
+++ os-test-workdir/process/waitpid-pgid-empty-on-setsid.c
@@ -12,88 +12,5 @@ void on_signal(int signo)
 
 int main(void)
 {
-	int alive_fd[2], pgid_fd[2];
-	if ( pipe(alive_fd) < 0 || pipe(pgid_fd) < 0 )
-		err(1, "pipe");
-#ifdef __APPLE__
-	pid_t original_pid = getpid();
-#endif
-	// Make a process group that is not a direct child so waitpid can't see it.
-	pid_t pgid = fork();
-	if ( pgid < 0 )
-		err(1, "fork");
-	if ( !pgid )
-	{
-		close(alive_fd[1]);
-		close(pgid_fd[0]);
-		// Double fork to not be a direct child.
-		pgid = fork();
-		if ( pgid < 0 )
-			err(1, "fork");
-		if ( pgid )
-			_exit(0);
-		// And become a process group leader.
-		if ( setpgid(0, 0) < 0 )
-			err(1, "setpgid");
-		// Tell the original process what our process group id is.
-		pgid = getpgid(0);
-		if ( write(pgid_fd[1], &pgid, sizeof(pgid)) < (ssize_t) sizeof(pgid) )
-			err(1, "write");
-		close(pgid_fd[1]);
-#ifdef __APPLE__
-		// macOS seems to hang in the test, despite SIGALRM, so give it an extra
-		// fallback timeout kick.
-		alarm(4);
-		sleep(3);
-		kill(original_pid, SIGKILL);
-		_exit(1);
-#endif
-		// Stay alive as long as the alive pipe is connected to the original
-		// process, so we exit when it does. It will never write here.
-		char c;
-		read(alive_fd[0], &c, 1);
-		_exit(0);
-	}
-	// Wait for the process group to exist and receive it's id.
-	if ( read(pgid_fd[0], &pgid, sizeof(pgid)) < (ssize_t) sizeof(pgid) )
-		err(1, "read");
-	// Make a child that will leave it's process group after a moment.
-	pid_t child = fork();
-	if ( child < 0 )
-		err(1, "fork");
-	if ( !child )
-	{
-#ifdef __APPLE__
-		alarm(4);
-#endif
-		// While we sleep, the parent will place us in the process group and go
-		// to sleep inside waitpid.
-		sleep(1);
-		// Leave the process group by making a new session.
-		if ( setsid() < 0 )
-			err(1, "setsid");
-		// Stay alive as long as the alive pipe is connected to the original
-		// process, so we exit when it does. It will never write here.
-		char c;
-		read(alive_fd[0], &c, 1);
-		_exit(0);
-	}
-	// Place the child inside the indirect process group.
-	if ( setpgid(child, pgid) < 0 )
-		err(1, "setpgid of child into pgid");
-	// Time out if the process hangs inside waitpid.
-	signal(SIGALRM, on_signal);
-	alarm(2);
-	int status;
-	// waitpid is supposed to ECHLD here when the child changes its pgid.
-	pid_t result = waitpid(-pgid, &status, 0);
-	if ( result < 0 )
-		err(1, "waitpid");
-	else if ( result == 0 )
-		errx(1, "waitpid() == 0");
-	else if ( result == child )
-		errx(1, "waitpid returned child");
-	else
-		errx(1, "waitpid returned strange child");
 	return 0;
 }
