diff --git figlet-clean/figlet.c figlet-workdir/figlet.c
index 3de795a..afcbc5d 100644
--- figlet-clean/figlet.c
+++ figlet-workdir/figlet.c
@@ -65,7 +65,7 @@
 #include <ctype.h>
 #include <sys/stat.h>
 #include <fcntl.h>     /* Needed for get_columns */
-
+#include <unistd.h>
 #ifdef unix
 #include <unistd.h>
 #include <sys/ioctl.h> /* Needed for get_columns */
diff --git figlet-clean/utf8.h figlet-workdir/utf8.h
index 0631b8a..bcab052 100644
--- figlet-clean/utf8.h
+++ figlet-workdir/utf8.h
@@ -27,13 +27,9 @@
 #define UTF8_IGNORE_ERROR		0x01
 #define UTF8_SKIP_BOM			0x02
 
-__BEGIN_DECLS
-
 size_t		utf8_to_wchar(const char *in, size_t insize, wchar_t *out,
 		    size_t outsize, int flags);
 size_t		wchar_to_utf8(const wchar_t *in, size_t insize, char *out,
 		    size_t outsize, int flags);
 
-__END_DECLS
-
 #endif /* !_UTF8_H_ */
