diff --git tar-clean/configure.ac tar-workdir/configure.ac
index 338e183..952dc42 100644
--- tar-clean/configure.ac
+++ tar-workdir/configure.ac
@@ -21,7 +21,7 @@ AC_INIT([GNU tar], [1.35], [bug-tar@gnu.org])
 AC_CONFIG_SRCDIR([src/tar.c])
 AC_CONFIG_AUX_DIR([build-aux])
 AC_CONFIG_HEADERS([config.h])
-AC_PREREQ([2.71])
+#AC_PREREQ([2.71])
 AM_INIT_AUTOMAKE([1.15 gnits tar-ustar dist-bzip2 dist-xz std-options silent-rules])
 
 # Enable silent rules by default:
