diff --git graphite2-clean/tests/featuremap/featuremaptest.cpp graphite2-workdir/tests/featuremap/featuremaptest.cpp
index 7ba52a7..8033d80 100644
--- graphite2-clean/tests/featuremap/featuremaptest.cpp
+++ graphite2-workdir/tests/featuremap/featuremaptest.cpp
@@ -20,6 +20,7 @@
     internet at http://www.fsf.org/licenses/lgpl.html.
 */
 #include <cassert>
+#include <cstdint>
 #include <cstdlib>
 #include <fstream>
 #include <iostream>
